{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/CAGEF_services_slide.png?raw=true\"  width=700></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Python for Data Science\n",
    "\n",
    "# Lecture 04: Exploratory data analysis and plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Student Name:\n",
    "\n",
    "## Student ID:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.1.0 About Introduction to Python\n",
    "\n",
    "Introduction to Python is brought to you by the **Centre for the Analysis of Genome Evolution & Function (CAGEF)** bioinformatics training initiative. This course was developed based on feedback on the needs and interests of the Department of Cell & Systems Biology and the Department of Ecology and Evolutionary Biology.\n",
    "\n",
    "The structure of this course is a code-along style so it is 100% hands on! A few hours prior to each lecture, the materials will be avaialable for download at [QUERCUS](https://q.utoronto.ca/). The teaching materials will consist of a Jupyter Lab Notebook with concepts, comments, instructions, and blank spaces that you will fill out with Python code along with the instructor. Other teaching materials include a live version of the notebook, and datasets to import into Python - when required. This learning approach will allow you to spend the time coding and not taking notes! \n",
    "\n",
    "As we go along, there will be some in-class challenge questions for you to solve either individually or in cooperation with your peers. Post lecture assessments will also be available (see syllabus for grading scheme and percentages of the final mark) through [DataCamp](https://Datacamp.com) to help cement and/or extend what you learn each week.\n",
    "\n",
    "### 0.1.1 Where is this course headed?\n",
    "\n",
    "We'll take a blank slate approach here to Python and assume that you pretty much know _nothing_ about programming. From the beginning of this course to the end, we want to get you from some potential scenarios:\n",
    "\n",
    "- A pile of data (like an excel file or tab-separated file) full of experimental observations and you don't know what to do with it. \n",
    "\n",
    "- Maybe you're manipulating large tables all in excel, making custom formulas and pivot table with graphs. Now you have to repeat similar experiments and do the analysis again. \n",
    "\n",
    "- You're generating high-throughput data and there aren't any bioinformaticians around to help you sort it out.\n",
    "\n",
    "- You heard about Python and what it _could_ do for your data analysis but don't know what that means or where to start.\n",
    "\n",
    "and get you to a point where you can:\n",
    "\n",
    "- Format your data correctly for analysis\n",
    "\n",
    "- Produce basic plots and perform exploratory analysis\n",
    "\n",
    "- Make functions and scripts for re-analysing existing or new data sets\n",
    "\n",
    "- Track your experiments in a digital notebook like Jupyter!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1.2 How do we get there? Step-by-step.\n",
    "\n",
    "In the first two lessons, we will talk about the basic data structures and objects in Python, get cozy with the Jupyter Notebook environment, and learn how to get help when you are stuck. Because everyone gets stuck - a lot! Then you will learn how to get your data in and out of Python, how to tidy our data (data wrangling), subset and merge data. We'll take a break from data wrangling to spend our fourth lecture learning how to generate exploratory data plots. Then we'll slide into using the power of python and programming to use flow control before visiting text manipulation techniques in lectures 5 and 6. Data cleaning and string manipulation is really the battleground of coding - getting your data into the format where you can analyse it. Lastly, we will learn to write customized functions to help scale up your analyses. <br><br>\n",
    "\n",
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/Draw_an_Owl.jpg?raw=true\"  width=700></center>\n",
    "\n",
    "The structure of the class is a __code-along__ style: It is fully hands on. At the end of each lecture, the complete notes will be made available in a PDF format through the corresponding Quercus module so you don't have to spend your attention on taking notes. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### 0.1.3 What kind of coding style will we learn?\n",
    "\n",
    "There is no single path correct from A to B - although some paths may be more elegant, or more efficient than others. With that in mind, the emphasis in this lecture series will be on:\n",
    "\n",
    "1. **Code simplicity** - learn helpful functions that allow you to focus on understanding the basic tenets of good data wrangling (reformatting) to facilitate quick exploratory data analysis and visualization. \n",
    "2. **Code readability** - format and comment your code for yourself and others so that even those with minimal experience in R will be able to quickly grasp the overall steps in your code.\n",
    "3. **Code stability** - while the core Python code is relatively stable, behaviours of functions can still change with updates. There are well-developed packages we'll focus on for our analyses. Namely, we'll become more familiar with the `pandas` series of packages for working with tabular data. This resource is well-maintained by a large community of developers. While not always the \"fastest\" approach this additional layer can help ensure your code still runs (somewhat) smoothly later down the road."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.2.0 Lecture objectives\n",
    "\n",
    "Welcome to this fourth lecture in a series of seven. Today we will pick up where we left off last week with our merged data. We'll learn how to explore the data, summarize it, and plot it!\n",
    "\n",
    "At the end of this lecture we will aim to have covered the following topics:\n",
    "\n",
    "1. Exploratory data analysis with long-format data\n",
    "2. Data plotting with the `matplotlib.pyplot` package.\n",
    "3. Advanced visualizations with the `seaborn` package.\n",
    "4. Saving visualizations to file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.3.0 A legend for text format in Jupyter markdown\n",
    "\n",
    "`grey background` - a package, function, code, command or directory. Backticks are also use for in-line code.      \n",
    "*italics* - an important term or concept or an individual file or folder     \n",
    "**bold** - heading or a term that is being defined      \n",
    "<span style=\"color:blue\">blue text</span> - named or unnamed hyperlink\n",
    "\n",
    "`...` - Within each coding cell this will indicate an area of code that students will need to complete for the code cell to run correctly. \n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Blue box:</b> A key concept that is being introduced\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Yellow box:</b> Risk or caution\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Green boxes:</b> Recommended reads and resources to learn Python\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>Red boxes:</b> A comprehension question which may or may not involve a coding cell. You usually find these at the end of a section. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.4.0 Lecture and data files used in this course\n",
    "\n",
    "### 0.4.1 Weekly Lecture and skeleton files\n",
    "\n",
    "Each week, new lesson files will appear within your JupyterHub folders. We are pulling from a GitHub repository using this [Repository git-pull link](https://jupyter.utoronto.ca/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2Fuoft-csb-datasci%2F2026-01-IntroPython&urlpath=tree%2F2026-01-IntroPython%2F&branch=main). Simply click on the link and it will take you to the University of Toronto JupyterHub. You will need to use your UTORid credentials to complete the login process. From there you will find each week's lecture files in the directory `/2026-01-IntroPython/Lecture_XX`. You will find a partially coded `skeleton.ipynb` file as well as all of the data files necessary to run the week's lecture.\n",
    "\n",
    "Alternatively, you can download the Jupyter Notebook (`.ipynb`) and data files from JupyterHub to your personal computer if you would like to run independently of the JupyterHub.\n",
    "\n",
    "### 0.4.2 Live-coding HTML page\n",
    "\n",
    "A live lecture version will be available at [uoft-csb-datasci.github.io](https://uoft-csb-datasci.github.io/2026-01.IntroPython/index.html) that will update as the lecture progresses. Be sure to refresh to take a look if you get lost!\n",
    "\n",
    "### 0.4.3 Post-lecture PDFs\n",
    "\n",
    "As mentioned above, at the end of each lecture there will be a completed version of the lecture code released as a PDF file under the Modules section of Quercus.\n",
    "\n",
    "***\n",
    "### 0.4.4 Microsporidia infection data set description\n",
    "\n",
    "The following datasets used in this week's class come from a published manuscript on PLoS Pathogens entitled \"High-throughput phenotyping of infection by diverse microsporidia species reveals a wild _C. elegans_ strain with opposing resistance and susceptibility traits\" by [Mok et al., 2023](https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1011225). These datasets focus on the an analysis of infection in wild isolate strains of the nematode _C. elegans_ by environmental pathogens known as microsporidia. The authors collected embryo counts from individual animals in the population after population-wide infection by microsporidia and we'll spend our next few classes working with the dataset to learn how to format and manipulate it. \n",
    "\n",
    "### 0.4.4.1 Dataset 1: /data/embryo_long_merged.csv\n",
    "\n",
    "This is a result of our efforts (mostly) from last lecture. After transforming a wide-format version of our measurement data, we merged it with some metadata regarding our experiments and now it is ready to be visualized!\n",
    "\n",
    "### 0.4.4.2 Dataset 2: /data/infection_signal.tsv\n",
    "\n",
    "This is an imaging analysis of infected _C. elegans_ strains N2 and JU1400 measuring the overall number of pixels for each animals and the number of fluorescent (infected) pixels within the same area. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.5.0 Packages used in this lesson\n",
    "\n",
    "`IPython` and `InteractiveShell` will be access just to set the behaviour we want for iPython so we can see multiple code outputs per code cell.\n",
    "\n",
    "`numpy` provides a number of mathematical functions as well as the special data class of `arrays` which we'll be learning about today.\n",
    "\n",
    "`pandas` is built upon the `numpy` package and extends the capabilities to work with tabular data\n",
    "\n",
    "`matplotlib` is a package used to work with data in the style of the language MATPLOT\n",
    "\n",
    "`seaborn` is built upon a portion of the `matplotlib` to make plotting data more accessible and simplify it from a layered grammar of graphics perspective."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----- Always run this at the beginning of class so we can get multi-command output ----- #\n",
    "\n",
    "# Access options from the iPython core\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "\n",
    "# Change the value of ast_node_interactivity\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "# ----- Additional packages we want to import for class ----- #\n",
    "\n",
    "# Import the pandas package\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# seaborn should already be installed on Jupyter Hub\n",
    "# !pip install seaborn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 1.0.0 An introduction to exploratory data analysis\n",
    "\n",
    "Recall from last week, we spent our lecture converting data from a series of worm/pathogen interactions which contained observations about the formation of spores, meronts, and embryos within individual animals in an infected population. The process involved conversion from a wide-format to long-format along with the merging of this data to a set of metadata. \n",
    "\n",
    "Now that the wrangling is completed, we can perform some exploratory data analysis (EDA). The process of EDA investigates your data to identify abnormalities, summarize its main characteristics and identify potential patterns or trends for further validation. We did some initial statistical summarization on the numerical and non-numerical data but today we'll dig deeper using some additional tools in our Python pockets.\n",
    "\n",
    "With our EDA today we will try to answer questions like:\n",
    "\n",
    "1. Which worm strain or pathogen strain is most often measured or observed?\n",
    "2. What is the mean number of embryos produced by uninfected animals in our study? \n",
    "3. How variable is that mean across different sets of replicate experiments?\n",
    "4. How many pathogen strains have been tested on each worm strain?\n",
    "\n",
    "### 1.0.1 Import our data from last week\n",
    "\n",
    "Let's open a version of our dataset from last week. We'll find it in the file `data/embryo_long_merged.csv`. Recall that it is always good practice to explore your data to find out more about aspects such as probability distributions, outliers, and central tendency/dispersion measures.\n",
    "\n",
    "Now we are going to do some EDA on `embryo_long_merged.csv` from last week's wrangling. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in embryo_long_merged.csv\n",
    "embryo_merged = pd.read_csv(...) \n",
    "\n",
    "# Take a peek at the data\n",
    "embryo_merged.head()\n",
    "\n",
    "# How big is this dataset?\n",
    "embryo_merged..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "So as a reminder, we've imported a dataset that is 11,049 rows with 30 columns of data. At this point, the measurement data exists in only 3 columns: `merontsPresent`, `sporesPresent`, and `numEmbryos`.\n",
    "\n",
    "Before we go any further, however, let's drop some of the extraneous metadata that we won't need for our analyses. Pretty much everything from `Infection Round` onwards is unnecessary. From our call to `.info()` we can see that we just need the first 19 columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the data from our dataset for the first 19 columns\n",
    "embryo_merged_subset = embryo_merged.iloc[...]\n",
    "\n",
    "# What does the subset look like?\n",
    "embryo_merged_subset.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.1.0 Which worm and pathogen strains are most often used in our experiments?\n",
    "\n",
    "Let's begin with a deceptively simple question about our data. As we'll see, however, it requires more than a simple method call to our data in order to discern. To get there, we'll walk through the thought process so you can avoid potential pitfalls later in your own analyses.\n",
    "\n",
    "### 1.1.1 Get a summary of our information\n",
    "\n",
    "There is a lot of subgrouped data hidden with our dataset. Experiments are classified by their `date`, `wormStrain`, `pathogenStrain`, `pathogenDose`, and `timepoint`. The combination of all five of these can also be found in the `experiment` column, although that combination may not always be useful to us.\n",
    "\n",
    "Let's begin with the `.describe()` method to review any numerical data that we can. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a description of the numeric data\n",
    "embryo_merged_subset..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.2 Use the `.describe()` method to summarize non-numeric columns\n",
    "\n",
    "Recall that we can also summarize our non-numeric data, to a certain extent, as long as we provide it properly to the `describe()` method. This method can identify the \"top\" occuring entry in a column as well as it's frequency. We already know that the `wormStrain` column contains the strain information on each individual worm measured in our data. The same goes for `pathogenStrain` which will contain some similar information about our pathogens used. It should be simple enough to just create a summary of that information. Let's give it a try."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the describe method on the OTU column from our merged_subset\n",
    "embryo_merged_subset.loc[:,['wormStrain', ...]].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** \n",
    "At a quick glance, we can answer our first question - that the N2 animal is the most measured one in our infection studies while the pathogen LUAm1 is the most measured pathogen in our study. There are a few catches to these results, however:\n",
    "\n",
    "1. Although N2 is the most-often measured ___individual animal___, this is biased by the fact that it acts as a control strain in many experiments. If we were to look at specific ___groups___ of experiments and how often N2 was included, would it still be the most prevalent?\n",
    "\n",
    "2. Our measurements might be comprised of entries where LUAm1 is a `pathogenStrain`, but it's `pathogenDose` is 0! So it's not really being used to infect at all. \n",
    "\n",
    "Let's address the 2nd question and circle back around to the first in a little bit as it is slightly more complex."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.3 Filter your data with conditional booleans\n",
    "\n",
    "Last lecture we saw a few examples where we could subset our data using a simple conditional statement with a method like `.isna()`. While we referred to this as \"slicing\" our data, you can also think of it as a way to ___filter___ data. We can of course, filter our data using other conditional statements and before summarizing your data, you should consider the nature of your values! In our dataset it appears that the `pathogenDose` can denote whether or not an animal was exposed to a pathogen or was \"mock-infected\". A mock-infected sample would have a `pathogenDose` value of 0.\n",
    "\n",
    "To solve our issues with summarizing the pathogen usage conundrum we turn to filtering our data by the `pathogenDose` column. We'll do this in two steps:\n",
    "\n",
    "1. Filter the dataset by a conditional.\n",
    "2. Pass that dataset on for summarization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recall that we can broadcast a conditional query to multiple values in a DataFrame\n",
    "embryo_merged_subset..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Oops! Look at the final error message: __'>' not supported between instances of 'str' and 'int'__ .The `pathogenDose` column is a string! We need to fix that up first and convert it from values like \"0M\" to a float value like \"0.0\".\n",
    "\n",
    "Recall we have at our disposal the following methods:\n",
    "- `.pop()`: updates the `DataFrame` object and returns a `Series` object.\n",
    "\n",
    "- `Series.str.split()`: returns a `DataFrame` object if `expand = True`\n",
    "\n",
    "- `.astype()`: returns a `DataFrame` object\n",
    "  \n",
    "- `.insert()`: returns nothing BUT updates a data frame object. Note that this method's parameters include\n",
    "    - `loc`: the index position to insert at (causing it to become the variable at that position)\n",
    "    - `column`: the name of your new column\n",
    "    - `value`: the values that comprise the column you are insert.\n",
    "    - `allow_duplicates`: boolean variable (`False` by default) which will allow you to insert a column with a label that already exists. Note that doing so can be problematic as it does not replace the pre-existing column.\n",
    " \n",
    "We'll use those now, to fix the `pathogenDose` column and replace it in our `embryo_merged_subset`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the insert method\n",
    "embryo_merged_subset.insert(loc = 4,                   # Location to insert at\n",
    "                            column = 'pathogenDose',   # The \"new\" column name to insert\n",
    "                            \n",
    "                            # To calculate the value, we'll pop from the subset (which removes the chosen column!)\n",
    "                            value = (\n",
    "                                embryo_merged_subset.pop(...)\n",
    "                                    \n",
    "                                    # Break up the float from the \"M\" which will make 2 columns\n",
    "                                    .str.split(pat = \"M\", expand = True)\n",
    "                                    \n",
    "                                    # Convert the first column to a float, and then provide only that to insert                                    \n",
    "                                    .astype({0:...})[0]    # The [0] index slices our DataFrame returns just the first column as a Series\n",
    "                            ) # Our corrected column can now be inserted\n",
    "                           )\n",
    "                           \n",
    "# Check on the updated subset data\n",
    "embryo_merged_subset.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "We can see that the `pathogenDose` column is now classified as a `float64` datatype. Let's check on that conditional filtering again!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we can use a conditional comparison on the pathogenDose column\n",
    "embryo_merged_subset.loc[:, 'pathogenDose'] > 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Now we need to summarize the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supply the result of the conditional query as a filtering criteria\n",
    "print(\"Filtered pathogenStrain summary\")\n",
    "embryo_merged_subset.loc[..., # remember we're filtering by rows!\n",
    "                         # We only need the pathogenStrain column to summarize\n",
    "                         'pathogenStrain'   # No list notation means we'll get a Series back \n",
    "                        ].describe()\n",
    "\n",
    "# Note the \\n adds a blank line to our output\n",
    "print(\"\\nUnfiltered pathogenStrain summary\")\n",
    "\n",
    "# Compare that to the unfiltered subset\n",
    "embryo_merged_subset.loc[:, 'pathogenStrain'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.4 The `.describe()` method returns a Series object\n",
    "\n",
    "A useful aspect of the `.describe()` method is that it also returns a `Series` object, which means its information can be retrieved or saved for further use! Recall we can choose from the possible index names. In the case of non-numeric summaries, we can use `count`, `unique`, `top`, or `freq`. These are also 0-indexed!\n",
    "\n",
    "Therefore we can use the results of `.describe()` to further filter our data along different variables. For instance, we can now ask the question \"Among the most observed pathogen used, which worm strain (host) is most observed?\" Let's answer that question now!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for experiments with actual pathogen exposure\n",
    "pathogen_Summary = embryo_merged_subset.loc[embryo_merged_subset.loc[:, 'pathogenDose'] > 0,\n",
    "                         # We only need the pathogenStrain column to summarize\n",
    "                         'pathogenStrain'].describe()\n",
    "\n",
    "# Of the LUAm1 infections, which worm strain reigns supreme?\n",
    "embryo_merged_subset.loc[(embryo_merged_subset[\"pathogenStrain\"] == ...), # Filter rows by the \"top\" pathogenStrain \n",
    "                         'wormStrain'].describe()                                          # Keep the wormStrain column to summarize it"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.5 Know the difference between filtering and the `.filter()` method\n",
    "\n",
    "You may be wondering to yourself, surely there must be a `.filter()` method implemented with the DataFrame. It seems like such an essential part of working with DataFrames. You're right that such a method exists but you would be wrong to think that it is used for conditional filtering. \n",
    "\n",
    "The `.filter()` method is used merely for subsetting your DataFrame by ___selecting___ on column name or by regular expression pattern (See __Lecture 06__). __It does not use conditional logic to subset rows from your data__. While this can be helpful in certain contexts, it does NOT implement the idea of filtering rows of our data based on conditional criteria.\n",
    "\n",
    "Here's a quick example of how to use it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the wormStrain, pathogenStrain, and numEmbryos columns\n",
    "embryo_merged_subset...(items = ['wormStrain', 'pathogenStrain', 'numEmbryos'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.6 Use the `.query()` method to conditionally filter your data\n",
    "\n",
    "Thus far we have seen the use of boolean logic to help produce conditional filtering by selecting on the resulting arrays with the `.loc[]` or `.iloc[]` methods. However, you might have noticed that the code is rather clunky:\n",
    "\n",
    "`embryo_merged_subset.loc[embryo_merged_subset.loc[:, 'pathogenDose'] > 0, 'pathogenStrain']`\n",
    "\n",
    "As an alternative, we can use the `.query()` method which has its roots in database querying language. This method has two main parameters:\n",
    "- `expr`: a string object which encapsulates the filter query you want to use. This is defined within a set of __single-quotes__ ie `'pathogenDose > 0'`.\n",
    "    - When defining column names with spaces, use the back-tick (grave accent) character (\\`)\n",
    "    - When filtering a column using a string value, surround these with double-quotes (__\"__).\n",
    "    - When refering to a variable in your environment prefix with the `@` symbol.\n",
    "- `inplace`: a boolean on whether or not to replace the `DataFrame` and produce no output (`True`) or pass on the new `DataFrame` (`False`, default).\n",
    "\n",
    "Let's repeat our previous filtering from __section 1.1.3__ using `.query()` and method chaining."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter to see which pathogen strain is most often measured.\n",
    "# We'll encapsulate the call in () so we can separate across lines\n",
    "\n",
    "# Filter the dataset for observations with actual infection\n",
    "(embryo_merged_subset.query(...)\n",
    "                     # Select just pathogenStrain as a column\n",
    "                     .filter(['pathogenStrain'])\n",
    "                     # Get a summary\n",
    "                     .describe()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Our query from __section 1.1.4__ asked \"Among the most observed pathogen used, which worm strain (host) is most observed?\". We created an intermediate variable `pathogen_Summary` to help answer that question and we can still use the `.query()` method to simplify the process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To complete our query, we use the @ to identify our intermediate variable\n",
    "(embryo_merged_subset.query('pathogenStrain == ...')\n",
    "                     # Filter on the column we want to use\n",
    "                     .filter([\"wormStrain\"])\n",
    "                     # Summarize\n",
    "                     .describe()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.2.0 Which worm strain is most often included in our different experimental groups?\n",
    "\n",
    "Circling back to our first question, we discovered that __N2__ is the worm strain that is most often _measured_ across all of our datasets, but the dataset contains groups of different infection experiments. __<u>Is N2 the most prevalent strain when we look at its inclusion within individual experiments?</u>__ How do we explore this?\n",
    "\n",
    "### 1.2.1 Grouping your DataFrame using the `.groupby()` method\n",
    "\n",
    "Thinking about our problem, we already have an identifier that breaks down each experimental grouping - `date`. Each different `date` essentially sets up a different experimental replicate. We could use our newly-taught filtering techniques, but we would have to cycle through each potential value and summarize the subset. (Honestly this would have been my approach when first learning!).\n",
    "\n",
    "Luckily for you, the `.groupby()` method can ___sort all___ that data for you based on the criteria you provide. The important parameters to us today are:\n",
    "\n",
    "- `by`: A function, label, or list of labels that you want to use to determine grouping criteria. You can even provide a dictionary object where specific key:values pairs can be used to determine groupings.\n",
    "- `axis`: Whether to split along the rows (0, default) or columns (1)\n",
    "- `as_index`: A boolean to determine if the index labels should be based on group labels (True by default)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many individual dates/replicates are there?\n",
    "len(embryo_merged_subset.loc[:,'date'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group our subset data by the 'date' column\n",
    "embryo_merged_subset...\n",
    "\n",
    "# What is returned to us?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.2.2 Use the `.head()` method to view rows from each group\n",
    "\n",
    "As you can see above, we create a `DataFrameGroupBy` object but if we attempted to look at it, it would look pretty much like the original `merged_subset`. The major difference is that the data has now been essentially sorted by the `date` column. In order to view part of it, we can use the `head(n)` method which will return `n` rows from ___each___ group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group our subset data by the 'date' column and view 1 row from each group\n",
    "embryo_merged_subset.groupby(by = ['date'])..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Using the `.groupby()` method and `.head()`, we can view a representative observation from each group. To answer our above question, however, we need more information. If we repeat our above code but _also_ include `wormStrain` and `pathogenStrain` columns in our grouping, what will that produce?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group our subset data by the 'date' and 'wormStrain' columns and grab the first row from each\n",
    "# How big is the result?\n",
    "embryo_merged_subset.groupby(by = ['date', 'wormStrain', ...]).head(1).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.2.3 Subset and apply functions to your grouped DataFrame\n",
    "\n",
    "From our results, we see there are 99 separate combinations of `date` : `wormStrain` : `pathogenStrain` in our grouped DataFrame! Now that we have our groups we can begin to apply functions to summarize data from each group. \n",
    "\n",
    "Back to our question, we can now ask, __which wormStrain produces the largest number of `date`:`wormStrain`:`pathogenStrain` combinations__. \n",
    "\n",
    "We've already used some functions like `unique()` but other helpful functions to apply include `sum()`, `max()`, `min()` and `median()`. Functions like `idxmin()` and `idxmax()` will return the _index_ of the min and max values respectively but only the _first occurence_ of the value being sought.\n",
    "\n",
    "In this case what we really need is something that can help produce a frequency table. The `.value_counts()` method can be applied to a `Series` to produce such a table. We just need to provide it with the column(s) we'd like to use with the `subset` parameter. \n",
    "\n",
    "In this case, we are interested in `wormStrain` so let's see what that looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the \"wormStrain\" column from our grouped data to a frequency table\n",
    "(embryo_merged_subset.groupby(by = ['date', 'wormStrain', 'pathogenStrain'])\n",
    "                     # head(1) gets us a representative row from each group\n",
    "                     .head(1)\n",
    "                     # Create the frequency table of those rows we've isolated\n",
    "                     .value_counts(...)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "So, from our groupings it looks like N2 edges out JU1400 by just a single group. That's pretty close but now we've answered our question!\n",
    "\n",
    "## 1.3.0 What is the mean number of embryos produced by uninfected animals in our study? \n",
    "\n",
    "Let's approach this question by identifying our criteria:\n",
    "\n",
    "1. Filtering __for__ uninfected animals\n",
    "2. Measuring _mean_ embryos across all data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the mean embryo value of uninfected strains\n",
    "\n",
    "# Subset by uninfected animals\n",
    "(embryo_merged_subset.query('pathogenDose == 0')\n",
    " \n",
    "                     # Group by wormStrain and take the numEmbryos columns\n",
    "                     .groupby(by = ['wormStrain'])[...]\n",
    "                     \n",
    "                     # Calculate the mean numEmbryos for each group\n",
    "                     ...\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.3.1 Use `sort_values()` to organize your data\n",
    "\n",
    "It looks like we get the results we're looking for but the data is sorted by alphabetical _index_ order. What if we're interested, however, in finding the highest and lowest values in our dataset? Since it's small, a quick and easy way to ascertain this information is with the `sort_values()` function which will sort by ascending order. \n",
    "\n",
    "If we had a multi-column DataFrame, we could use the `by` parameter to select multiple columns to sort with.\n",
    "\n",
    "To sort in descending or reverse order, we can set the `ascending = False` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify the mean embryo value of uninfected strains AND sort them!\n",
    "\n",
    "# Subset by uninfected animals\n",
    "(embryo_merged_subset.query('pathogenDose == 0')\n",
    " \n",
    "                     # Group by wormStrain and take the numEmbryos colums\n",
    "                     .groupby(by = ['wormStrain'])['numEmbryos']\n",
    "                     \n",
    "                     # Calculate the mean numEmbryos for each group\n",
    "                     .mean()\n",
    "                     \n",
    "                     # Sort the data by descending order\n",
    "                     .sort_values(...)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "We now have an answer to our question and have identified the mean number of embryos per uninfected animal in each strain. This EDA gets us a baseline value for each strain that can be used in later comparisons!\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Section 1.3.1 Comprehension Question:</b> Among all of the mean embryo values we've generated, how do we determine the <b>median value</b> in the group of means? Which strain does it belong to? Modify the code below to determine the answer.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comprehension code answer 1.3.1\n",
    "# Identify the mean embryo value of uninfected strains AND sort them!\n",
    "\n",
    "# Subset by uninfected animals\n",
    "(embryo_merged_subset.query('pathogenDose == 0')\n",
    " \n",
    "                     # Group by wormStrain and take the numEmbryos columns\n",
    "                     .groupby(by = ['wormStrain'])['numEmbryos']\n",
    "                     \n",
    "                     # Calculate the mean\n",
    "                     .mean()\n",
    "                     \n",
    "                     # How do we calculate the median?\n",
    "                     ...\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comprehension Question 1.3.1 Answer:\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.4.0 What is the dispersion of the <u>uninfected mean</u> across different sets of replicate experiments?\n",
    "\n",
    "Now that we've got a few very helpful tools under our belt, we can take our query to the next level and ask what the mean and standard deviation of any individual worm strain is across multiple replicates. \n",
    "\n",
    "### 1.4.1 Plan your analysis strategy to avoid mistaken assumptions\n",
    "\n",
    "Time to think about your dataset in relationship to your question. We already know that \n",
    "- each worm strain may appear within any infection experiment\n",
    "- a dose of 0 represents uninfected animals.\n",
    "\n",
    "Again, it will be important to filter our data before summarizing it. Next you must understand ___which groupings___ you are looking for and ___what measurements___ you'd like to summarize. Let's break down the problem:\n",
    "\n",
    "1. Filter __for__ uninfected animals.\n",
    "2. Group data by `date`, `wormStrain`, and `pathogenStrain`. This will subgroup data into infection replicates!\n",
    "3. Calculate the mean number of embryos.\n",
    "4. Summarize the data grouped by wormStrain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine the standard deviation of the mean embryo counts across each strain\n",
    "# IN the uninfected state - ie a baseline embryo count.\n",
    "\n",
    "# Query for only uninfected data\n",
    "(embryo_merged_subset.query('pathogenDose == 0')\n",
    " \n",
    " # Group by infection experiment \n",
    " .groupby(by = ['date', 'wormStrain', 'pathogenStrain'])\n",
    " \n",
    " # Isolate numEmbryos and generate a mean for each group\n",
    " # Indexing like this returns a SeriesGroupBy object\n",
    " ['numEmbryos']\n",
    " .mean()\n",
    " \n",
    " # Group the series of means again by wormStrain\n",
    " .groupby(...)\n",
    " \n",
    " # Summarize each group of numEmbryo means\n",
    " .describe()\n",
    " \n",
    " # Sort the data by descending order\n",
    " .sort_values(by = 'mean', ascending = False)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Notice the presence of `NaN` values in our standard deviation columns? Can you tell why this is the case? What is the relationship between all of the strains with such a value?\n",
    "\n",
    "## 1.5.0 How many pathogens have been tested on each strain?\n",
    "\n",
    "Circling back again towards our initial questions, one in a similar vein would be to see how many pathogens have been tested on each strain. Let's remember to plan our analysis:\n",
    "\n",
    "1. We only want to examine data where the `pathogenDose` is >0\n",
    "2. We want to examine data based on each individual worm strain\n",
    "3. We are looking to identify the list of unique pathogen strains tested on each grouping\n",
    "4. We want to determine the _size_ of each of those lists.\n",
    "\n",
    "Let's start with the first 3 criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Isolate pathogen strain information in each group\n",
    " ['pathogenStrain']\n",
    " \n",
    " # Determine the unique values in each\n",
    " ...\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "You can see from our above results that we have generated a Series object, where the pathogen strains associated with each worm strain are stored as a `list`-like object which turns out to be an `np.array`. We know that we can extract the `.size` property from those objects so that should get us our answer!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Isolate pathogen strain information in each group\n",
    " ['pathogenStrain']\n",
    " \n",
    " # Determine the unique values in each\n",
    " .unique()\n",
    " \n",
    " # Extract the size of each array\n",
    " ...\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "__Uh oh__, just a single number - 18. That's actually how many _worm strains_ we had in the Series object generated by the call to `.unique()`. What we wanted, instead, was the size of ___each___ element in the Series. How do we obtain that from what we have so far?\n",
    "\n",
    "### 1.5.1 Use the `.apply()` method to broadcast a function to individual elements\n",
    "\n",
    "We haven't spent much time discussing this about DataFrames, but you may recall that `np.array` objects can perform element-wise arithmetic. The general term for this ability is called ___broadcasting___. In the case of DataFrames, we can perform similar functions. \n",
    "\n",
    "We can:\n",
    "- set the value of an entire dataset to a static value\n",
    "- Use specific values/criteria to identify subsets of cells and modify those (via booleans!)\n",
    "- apply pre-supplied mathematical functions to elements\n",
    "- `apply()` our own custom funcitons to elements\n",
    "\n",
    "The `apply()` method takes the form of `apply(func, args=(), **kwargs)`. We'll talk a little more about some of these parameters in later lectures but for now we have:\n",
    "\n",
    "- `func`: the name of the function you want to use\n",
    "- `args`: a tuple of any additional arguments that are needed for `func` to work.\n",
    "\n",
    "Note that the `.apply()` method is ___object-dependent___ and will work differently depending on the `pd.object` for which it is being called. In our case, since we have a `Series` object, it will apply the `.size` attribute to __each entry__ (array) in our `Series`. \n",
    "\n",
    "You'll see one more unfamiliar piece of syntax: `lambda`. This is how we can tell Python we want to make a \"quick function\" on the spot. We'll cover this more in later lectures as well __but__ simply put, it removes the need for a formal declaration of a function (__Lecture 7__). For now, we'll just roll with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Isolate pathogen strain information in each group\n",
    " ['pathogenStrain']\n",
    " \n",
    " # Determine the unique values in each\n",
    " .unique()\n",
    " \n",
    " # Extract the size of each array using a temporary function\n",
    " .apply(...)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.5.2 Use `.nunique()` on a grouped dataframe to return the number of unique elements\n",
    "\n",
    "Looking at our code above, we went through 5 steps to get a final answer:\n",
    "1. Query the data.\n",
    "2. Group the data.\n",
    "3. Isolate the column we want to analyse for each group.\n",
    "4. Determine the unique values in the data.\n",
    "5. Retrieve the size of each array element.\n",
    "\n",
    "What if instead, we used a helpful method - `.nunique()` to count the number of unique elements in our grouped dataframe. This method simplifies the process by combining the `unique()` and `len()` methods together (which we have used previously to achieve the same goal). Furthermore you can ignore the `NA` values by default. Using this method simplifies our process slightly into just __3 steps of code__ as we'll see."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Determine the length of unique values in each\n",
    " ...\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "In 3 commands, we were able to determine not only the number of unique values in `pathogenStrain` but also the unique count __across all of the data columns__! What happens if we want to do more than just calculate a single value?\n",
    "\n",
    "### 1.5.3 Use the `.agg()` method to generate a summary from multiple functions\n",
    "\n",
    "Rather than using just a single method like `.nunique()` you can actually apply multiple functions to a grouped dataframe to generate a summary of the information. In this case we'll return to looking at the `pathogenStrain` column to simplify our output. We are also a little limited by the kind of summary we can achieve from string data. We can, however, still `count()` the number of entries in each group. \n",
    "\n",
    "To combine both of these methods to produce a summary, we'll use the `.agg()` method, which will accept a list of function names like \"count\" and \"nunique\" but also \"sum\", \"min\", \"max\" and other methods found in the `GroupBy` object. You can also use your own custom function just like the `.apply()` method.\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>How do I know what to use?</b> The GroupBy object has a number of it's own methods and the only way to get better at know what to use, is by studying the commands at your disposal. For a brief description of the various methods available, you can check out the <a href=\"https://pandas.pydata.org/docs/reference/groupby.html\">online Pandas reference manual</a>.\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Isolate pathogen strain information in each group\n",
    " ['pathogenStrain']\n",
    " \n",
    " # Determine the total number of values in each group, the number of unique, and what they are\n",
    " .agg([lambda x: ..., \n",
    "       \"count\", \n",
    "       \"nunique\", \n",
    "       \"unique\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.5.4 Use different methods for different variables with the `.agg()` method\n",
    "\n",
    "In our previous example we isolated the `pathogenStrain` column for the `.agg()` method but we can actually choose ___one or more columns___ to apply specific functions by using a `dictionary` object where the `key:value` pairs take on the form of `column_name:[func_1, ..., func_n]`.\n",
    "\n",
    "Let's modify our above code to gather the sum, mean and standard deviation of `numEmbryos` for each infected worm strain (regardless of pathogen), while also counting the number of animals measured for each strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Determine the total number of values in each group, the number of unique, and what they are\n",
    " .agg({'worm.number':..., \n",
    "       'numEmbryos':...})\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Notice the above output generates subgroupings _within_ our columns so each individual columns is multi-level in it's labeling! This is an important aspect when trying to pass around the various columns. To identify individual columns, they must be described as a `tuple` which takes the form of `('level_1', 'level_2', ... 'level_n')`. For example from our above output, the `std` column is actually labeled as `('numEmbryos', 'std')`. \n",
    "\n",
    "You can confirm this by adding a `.columns` call to our above code!\n",
    "\n",
    "### 1.5.5 Use more complex filtering with the `.query()` method\n",
    "\n",
    "Thus far we have limited our `query()` filtering to simply choosing between `pathogenDose == 0` or `pathogenDose >0`. What if we were also interested in only looking at a subset of our data for this analysis? The two most prevalent strains are N2 and JU1400. How can we filter for only that data in our analysis?\n",
    "\n",
    "Suppose we wanted to know the count, mean, and standard deviation in number of embryos for just N2 and JU1400 animals tested in mock infection replicates - basically retrieving their baseline uninfected states. We can do a slightly more complex filtering with a few more predicates. To accomplish this we'll revisit 2 keywords:\n",
    "- `in` this can search for membership of values within a specified `list`. ie `values in ['query1', 'query2']`\n",
    "- `&` this can be used to join predicates together so that one or more conditions must be met before returning a `True` value (more on this in __Lecture 5__). Normally, this is known as the \"bitwise AND\" which operates in a particular way, but for us, we'll use it as a \"logical AND\". More on this in **section 3.3.0**.\n",
    "\n",
    "In our case we want to meet 2 conditions: `wormStrain in ['N2', 'JU1400']` AND `pathogenDose == 0`. Let's see how we can join these into a single query call BUT look closely at how we use single and double-quotation marks in the `.query()` method!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Query for only infected data\n",
    "(embryo_merged_subset.query(...)\n",
    " \n",
    " # Group our data by worm strain\n",
    " .groupby('wormStrain')\n",
    " \n",
    " # Determine the number of observations in each group, and summary stats\n",
    " .agg({'worm.number':['count'], \n",
    "       'numEmbryos':['mean', 'std']})\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.5.6 Find the right tool to save yourself time\n",
    "\n",
    "Now that we've walked through a few analysis directions and introduced a number of great tools, you'll notice a basic pattern to most analyses:\n",
    "\n",
    "1. Filter data with `.query()`.\n",
    "2. Group data with `.groupby()`.\n",
    "3. Summarize data with premade functions separately or in combination with `.agg()` or with custom functions via `apply()`.\n",
    "\n",
    "After more practice and experience some of these functions will become second nature to your coding choices. \n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>Section 1.5.6 Comprehension Question:</b> Now that you know how to group data by specific variables, how would you calculate the number of groups created when combining the <b>pathogenStrain</b> and <b>pathogenDose</b> variables for actual (non-zero dose) infection experiments? Use code to output how many total worm strains have been tested in each scenario and sort for the largest number of unique entries.\n",
    "\n",
    "Use the code cell below to help generate an answer.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comprehension code answer 1.5.6\n",
    "\n",
    "# Query for only infected data\n",
    "(embryo_merged_subset.query('pathogenDose > 0')\n",
    " # Group our data by worm strain\n",
    " .groupby(by = ['pathogenStrain', 'pathogenDose'])\n",
    "\n",
    " # Determine the total number of unique worm strain values in each group, the number of unique, and what they are\n",
    " .agg(...)\n",
    " \n",
    " # Sort the values in nunique to simplify your search\n",
    " .sort_values(...)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 2.0.0 Plotting with the `pyplot` module\n",
    "\n",
    "### 2.0.1 Use exploratory plots to assess your data\n",
    "\n",
    "Now that we've had a chance to look at our data up close, let's talk about how we can use exploratory plots to give us a quick visual assessment of our data. We can use these visualizations to help make decisions about how to further analyse our data. Is there a difference between different groups of data? Does it look like there might be any bias between our datasets? What does the overall distribution of our sampling look like? \n",
    "\n",
    "### 2.0.2 Types of plots\n",
    "\n",
    "Often when trying to convey a message about our data through a visualization, we want to choose the _right kind_ of visualization. These visualizations can also be referred to as figures or plots. Within the `maplotlib` package is the `pyplot` module. It is a collection of functions that give the `matplotlib` package capabilities that are very similar to the programming language MATLAB. The `pyplot` module has functions that can create some of the following _basic_ plots:\n",
    "\n",
    "|Plot type|Command|What to use it for|\n",
    "|:-:|:-:|:-|\n",
    "|Bar plot / histogram|`bar()`|Population data summaries. Helpful for contrasting between groups|\n",
    "|Scatter plot| `scatter()`|Multiple independent measurements across different variables|\n",
    "|Line plot| `plot()`|Multiple measurements that represent the same sample(s)|\n",
    "|Histogram| `hist()`|Generate a distribution by binning your data|\n",
    "|Stem or lollipop| `stem()`| A twist on the bar plot that may be more compact and visually pleasing|\n",
    "|Boxplots| `boxplot()`| Create a visual summary of your datapoints based on their distribution|\n",
    "|Violin plots| `violinplot()`| Create a visual kernel density (distribution) estimate of datapoints|\n",
    "\n",
    "### 2.0.3 Plot components\n",
    "   \n",
    "Within each plot are a number of basic components: titles, axis properties, legends, etc. Here is a helpful table outlining _some_ of the basic plot components.\n",
    "\n",
    "|Component|Description|Command|Parameters|\n",
    "|:-:|:-|:-|:-|\n",
    "|Title|The title of your plot|`title()`| \n",
    "|X- or Y-axis title| The axis titles of your plot| `xlabel()`, `ylabel()`| xlabel=str | \n",
    "||||loc={'left', 'center', 'right'}| \n",
    "||||Text properties|\n",
    "|X- or Y-axis ticks| Alter your axis tick positions/locations and labels | `xticks()`, `yticks()`| ticks=[a,..n]|\n",
    "||||labels=[label1, ..., labeln]|\n",
    "|Axis limits| A list defining the x- and y-axis limits| `axis()`|[xMin, xMax, yMin, yMax]| \n",
    "|Axis scale| Set the kind of axis scale for your data|`xscale()`, `yscale()`|\"linear\", \"log\", \"symlog\", \"logit\"|\n",
    "|Text properties|Labels can take text parameters too||color, fontsize, fontstyle, rotation|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.1.0 Build a basic barplot with the `bar()` function\n",
    "\n",
    "We'll use our worm embryo values as an example to try and plot some of our data as a bar plot. The `bar()` method generally requires two sets of data to be supplied along with some optional data:\n",
    "\n",
    "- `x`: An array of x-coordinates (group labels, or x values)\n",
    "- `height`: A float or array of bar heights - these are usually the measured/summarized values\n",
    "- `width`: A float or array of bar widths (default is 0.8)\n",
    "- `bottom`: The y-coordinates of the bars (default is 0)\n",
    "- `align`: The alignment of the bars to your x-coordinate labels (default is `center`)\n",
    "\n",
    "Let's start by building a basic barplot examining the baseline mean number of embryos in our various worm strains. First we'll make a `DataFrame` object summarizing the relevant values. Then we'll build our barplot and see what needs to be altered as we move forward. Note that we use the `plt.show()` method to display our plot after putting a lot of pieces together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll reset the code cells to only show the last code call. This will de-clutter the plotting process for us\n",
    "InteractiveShell.ast_node_interactivity = \"last\"\n",
    "\n",
    "# Determine the mean number of embryos across all non-infected observations per strain\n",
    "wormStrain_mean_embryos = (embryo_merged_subset.query('pathogenDose == 0')\n",
    "                           \n",
    " # Group data by worm strains \n",
    " .groupby('wormStrain')\n",
    "                           \n",
    " # Get the mean value in each group\n",
    " .agg({'numEmbryos':'mean'})\n",
    "\n",
    " # Sort the data\n",
    " .sort_values(by = 'numEmbryos', ascending = False)\n",
    ")\n",
    "\n",
    "# Check on the results\n",
    "wormStrain_mean_embryos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build our barplot by giving the index as the x-label, and values as the height\n",
    "# We'll need to supply the data directly to each parameter\n",
    "plt.bar(x = ..., \n",
    "        height = ...)\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.1.1 Use the `figure()` function to set your figure size\n",
    "\n",
    "As we can see from our first attempt, the plot is rather small. We definitely have some problems with the basics of this plot and we'll address the first one that might help. This figure could be a little larger so we can see the x-axis labels better. Use the `figure()` method to set your figure size using the `figsize` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 2.1.1 Fix the size of the plot\n",
    "plt.figure(figsize = (...))\n",
    "\n",
    "# Build our barplot\n",
    "plt.bar(x=wormStrain_mean_embryos.index, \n",
    "        height=wormStrain_mean_embryos['numEmbryos'])\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.1.2 Rotate your x-axis text with the `xticks()` function\n",
    "\n",
    "Now that our plot is larger, we can still see the x-axis labels are still too large. We can, however, rotate the text and see if that helps. Let's rotate to a 90-degree angle. We'll alter these axis properties through the `xticks()` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the size of the plot\n",
    "plt.figure(figsize = (12, 5))\n",
    "\n",
    "# Build our barplot\n",
    "plt.bar(x=wormStrain_mean_embryos.index, \n",
    "        height=wormStrain_mean_embryos['numEmbryos'])\n",
    "\n",
    "## 2.1.2 Rotate the x-axis text\n",
    "plt.xticks(...)\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.1.3 Add labels to your plot\n",
    "\n",
    "Okay, the plot is larger and we've fixed our x-axis label issues. No more overcrowding! Let's add a main title and axis titles to our dataset. We can use the `title()`, `xlabel()`, and `ylabel()` functions in this case. While we set the labels, we can also set their properties such as `fontsize`, `fontstyle`, and `color`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the size of the plot\n",
    "plt.figure(figsize = (12, 5))\n",
    "\n",
    "# Build our barplot\n",
    "plt.bar(x=wormStrain_mean_embryos.index, \n",
    "        height=wormStrain_mean_embryos['numEmbryos'])\n",
    "\n",
    "# Rotate the x-axis text\n",
    "plt.xticks(rotation = 90)\n",
    "\n",
    "## 2.1.3 Add titles\n",
    "plt.title(\"Baseline mean embryos per worm strain\", fontsize = ...) \n",
    "# You can play with font style\n",
    "plt.xlabel(\"worm strain\", fontstyle = ...)\n",
    "# And even the font colour!\n",
    "plt.ylabel(\"mean embryo count\", color = ...)\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.1.4 Alter properties of your barplot\n",
    "\n",
    "Similar to altering text properties, most of the plots have the ability to alter various properties like fill and line \n",
    "color or other plot-specific attributes. Let's update the fill and line colours for our barplot using the `color` and `edgecolor` parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix the size of the plot\n",
    "plt.figure(figsize = (12, 5))\n",
    "\n",
    "# Build our barplot\n",
    "plt.bar(x=wormStrain_mean_embryos.index, \n",
    "        height=wormStrain_mean_embryos['numEmbryos'],\n",
    "        ## 2.1.4 you can also change some barplot aspects\n",
    "        color=...,\n",
    "        edgecolor=\"black\")\n",
    "\n",
    "# Rotate the x-axis text\n",
    "plt.xticks(rotation = 90)\n",
    "\n",
    "# Add titles\n",
    "plt.title(\"Baseline mean embryos per worm strain\", fontsize = \"x-large\") \n",
    "plt.xlabel(\"worm strain\", fontstyle = \"italic\")\n",
    "plt.ylabel(\"mean embryo count\", color = \"r\")\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>Section 2.0.0 Comprehension Question</b> What if we would like to change our x-axis to limit the number of subsites we are displaying? How would you plot just the first 10 values using the <b>plt.axis()</b> method? Modify the code below to try and alter the x-axis limits. For more information on how to use this method <a href=\"https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.axis.html?highlight=pyplot%20axis#matplotlib.pyplot.axis\">check out the documentation</a>.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "?plt.axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comprehension answer code 2.0.0\n",
    "# Fix the size of the plot\n",
    "plt.figure(figsize = (12, 5))\n",
    "\n",
    "# Build our barplot\n",
    "plt.bar(x=wormStrain_mean_embryos.index, \n",
    "        height=wormStrain_mean_embryos['numEmbryos'],\n",
    "        color=\"orchid\",\n",
    "        edgecolor=\"black\")\n",
    "\n",
    "# Rotate the x-axis text\n",
    "plt.xticks(rotation = 90)\n",
    "\n",
    "# Add titles\n",
    "plt.title(\"Baseline mean embryos per worm strain\", fontsize = \"x-large\") \n",
    "plt.xlabel(\"worm strain\", fontstyle = \"italic\")\n",
    "plt.ylabel(\"mean embryo count\", color = \"r\")\n",
    "\n",
    "# Alter the axis limits\n",
    "plt.axis(...)\n",
    "\n",
    "# Show our plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learn more about Pyplot:</b> We've barely scratched the surface of working with the Pyplot module and depending on your needs it may be enough to get along on. You can find more at the <a href=\"https://matplotlib.org/stable/tutorials/introductory/pyplot.html#sphx-glr-tutorials-introductory-pyplot-py\">matplotlib tutorial for pyplot</a>.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 3.0.0 Basic visualizations with the `seaborn` package\n",
    "\n",
    "Building upon our visualizations in the last section, there are some common themes you might recognize about them. We have a plot area, x- and y-axis data, axis limits, and plot colors. Using `matplotlib` to help generate your visualizations, you can control many small details but it can also be tedious at times to micromanage so many aspects of your plot.\n",
    "\n",
    "The `seaborn` package is actually built upon the `pyplot` module and tries to bring a high-level approach to statistical plots. As we'll see later on, this means updating certain details of our plots will require an understanding of the base `matplotlib` and `pyplot` functions.\n",
    "\n",
    "### 3.0.1 The `seaborn` package subdivides plot types into 3 categories\n",
    "\n",
    "The `seaborn` package takes a dual-pronged approach to generating plots. There are functions considered to work at the __Figure__ level and then there are functions that affect what is known as the __Axes__ level. To simplify the concept:\n",
    "\n",
    "- `Axes`: A single plot defined by an x and y axis grid. This includes all of the basic plots like scatter and box plots.\n",
    "- `Figure`: A plot space that can range from 1 to multiple `Axes`. The setup of `Axes` can be simple to complex.\n",
    "\n",
    "At the `Axes` levels, there are 3 categories of plot types based on their similarity: __relational, distribution, and categorical__. For each of these categories there is a figure-level function that can be used to create multi-panel (faceted) versions of these plots by splitting the data further based on categorical variables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/seaborn_function_overview_8_0.png?raw=true\"  width=700>|\n",
    "|:--:|\n",
    "|From the `seaborn` overview: for most simple plots, one of the above figure or axes-level plots can be utilized.|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above functions are used to initialize figure and axes objects by identifying a number of properties. Within these, some of the options can vary greatly based on plot type. Of the two levels of functions, their influence on figure attributes can vary:\n",
    "\n",
    "- __Figure-level__: Return a `FacetGrid` object that has some additional methods for altering attributes of the plot in a way that makes sense to the subplot organization.\n",
    "- __Axes-level__: Add axis labels and legends to the `Axes` they are drawn onto but do not alter the figure in any other way. You can choose to draw onto the current axes in memory OR specify the reference to an axes which may be within a larger figure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1.0 Introduction to the Grammar of Graphics\n",
    "\n",
    "One approach to effective data visualization relies on the Grammar of Graphics framework originally proposed by Leland Wilkinson (2005). The idea of grammar can be summarized as follows:\n",
    "\n",
    "- Grammar is the foundational set of rules that define the components of a language.\n",
    "- A language is built on a structure that consists of syntax and semantics."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/ggplot2.png?raw=true\" width=700></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The __grammar of graphics__ is a language to communicate about what we are plotting programmatically\n",
    "\n",
    "- It begins with a **tidy** data frame. It will have a series of __observations__ (rows) each of which will be described across multiple __variables__ (columns). Variables can actually represent qualitative or quantitative measurements or they could be descriptive data about the experiments or experimental groups.\n",
    "\n",
    "- The data units may undergo conversion through a process called **scaling** (transformation) before being used for plotting. \n",
    "\n",
    "- A subset of data columns are then passed on to be presented in various data plots (scatterplots, boxplots, kernel density estimates, etc.) by using the data to describe visual properties of the plot. We call these visual properties, the **aesthetics** of the plot. For example, the data being plotted or represented can be visually altered in shape or colour based on accompanying column data. \n",
    "\n",
    "- A plot can have multiple layers (for example, a scatter plot with a regression line) and each of these plot types is referred to as a **geom** (short for geometric object)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.1.1 The grammar of graphics with `seaborn`\n",
    "\n",
    "The grammar of graphics facilitates the concise description of any components of any graphics. Hadly Wickham of R `tidyverse` fame has proposed a variant on this concept - the layered grammar of graphics framework in the `ggplot2` package for R. By following a layered approach of defined components, it can be easy to build a visualization. \n",
    "\n",
    "In a similar manner, the `seaborn` package has some methods that facilitate a layering approach to building your visualizations. However, many of the details are built upon the foundation of layering `Axes` objects or alterations upon `Figure` objects.  \n",
    "\n",
    "Each Axes-level function usually takes in:\n",
    "\n",
    "1. `Data`: your visualization always starts here. What are the dimensions you want to visualize. What aspect of your data are you trying to convey?\n",
    "\n",
    "2. `Aesthetics`: assign your axes based on the data dimensions you have chosen. Where will the majority of the data fall on your plot? Are there other dimensions (such as categorically encoded groupings) that can be conveyed by aspects like size, shape, colour, fill, etc. \n",
    "\n",
    "3. `Geometric objects`: how will you display your data within your visualization. Which `*plot` will you use?\n",
    "\n",
    "The figure-level methods can be used to alter or update:\n",
    "\n",
    "1. `Scale`: do you need to alter your x or y-axis limits? What about scaling/transforming any values to fit your data within a range? Sometimes, depending on the Geometric object, you are better off transforming your data ahead of time.\n",
    "\n",
    "2. `Facets`: will generating subplots of the data add a dimension to your visualization that would otherwise by lost or hard to discern?\n",
    "\n",
    "3. `Coordinate system`: will your visualization follow a classic cartesian, semi-log, polar, etc. coordinate system?\n",
    "\n",
    "Let's jump into our first dataset and start building some plots with it shall we?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1.2 Import `infection_signal.tsv` \n",
    "\n",
    "Before we dig into the `seaborn` package, we will import a new dataset that has a slightly more diverse array of data that we can use for showcasing the plotting power of `seaborn`. To accomplish our task, we'll begin by importing an alternative dataset of measurements from `infection_signal.tsv` which consists of area measurements (total pixels) of animals from images after pathogen infection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll reset the code cells to only show FINAL code output.\n",
    "InteractiveShell.ast_node_interactivity = \"last\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the pitlatrine data in from file\n",
    "infectionSig_data = pd.read_csv(\"data/infection_signal.tsv\", sep = ...)\n",
    "\n",
    "# Look at the first 5 rows of data\n",
    "infectionSig_data.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How big is this dataframe?\n",
    "infectionSig_data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What are the unique values in each column?\n",
    "infectionSig_data.apply(...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Taking a quick look at our data we can summarize it briefly here and see that there are a few categories we can explore across variables like `strain` (4 types) and `fixing.date` (3 replicates) for measured variables like `area` (total area of a worm), `area.infected` (infection signal area) and `percent.infected` (infected area as a % of the total).\n",
    "\n",
    "## 3.2.0 Build a scatterplot using the `relplot()` function\n",
    "\n",
    "We'll start by building a basic scatterplot. We'll focus on comparing the total chemical oxygen demand versus the total solids count. Rather than working at the axes-level, we'll work with the encompassing `relplot()` method which will give us flexibility in our visual exploration down the road.\n",
    "\n",
    "For the basic plot `relplot()`, we'll start with the following parameters:\n",
    "- `data`: The tidy (long-form) data set we want to visualize.\n",
    "- `x`, `y`: The variable names for assigning x- and y-axis values.\n",
    "- `height`: The total height of our figure.\n",
    "- `aspect`: A scalar value used to determine the width of your figure (width = height * aspect).\n",
    "- `kind`: The type of plot we want to produce: scatter (default) vs line\n",
    "- `kwargs`: This is a catch-all for _any other_ keyword arguments that could be passed onto underlying functions like the `Axes`-level methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import your seaborn package\n",
    "import seaborn as sns\n",
    "\n",
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = ...,   # Set the data\n",
    "                      height = 6, aspect = 1,   # Set the size of the figure\n",
    "                      kind = ...          # Set the figure type\n",
    "                     )\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Looking at the output we can see that __without having specified the x and y axis values__, it simply plotted all of the variables along a default x-axis of index number (456 rows total). Many of these values have no real relationship to each other at all. Let's try again by setting our axis variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,  \n",
    "                      x = ..., y = ...,     # Set the axis variables\n",
    "                      height = 6, aspect = 1,   \n",
    "                      kind = \"scatter\"          \n",
    "                     )\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.1 Specify colouring by category with the `hue` parameter\n",
    "\n",
    "Now we begin to see the power of having a tidy DataFrame. Since each of our observations is in its own row, we can classify each observation by variables like `strain`! In other contexts, these may be considered __factors__ by which we categorize our data. Using the `seaborn` package, we can specify the colour of our points using the `hue` parameter. Since we have set the `data = infectionSig_data` parameter, we can tell `seaborn` to look at a specific variable/column when determining the `hue` parameter.\n",
    "\n",
    "At the same time, we'll play with two more parameters: \n",
    "\n",
    "1. The `alpha` parameter is essentially the opacity of each datapoint. Setting a lower value increases transparency which allows us to see overlapping datapoints better. This parameter becomes especially helpful when working with extremely dense datapoints.\n",
    "2. The `aspect` parameter will be increased so that our figure is a little wider instead of being square in shape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = ...,          # Set the point-colour by strain\n",
    "                      alpha = ...,              # Set the transparency of the points\n",
    "                      height = 6, aspect = 1.5,  \n",
    "                      kind = \"scatter\"          \n",
    "                     )\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.2 Alter your axis scale with the `.set()` method\n",
    "\n",
    "The `.set()` method is a gateway to altering a number of aspects of your plot. Once we have our plot saved as an object named `snsplot` we can alter or set some of it's properties this way. In particular we will be using the `yscale` parameter to adjust the y-axis log scale.\n",
    "\n",
    "Note that our plot object is actually kept as a type of `matplotlib.axes.Axes` and that's where we are calling the `.set()` method from. We can actually set quite a few figure attributes through this method.\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>More on figure attributes:</b> For a full list of plot attributes you can update through this method you can visit the <a href=\"https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set.html?highlight=set#matplotlib.axes.Axes.set\"> Matplotlib help page</a>.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = \"strain\",          # Set the point-colour by strain\n",
    "                      alpha = 0.6,              # Set the transparency of the points\n",
    "                      height = 6, aspect = 1.5,  \n",
    "                      kind = \"scatter\"          \n",
    "                     )\n",
    "\n",
    "## 3.2.2 Change your y-axis to a log scale\n",
    "snsPlot.set(...)\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.3 Facet your data into different plots with the `relplot()` function\n",
    "\n",
    "Depending on the complexity of your data, you may want to split or subgroup your data into multiple new plots based on certain variables. Splitting into multiple subplots is known as faceting your data. Usually this results in a grid-like pattern where data is grouped by categories of one variable as columns, and another variable by rows. The data could also simply be grouped by a single variable instead. Either way, this generates a figure-level object known as a `FacetGrid()`.\n",
    "\n",
    "The `relplot()` method already has the capability to handle this splitting of axes within the figure it generates. The `relplot()` method can facet data across two variables using the `row` and `col` parameters. We simply need to name the variable(s) that will be used to categorize the data.\n",
    "\n",
    "To summarize, the parameters to use for this operation are:\n",
    "\n",
    "- `col`: The variable name that will be used to group the columns of your grid.\n",
    "- `row`: The variable name that will be used to group the rows of your grid.\n",
    "\n",
    "Below, we'll remove the colouring of points based on `strain` and instead, split the data into four `Axes` based on this information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      alpha = 0.6,               # Set the transparency of the points\n",
    "                      height = 6, aspect = 0.6,  # Set the size of the figure\n",
    "                      kind = \"scatter\",\n",
    "                      col = ...             ## 3.2.3 Split the columns of the grid by country\n",
    "                     )\n",
    "\n",
    "# Change your y-axis to a log scale\n",
    "snsPlot.set(yscale=\"log\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.4 Use a continuous variable to colour your data\n",
    "\n",
    "Note that we typically need to just apply one attribute to each dimension of data we are investigating. By splitting the data by `strain` we no longer need to colour it based on this category. We can, however, add additional information to our visualization by using another dimension in our data. Instead of colouring the points based on a categorical variable (strain), we can use a continuous variable like `percent.infected` from our dataset to see if there could be a trend in relation to our data.\n",
    "\n",
    "It is easy enough to set this dimension using the `hue` parameter in our initial `relplot()` call. We'll also set the `palette` parameter to a different colour and set the `edgecolor` parameter so that our lower-value/white points can still be seen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = ..., palette = ...,   ## 3.2.4 Set the point-colour by percent infected\n",
    "                      edgecolor = \"black\",       ## 3.2.4 Set the point border colour so we can see them all\n",
    "                      alpha = 0.6,               \n",
    "                      height = 6, aspect = 0.6,  \n",
    "                      kind = \"scatter\",\n",
    "                      col = \"strain\"             # Split the columns of the grid by country\n",
    "                     )\n",
    "\n",
    "# Change your y-axis to a log scale\n",
    "snsPlot.set(yscale=\"log\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By colouring our datapoints using `percent.infected`, we can now see more clearly on the same plot that samples with higher area.infected values but lower overall area, tend to trend towards having a higher overall infected area percentage. Not a surprising result BUT very cool to see when colouring our datapoints. Rather than generating additional plots comparing different pairs of variables, we've simply added an additional dimension of information to our visualization.\n",
    "\n",
    "***\n",
    "### 3.2.5 Update your axis using the `.set_axis_labels()` method\n",
    "\n",
    "The names of our axis titles are drawn from the variable names we used for the original `DataFrame` but we may be limited in how those variables are originally named. In other cases you may wish to add units, or simply make your axis titles more descriptive. To accomplish this we can alter our labels directly using the `set_axis_labels()` method. The parameters to set are `x_var` and `y_var` in that order. Set them directly if you only want to change a single axis title."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = \"percent.infected\", palette = \"Reds\",   # Set the point-colour by percent infected\n",
    "                      edgecolor = \"black\",       # Set the point border colour so we can see them all\n",
    "                      alpha = 0.6,               \n",
    "                      height = 6, aspect = 0.6,  \n",
    "                      kind = \"scatter\",\n",
    "                      col = \"strain\"             # Split the columns of the grid by country\n",
    "                     )\n",
    "\n",
    "## 3.2.5 Set the axis titles (Aesthetics)\n",
    "snsPlot.set_axis_labels(x_var = ..., \n",
    "                        y_var = ...)\n",
    "\n",
    "# Change your y-axis to a log scale\n",
    "snsPlot.set(yscale=\"log\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.6 Set the shape of your points with the `style` parameter\n",
    "\n",
    "We'll switch gears a little at this point and ask what our data looks like when we compare across our `fixing.date` replicates instead. For each strain facet, we'll separate the data by replicate by changing their shape using the `style` parameter instead. Thus the visualization of our data will be grouped further into replicate values while still faceting on actual worm strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = \"percent.infected\", palette = \"Reds\",   # Set the point-colour by percent infected\n",
    "                      edgecolor = \"black\",       # Set the point border colour so we can see them all\n",
    "                      style = ...,     # 3.2.6 Change the style of our points based on variable values\n",
    "                      alpha = 0.6,               \n",
    "                      height = 6, aspect = 0.5,  \n",
    "                      kind = \"scatter\",\n",
    "                      col = \"strain\"             # Split the columns of the grid by country\n",
    "                     )\n",
    "\n",
    "# Set the axis titles (Aesthetics)\n",
    "snsPlot.set_axis_labels(x_var = \"Total area\", \n",
    "                        y_var = \"Area infected\")\n",
    "\n",
    "# Change your y-axis to a log scale\n",
    "snsPlot.set(yscale=\"log\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** \n",
    "### 3.2.7 Using the right colours for your visualization with the `color_palette()` function\n",
    "\n",
    "From the output of our last visualization you may wonder if we can do better with colouring based on the `percent.infected` values. As our values approach 0, things get very hard to see and this prompted us to use a dark outline for our points. Another helpful considering is the use of an appropriate colour palette. These can be accessed via the the `seaborn.color_palette()` function which includes the parameters:\n",
    "\n",
    "- `palette`: the name of the palette you wish to access (add \"_r\" to the name to reverse it!)\n",
    "- `n_colours`: the number of colours from the palette you'd like to access (most named palettes have 6 different colours)\n",
    "- `as_cmap`: a boolean to determine if you'd like to create a continuous mapping of values across a gradient of palette colours (think heatmaps!)\n",
    "\n",
    "When calling on this function, it will return to you a series of colour values either in a discrete or continuous gradient.\n",
    "\n",
    "We'll update our figure with a colour-blind friendly \"viridis\" palette in a reverse so our higher values are darker in colour."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a scatterplot\n",
    "snsPlot = sns.relplot(data = infectionSig_data,   \n",
    "                      x = \"area\", y = \"area.infected\",     \n",
    "                      hue = \"percent.infected\",  # Set the point-colour by percent infected\n",
    "                      palette = sns.color_palette(palette = ..., \n",
    "                                                  as_cmap = ...),  ## 3.2.7 Use a better colour palette like viridis\n",
    "                      edgecolor = \"black\",       # Set the point border colour so we can see them all\n",
    "                      style = \"fixing.date\",     # 3.2.6 Change the style of our points based on variable values\n",
    "                      alpha = 0.7,               \n",
    "                      height = 6, aspect = 0.5,  \n",
    "                      kind = \"scatter\",\n",
    "                      col = \"strain\"             # Split the columns of the grid by country\n",
    "                     )\n",
    "\n",
    "# Set the axis titles (Aesthetics)\n",
    "snsPlot.set_axis_labels(x_var = \"Total area\", \n",
    "                        y_var = \"Area infected\")\n",
    "\n",
    "# Change your y-axis to a log scale\n",
    "snsPlot.set(yscale=\"log\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>How do I know which colours to use?</b> Unfortunately we don't have time to explore the types of colour palettes that you'd like to use but the <a href=\"https://seaborn.pydata.org/tutorial/color_palettes.html\">searborn help page</a> has a brief tutorial that explains the importance of palettes and how you can use qualitative, sequential, and diverging colour palettes to help convey your message!\n",
    "</div>\n",
    "\n",
    "## 3.3.0 Plotting theoretical distributions\n",
    "\n",
    "Now that we have some of the basics, it's time to take a closer look at using other types of plots. Let's return to our embryo data in `embryo_merged_subset`. It has a lot of nice population-based data that we can dissect to look at theoretical distributions.\n",
    "\n",
    "### 3.3.1 More on filtering your data with the boolean AND operator: `&`\n",
    "\n",
    "We'll focus our dataset first by filtering for just the `N2` strain in the mock infection condition. To accomplish this we'll use the conditional AND `&` operator which can combine our boolean expressions. What we skipped last time around was that to accompany this, you can also use the conditional OR `|` and we've already seen the logical NOT `~` which can convert our boolean to it's opposites. \n",
    "\n",
    "If we are generating this code using `.loc` conditional selection, each boolean expression _must be_ separately enclosed by the parentheses syntax `( )`. We'll talk more about this next time in __Lecture 05__. For now, we'll continue to use the `.query()` method as before and save the filtered data into the object `N2_mock_data`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data by N2 animals with a pathogenDose of 0\n",
    "N2_mock_data = embryo_merged_subset.query('wormStrain == \"N2\" ...')\n",
    "\n",
    "# Filter code for using loc instead:\n",
    "# N2_mock_data = embryo_merged_subset.loc[(embryo_merged_subset['wormStrain'] == \"N2\") &\n",
    "#                                         (embryo_merged_subset['pathogenDose'] == 0), \n",
    "#                                        :]\n",
    "\n",
    "# Check on the data created\n",
    "N2_mock_data.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.3.2 Look at your distribution as a kernel density estimate with `kdeplot()` or `displot()`\n",
    "\n",
    "There are a lot of datapoints in our new dataset. This time our measurements are based on the `numEmbryos` for specific uninfected worms within different replicate experiments. A quick question we already investigated was to identify the overall distributions of our embryo numbers across different replicates for all of our strains. We can, however, also visualize this data as a distribution. \n",
    "\n",
    "A quick way to answer this question is by generating a kernel density estimate (KDE) using the `displot()` method from `seaborn`. We'll need to provide an `x` value (`numEmbryos` in this case) and we'll colour our plots based on the `date` variable which is coded as a series of integers which should represent our separate experimental replicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a KDE plot\n",
    "snsPlot = sns.displot(data = N2_mock_data,          \n",
    "                      x = ...,\n",
    "                      hue = ..., \n",
    "                      height = 6, aspect = 2,        \n",
    "                      kind = ...,\n",
    "                      fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "                     )\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Why do we only have 2 colours in our distribution plot? We can see all of the various dates, BUT only the first is really coloured and the others all have a darker hue. It is very reminiscent of when we coloured our scatterplot by the `hue` parameter as well. However, should the values in `date` be considered as numbers __or__ as separate individual categories?\n",
    "\n",
    "### 3.3.3 The `category` data type reassigns order or meaning to your values\n",
    "\n",
    "Sometimes when we work with our data, we may produce what _look_ like numerical values for a variable like a replicate number or serial number. However, these values aren't just numbers but really grouping values just like we have in our `wormStrain` variable. Python/pandas doesn't differentiate on this idea inherently when passing data around because **it has no insight into our intentions**. Moreover when we work with the `seaborn` package - it will also treat integers and floats just like integers and floats. \n",
    "\n",
    "Recall in __section 3.3.1__ we looked at the dtype values in our DataFrame and the `date` variable was an `int64`. What we want is some other way to represent this data. We could convert it to a string `str` BUT we'll introduce a better dtype called the `category`. \n",
    "\n",
    "[Categorical variables](\"https://pandas.pydata.org/docs/user_guide/categorical.html?highlight=category\") are very handy when working with statistical analysis and help to define groups but can also give them an order of importance. This means when analysing or plotting the data, **this specific order can be used to determine how that data is used or displayed**. \n",
    "\n",
    "To start off simple, let's convert our `date` variable to a `category` data type and see how that affects our KDE plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert our date variable with the astype() method\n",
    "N2_mock_data = N2_mock_data.astype({'date':...})\n",
    "\n",
    "# rebuild a KDE plot\n",
    "snsPlot = sns.displot(data = N2_mock_data,          \n",
    "                      x = \"numEmbryos\",\n",
    "                      hue = \"date\", \n",
    "                      height = 6, aspect = 2,        \n",
    "                      kind = \"kde\",\n",
    "                      fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "                     )\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** \n",
    "It worked! Now we can see that each date has been given a distinct colour!\n",
    "\n",
    "### 3.3.3.1 Reverse the order of a list using slicing notation\n",
    "\n",
    "So right now we can see the order of our data has been set by the date values with our earliest date (190426) on top and our latest date (200918) on the bottom. When defining categorical data, you can also _define the order_ of your data. To help us out, we'll grab an array of the unique elements in our `date` variable and then we can use the slicing notation `[::-1]` to reverse our array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take the unique values and flip the order\n",
    "date_list = N2_mock_data.date.unique()[...].tolist()\n",
    "\n",
    "# View the reversed date list\n",
    "date_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.3.3.2 Reorder your categories with `reorder_categories()`\n",
    "\n",
    "Now that we have a reverse list of our categories (or you could make a custom list of course), you can replace the `date` column in our dataset. To alter the categorical information, you must access the `.cat` property and use the `.reorder_categories()` method on it. This will return a ___new___ categorical object which you must use to ___replace___ the original `date` data.\n",
    "\n",
    "We'll take a look at our updated KDE plot afterwards and see if it worked!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll need to pull out and replace the date column\n",
    "# In this case we will not \"pop\" the column out but work on it as part of the DataFrame\n",
    "N2_mock_data['date'] = (N2_mock_data['date']\n",
    "                        # access the categorical property\n",
    "                        .cat\n",
    "                        # Reorder the categorical data\n",
    "                        .reorder_categories(new_categories = ...)\n",
    "                       )\n",
    "\n",
    "# Double-check the date column results\n",
    "N2_mock_data['date']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rebuild a KDE plot with the new category order\n",
    "snsPlot = sns.displot(data = N2_mock_data,          \n",
    "                      x = \"numEmbryos\",\n",
    "                      hue = \"date\",            # Now the date variable is a category!\n",
    "                      height = 6, aspect = 2,        \n",
    "                      kind = \"kde\",\n",
    "                      fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "                     )\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Learn more about categories!</b> There are many additional methods you can use for altering, renaming, and combining categories! Check out more information on the <a href=\"https://pandas.pydata.org/docs/user_guide/categorical.html\">pandas CategoricalDtype help page</a>\n",
    "</div>\n",
    "\n",
    "\n",
    "### 3.3.4 Add a `rugplot()` to the margin\n",
    "\n",
    "Within `seaborn` there are a few ways marginal plots that can be added to your visualizations. Marginal plots usually add distribution summaries like a histogram, kde or in our case, a rugplot. More specifically, we'll be using the `.rugplot()` method but there is also the ability to create certain plot combinations using the `.jointplot()` method.\n",
    "\n",
    "A rugplot is simply a series of vertical or horizontal tick-marks representing our actual data points along the x- and/or y-axis. For our rugplot, we'll add it outside of our plot area by manipulating the parameters:\n",
    "\n",
    "- `height`: determine how \"tall\" our rug plot is as a _proportion_ of the plot. Positive values fall within the axis vs negative values.\n",
    "- `clip_on`: boolean parameter to denote if the plot should cut off objects falling outside the axes limits (`True`) or let them remain rendered (`False`) outside the plot axes.\n",
    "- `alpha`: a common parameter to set which determines the opacity of plot points. In this case, it will make viewing the density of our tick marks a little better.\n",
    "\n",
    "Despite where it is plotted, we are adding this plot on the underlying `Axes` object of the current `snsPlot` figure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll build a KDE plot\n",
    "snsPlot = sns.displot(data = N2_mock_data,          \n",
    "                      x = \"numEmbryos\",\n",
    "                      hue = \"date\",            # Now the date variable is a category!\n",
    "                      height = 6, aspect = 2,        \n",
    "                      kind = \"kde\",\n",
    "                      fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "                     )\n",
    "\n",
    "## 3.3.4 Add a rugplot - this is plotted on top of the current axes object\n",
    "sns.rugplot(data = N2_mock_data,\n",
    "            x = ..., \n",
    "            height = ...,         # Draw the rug plot BELOW the x-axis\n",
    "            clip_on = False,        # Ensure that it is rendered\n",
    "            alpha = 0.5\n",
    "           )\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.4.0 Boxplots provide visual summary statistics of your data\n",
    "\n",
    "Boxplots are categorical type plots and a great way to visualize summary statistics for your data. As a reminder, the thick line in the center of the box is the __median__. The upper and lower ends of the box are the first and third __quartiles__ (or 25th and 75th percentiles) of your data. The whiskers extend to the largest value no further than 1.5\\*IQR (inter-quartile range - the distance between the first and third quartiles). \n",
    "\n",
    "Data beyond these whiskers are considered outliers and plotted as individual points. This is a quick way to see how comparable your samples or variables are.<br><br>\n",
    "\n",
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/boxplot_dissection.png?raw=true\"  width=700></center>\n",
    "\n",
    "<br>\n",
    "We are going to use boxplots to see the distribution of embryos per worm strain across all uninfected worm samples.\n",
    "\n",
    "### 3.4.1 Generate a basic boxplot with the `catplot()` function\n",
    "\n",
    "To build the basic boxplot we begin with the main variables. We want to summarize the mean embryo values distribution of each worm strain across all replicate experiments. We can use the `catplot()` method to build our visualization. You'll note that the plot is not automatically coloured to differentiate between the x-axis groups.\n",
    "\n",
    "The `catplot()` method is the gateway to more categorical plots and behaves similarly to the other two figure-level plots we've encountered. \n",
    "\n",
    "Before we begin, we'll summarize the data again as we did earlier in __section 1.3.0__ but this time we'll also include an extra level to `.groupby()` with `doseLevel`. This data was not included in our work from last week but represents a categorical series of values that interpret specific pathogenStrain-pathogenDose combinations as qualitative levels of \"Mock\", \"Medium\", \"High\", etc.\n",
    "\n",
    "We'll again use the `in` keyword in our `query()` to help us filter and use the `.agg()` method to calculate the mean for each of our subgroups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter by doseLevel as either Mock or Medium infections\n",
    "mean_embryo_data = (embryo_merged_subset.query('doseLevel in [\"Mock\", \"Medium\"]')\n",
    "\n",
    "                    # Group by infection experiment \n",
    "                    .groupby(by = ['date', 'wormStrain', 'pathogenStrain', 'doseLevel'])\n",
    "\n",
    "                    # Create the frequency table on numEmbryos\n",
    "                    # ['numEmbryos']\n",
    "                    .agg({'numEmbryos':'mean'})\n",
    "\n",
    "                    # Recall, when we reset the index, it converts the indices back into columns\n",
    "                    .reset_index()\n",
    "\n",
    "                    # Set our wormStrain and doseLevel variables as categories!\n",
    "                    .astype({'wormStrain':'category', ...})\n",
    ")\n",
    "\n",
    "# Check out our summarized data\n",
    "mean_embryo_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Now we can build our boxplot using the summarized data!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = ..., y = ...,                       \n",
    "                      kind = ..., # Make a boxplot\n",
    "                      height = 6, aspect = 2 # Set the width of our plot to 4 and width at 12 (6x2)\n",
    "                     )\n",
    "                    \n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.2 Rotate your x-axis through the `set_xticklabels()` method\n",
    "\n",
    "We've encountered this problem before when working with `matplotlib.pyplot` and the solution is actually the same. Recall that `seaborn` is built upon the back of the `pyplot` module so we can actually modify the plot directly through pyplot!\n",
    "\n",
    "We'll rotate the x-axis text with the `set_xticklabels()` method on our plot object to 90 degrees. At the same time, let's update the `hue` parameter to colour our boxes by wormStrain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", y = \"numEmbryos\", \n",
    "                      kind = \"box\",                        # Make a boxplot\n",
    "                      height = 6, aspect = 2,\n",
    "                      hue = \"wormStrain\",                  # Colour our wormStrains\n",
    "                     )\n",
    "\n",
    "## 3.4.2 Alter the xtick attributes\n",
    "snsPlot...                   \n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.4 Generate nested boxplots using the `hue` parameter\n",
    "\n",
    "We already know that our data is measured across two dose levels (Mock and Medium) so we can take advantage of this information to create a nested (paired/grouped) set of boxplots. When you have a smaller number of categories (ie treatments), this allows you to more directly compare the characteristics of your two subpopulations. Let's see what happens when we use the `hue` parameter to distinguish between our doseLevel subsets.\n",
    "\n",
    "We'll also amke the following additional changes:\n",
    "\n",
    "1. set the `boxplot()` parameter `width` to put a little more distance between each category along the x-axis. This value usually ranges between 0 and 1.\n",
    "2. To save on some space we'll additionally move the legend ___into___ the plot space using the `legend_out` boolean parameter.\n",
    "3. We'll update the y-axis title to correctly reflect that these are the mean number of embryos from each grouped experiment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", y = \"numEmbryos\", \n",
    "                      kind = \"box\",                        # Make a boxplot\n",
    "                      height = 6, aspect = 2,\n",
    "                      hue = ...,         ## 3.4.4 Set the hue by doseLevel\n",
    "                      legend_out = False,        ## 3.4.4 Move the legend inside the plot\n",
    "                      width = 0.6                ## 3.4.4 Put some more distance between categories by decreasing width\n",
    "                     )\n",
    "\n",
    "# Set the axis titles (Aesthetics)\n",
    "snsPlot.set_axis_labels(y_var = \"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(rotation = 90)                    \n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Let's take a moment to fix our `doseLevel` category so that we plot the \"Mock\" data __before__ the infected data for our boxplot. Remember, we can use the `.reorder_categories()` method to accomplish this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull out our doseLevel data and replace it with a newly categorized version\n",
    "mean_embryo_data['doseLevel'] = (mean_embryo_data['doseLevel']\n",
    "                                 .cat\n",
    "                                 .reorder_categories(new_categories = [...])\n",
    "                                )\n",
    "\n",
    "# Double check it worked\n",
    "mean_embryo_data['doseLevel'].cat.categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", y = \"numEmbryos\", \n",
    "                      kind = \"box\",                        # Make a boxplot\n",
    "                      height = 6, aspect = 2,\n",
    "                      hue = \"doseLevel\",         # Set the hue by doseLevel\n",
    "                      legend_out = False,        # Move the legend inside the plot\n",
    "                      width = 0.6                # Put some more distance between categories by decreasing width\n",
    "                     )\n",
    "\n",
    "# Set the axis titles (Aesthetics)\n",
    "snsPlot.set_axis_labels(y_var = \"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(rotation = 90)                    \n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.5 Facet your crowded data with the `row` and `col` parameters\n",
    "\n",
    "So our plot above is quite busy. Whereas previously, we used the `relplot()` method to generate a faceted scatterplot (or relational plot), here we'll use the `catplot()` method to accomplish something similar. The `catplot()` method handles the distribution or faceting of _categorical_ plots using similar parameters:\n",
    "\n",
    "- `col`: The variable name that will be used to group the columns of your grid.\n",
    "- `row`: The variable name that will be used to group the rows of your grid.\n",
    "\n",
    "This time around we'll facet our data into a stacked set of plots using the `row` parameter. You'll notice that since we are no longer grouping by `hue`, the legend will also disappear.\n",
    "\n",
    "At the same time, we'll play with a few additional methods:\n",
    "- `set_axis_labels()`: can accept an x and y-axis label in the form of a string. You can also set separate labels with the `set_xlabels` and `set_ylabels` methods.\n",
    "- `set_xticklabels()`: we already used this to rotate our x-axis text but we also can set the `labels` parameter.\n",
    "- `set_titles()`: we'll use this to simplify the name of each panel to use just the value from our variable. This is found in string `\"{col_name}\"`. `\"{col_var}\"` would use the variable name we're setting our facet by, ie. \"wormStrain\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll need to increase the font scale for this very large plot\n",
    "sns.set(font_scale = 4)\n",
    "\n",
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = ..., y = \"numEmbryos\", \n",
    "                      kind = \"box\",              # Make a boxplot\n",
    "                      height = 10, aspect = 1,\n",
    "                      width = 0.6,               # Put some more distance between categories by decreasing width\n",
    "                      col = \"wormStrain\",        ## 3.4.5 Set our facets based on wormStrain \n",
    "                      ...               ## 3.4.5 Only use 5 facets per row\n",
    "                     )\n",
    "\n",
    "## 3.4.5 Alter the xtick attributes\n",
    "snsPlot...(labels = [\"Mock\", \"Medium\"])                    \n",
    "               \n",
    "## 3.4.5 Change our y-axis label\n",
    "snsPlot.set_axis_labels(\"dose level\", \"mean embryos per replicate\")\n",
    "\n",
    "## 3.4.5 set the title\n",
    "snsPlot...(\"{col_name}\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.6 Overlay on a faceted plot using `map_dataframe()`\n",
    "\n",
    "Even though boxplots give us summary statistics on our data, it is useful to readers (and reviewers) to be able to see where our individual data points are. We've already used `rugplot()` to help visualize our data distribution in density plots. In that case, we simply plotted on top of the already present _simple_ plot.\n",
    "\n",
    "Similarly, for a boxplot we can add the data as another layer using using an `sns.swarmplot()` to place dots on top of our boxplot. A swarmplot places overlapping data points in a such a way that they are next to each other instead. This gives us a better picture of the distribution of our data. \n",
    "\n",
    "In the case of our ___facted_boxplot___ however, we cannot simply overlay with `sns.swarmplot()`. Instead, we need to map our data using the `.map_dataframe()` method. It will preserve the underlying panel/graph and use it's characteristics to overlay a new plot, __potentially with a different DataFrame__. It uses the following parameters:\n",
    "\n",
    "- `func`: the function we want to overlay. This would be the `sns.swarmplot` in our case. Note the lack of parentheses!\n",
    "- `args` and `kwargs`: we'll talk more about this in __Lecture 5__, but essentially any additional arguments you would normally use for your plotting function, are simply supplied here as named parameters.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"doseLevel\", y = \"numEmbryos\", \n",
    "                      kind = \"box\",                        # Make a boxplot\n",
    "                      height = 10, aspect = 1,\n",
    "                      width = 0.6,                # Put some more distance between categories by decreasing width\n",
    "                      hue = \"doseLevel\",          ## 3.4.6 Re-colour our boxes by doseLevel again!\n",
    "                      col = \"wormStrain\",\n",
    "                      col_wrap = 5,\n",
    "                      fliersize = 0               ## 3.4.6 Hide our outliers by making them size 0\n",
    "                     )\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(labels = [\"Mock\", \"Medium\"])                    \n",
    "               \n",
    "# set the title\n",
    "snsPlot.set_titles(\"{col_name}\")\n",
    "\n",
    "## 3.4.6 Overlay a swarmplot on our catplot\n",
    "snsPlot...(..., data = mean_embryo_data,\n",
    "                      x = \"doseLevel\", \n",
    "                      y = \"numEmbryos\", \n",
    "                      hue = ...,        # split the points by dose level just like nested boxplots\n",
    "                      palette=\"dark:black\",     # Recolour all of the points to black\n",
    "                      size=15                   # Set the size of our points so they can be seen\n",
    ")\n",
    "\n",
    "## 3.4.6 Change our y-axis label. It appears to work best AFTER the map_dataframe() call\n",
    "snsPlot.set_axis_labels(\"dose level\", \"mean embryos per replicate\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.5.0 Use the violin plot to visualize distributions\n",
    "\n",
    "If you could combine aspects of the boxplot and the KDE into a single visualization, you would get something like the violin plot. Another way to think of the violin plot is to imagine a KDE plot that's been shrunk down and placed categorically.\n",
    "\n",
    "It's actually quite easy to switch over since many of the aspects are similar to the boxplot. We need only change the `kind` parameter in our `catplot()` code. The `fliersize` parameter will also be removed since this no longer applies to our plot and will throw a warning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"doseLevel\", y = \"numEmbryos\", \n",
    "                      kind = ...,            ## 3.5.0 Make a violin plot\n",
    "                      height = 10, aspect = 1,\n",
    "                      width = 0.6,                # Put some more distance between categories by decreasing width\n",
    "                      hue = \"doseLevel\",          # Re-colour our violins by doseLevel again!\n",
    "                      col = \"wormStrain\",\n",
    "                      col_wrap = 5\n",
    "                     )\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(labels = [\"Mock\", \"Medium\"])                    \n",
    "               \n",
    "# set the title\n",
    "snsPlot.set_titles(\"{col_name}\")\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "snsPlot.map_dataframe(sns.swarmplot, data = mean_embryo_data,\n",
    "                      x = \"doseLevel\", \n",
    "                      y = \"numEmbryos\", \n",
    "                      hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "                      palette=\"dark:black\",     # Recolour all of the points to black\n",
    "                      size=15                   # Set the size of our points so they can be seen\n",
    ")\n",
    "\n",
    "# Change our y-axis label. It appears to work best AFTER the map_dataframe() call\n",
    "snsPlot.set_axis_labels(\"dose level\", \"mean embryos per replicate\")\n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.5.1 Combine a binary grouping with the `split` parameter\n",
    "\n",
    "Sometimes a more direct comparison of your data can be applied through the violin plot by generating a split version of it. This is especially helpful when you are working with nested data that is binary and you would like to compare it visually. \n",
    "\n",
    "We'll initialize this visualization with the `split` boolean parameter. To help with this visualization we'll also:\n",
    "\n",
    "- change the overlay from a `swarmplot` to a `stripplot` to accommodate a more narrow width of the half-violins\n",
    "- include some `inner` markers of the quartile information in each violin\n",
    "- Use the `density_norm` parameter to ensure our violins are the same `width` (for easier viewing). Other scaling options for this are `area` and `count`.\n",
    "- switch up the colours here with the `palette` parameter which accepts a dictionary-like object too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(font_scale = 1)\n",
    "\n",
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", y = \"numEmbryos\", \n",
    "                      kind = \"violin\",                        # Make a boxplot\n",
    "                      height = 6, aspect = 2, \n",
    "                      hue = \"doseLevel\",                      \n",
    "                      split = True,            ## 3.5.1 This will create hybrid violin plots\n",
    "                      inner = \"quart\",         ## 3.5.1 Add quartile markers to each half of the violin\n",
    "                      density_norm=\"width\",    ## 3.5.1 Make all the widths the same\n",
    "                      palette = {\"Mock\":\"crimson\", ...}  ## 3.5.1 Update the palette of our violins\n",
    "                     )\n",
    "\n",
    "## 3.5.1 Overlay a stripplot on our catplot\n",
    "snsPlot.map_dataframe(sns.stripplot, data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", \n",
    "                      y = \"numEmbryos\", \n",
    "                      hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "                      palette=\"dark:black\",     # Recolour all of the points to black\n",
    "                      size=5,                   # Set the size of our points so they can be seen\n",
    "                      dodge = True              # Reduce the overlap on any points\n",
    ")\n",
    "\n",
    "# Change our y-axis label. It appears to work best AFTER the map_dataframe() call\n",
    "snsPlot.set_axis_labels(\"dose level\", \"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(rotation = 90)  \n",
    "\n",
    "# Show the plot (avoid extra object information)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 4.0.0 Saving your plots to file\n",
    "\n",
    "Up until now, we have taken for granted that our plots have been displayed using a *Graphic Device*. For our Jupyter Notebooks we can see the graphs right away and update our code. You can even save them manually from the output display but sometimes you may be producing multiple visualizations based on large data sets. In this case it is preferable to save them directly to file.\n",
    "\n",
    "## 4.1.0 Save your figures with `plt.savefig()` method\n",
    "\n",
    "Once you have a figure the way you want it, you can save in any number of graphical and non-graphical formats. The `savefig()` method from the `pyplot` package is here to save the day. Saving the current figure, you can use some of the following parameters:\n",
    "\n",
    "- `fname`: The path to your file you want to save, including the extension. If `format` is not set, then the file extension will be used to _infer_ the format instead.\n",
    "- `dpi`: The resolution in dots per inch for your figure.\n",
    "- `format`: The file format you'd like to use. Supported filetypes include svg, jpg, eps, and pdf.\n",
    "\n",
    "Let's save our faceted scatterplot from way back in section 3.2.6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use catplot to make our boxplot\n",
    "snsPlot = sns.catplot(data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", y = \"numEmbryos\", \n",
    "                      kind = \"violin\",                        # Make a boxplot\n",
    "                      height = 6, aspect = 2,\n",
    "                      hue = \"doseLevel\",\n",
    "                      split = True,            # This will create hybrid violin plots\n",
    "                      inner = \"quart\",         # Add quartile markers to each half of the violin\n",
    "                      density_norm=\"width\",    # Make all the widths the same\n",
    "                      palette = {\"Mock\":\"crimson\", \"Medium\": \"deepskyblue\"}  # Update the palette of our violins\n",
    "                     )\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "snsPlot.map_dataframe(sns.stripplot, data = mean_embryo_data,\n",
    "                      x = \"wormStrain\", \n",
    "                      y = \"numEmbryos\", \n",
    "                      hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "                      palette=\"dark:black\",     # Recolour all of the points to black\n",
    "                      size=5,                   # Set the size of our points so they can be seen\n",
    "                      dodge = True\n",
    ")\n",
    "\n",
    "# Change our y-axis label. It appears to work best AFTER the map_dataframe() call\n",
    "snsPlot.set_axis_labels(\"dose level\", \"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "snsPlot.set_xticklabels(rotation = 90)  \n",
    "\n",
    "# Save the plot\n",
    "plt.savefig(..., \n",
    "            format = ..., \n",
    "            dpi = ...\n",
    "           )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 5.0.0 Plotting multi-panel figures\n",
    "\n",
    "Up until now we've been generating a combination of either faceted plots or simply layering elements upon single plots. Throughout all of these we have not really been mixing the _types_ of plots we've generated. Luckily for us, the `matplotlib.pyplot` module provides a means for us to put together multiple plot axes in a single figure.\n",
    "\n",
    "We have already seen some of these __figure-level__ functions in action with `relplot()` and `catplot()` which provided an interface to __axes-level__ methods like `scatterplot()` or `boxplot` when creating faceted plots."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/seaborn_function_overview_8_0.png?raw=true\"  width=700>|\n",
    "|:-:|\n",
    "|A handy figure from the seaborn overview at: https://seaborn.pydata.org/tutorial/function_overview.html|\n",
    "\n",
    "What if, however, we would like to create a figure with multiple axes of different types?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.1.0 Use `subplot2grid()` to generate a multi-grid plot\n",
    "\n",
    "When generally considering the layout of our data, we want to think of breaking up the figure into a grid. This can start off simply as a 1x1 panel and expand outwards with nested panels within a 2x2 or 3x3 or larger figure. The dimensions also are not limited to square shapes but can be rectangular as well.\n",
    "\n",
    "The `pyplot.subplot2grid()` function takes the following parameters:\n",
    "- `shape`: The dimensions of the __figure__ given in a (numRow, numCol) tuple. This is essentially the backdrop of all the panels.\n",
    "- `loc`: The location of the subplot (Axes object) you are creating in relation to the base __figure__. Position (0,0) is the top left corner.\n",
    "- `rowspan`: The dimensions of the subplot in number of rows.\n",
    "- `colspan`: The dimensions of the subplot in number of columns.\n",
    "- `fig`: A figure object to place the Axes object in. Otherwise the _current_ figure is used."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/ggarrange_examples.png?raw=true\"  width=700>|\n",
    "|:-:|\n",
    "|Some simple layouts demonstrating how a figure can be subdivided|\n",
    "\n",
    "Let's make a figure with 3 plots as seen in the 4th example above. We'll begin just by generating the specific layout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure()\n",
    "\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid(shape = (2, 2), loc = (...), colspan = 1, rowspan = 2)\n",
    "ax2 = plt.subplot2grid(shape = (2, 2), loc = (...), colspan=1)\n",
    "ax3 = plt.subplot2grid(shape = (2, 2), loc = (...), colspan=1)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5.1.1 Use the `Axes` objects as a canvas to plot onto `fig`\n",
    "\n",
    "Now you can see that our figure encompasses the 3 panels that we envisioned. You'll notice that we named each panel with a reference/variable. We could have also added them to a single list object to save on variable names but to simplify our understanding they've been named separately. \n",
    "\n",
    "Why do we need these objects? Each `Axes`-level function we use to plot with, can take a parameter `ax` where we pass along an `Axes` object. This identifies _which_ panel we want to plot onto in our overall figure. Otherwise it will use the last `Axes` object generated (ie `ax3` was the last one we generated in our above code). We'll fill our `Axes` as follows:\n",
    "\n",
    "- `ax1`: A scatterplot object from our infection signal data\n",
    "- `ax2`: Our split-violin plot of the mean embryo values across strains\n",
    "- `ax3`: A kde plot of the uninfected embryo counts per strain\n",
    "\n",
    "Each of these is a [matplotlib.axes.Axes object](\"https://matplotlib.org/stable/api/axes_api.html\")\n",
    "\n",
    "There is, however, a quick hitch. We can no longer rely directly on the figure-level functions of `relplot`, `catplot` and `displot`. In order to plot correctly on these subpanels, we'll need to use the ___axes-level___ function versions instead. We'll add our plots one at a time so you can see the effect of each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0), colspan=1, rowspan=2)\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1), colspan=2)\n",
    "ax3 = plt.subplot2grid((2, 3), (1, 1), colspan=2)\n",
    "\n",
    "# ---------- Plot 1 ----------#\n",
    "## 5.1.1 Add the first plot - scatterplot\n",
    "# We'll build a scatterplot\n",
    "sns.scatterplot(data = infectionSig_data,   \n",
    "                x = \"area\", \n",
    "                y = \"area.infected\",     \n",
    "                hue = \"strain\",          # Set the point-colour by strain\n",
    "                alpha = 0.6,             # Set the transparency of the points\n",
    "                ax = ...                 # Set this plot to ax1 within the fig object\n",
    "               )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis in ax1\n",
    "...set_xlabel(\"total area\")\n",
    "...set_ylabel(\"area infected\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5.1.2 Add multiple layers to the second `Axes`\n",
    "\n",
    "Next we'll populate the second panel (top-right) with a combination of `violinplot()` and `swarmplot()`. We'll adjust the x-axis tick labels through the `.tick_params()` method as we will be dealing with the `matplotlib.axes.Axes` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0), colspan=1, rowspan=2)\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1), colspan=2)\n",
    "ax3 = plt.subplot2grid((2, 3), (1, 1), colspan=2)\n",
    "\n",
    "# ---------- Plot 1 ----------#\n",
    "# Add the first plot - scatterplot\n",
    "# We'll build a scatterplot\n",
    "sns.scatterplot(data = infectionSig_data,   \n",
    "                x = \"area\", \n",
    "                y = \"area.infected\",     \n",
    "                hue = \"strain\",          # Set the point-colour by strain\n",
    "                alpha = 0.6,             # Set the transparency of the points\n",
    "                ax = ax1                 # Set this plot to ax1 within the fig object\n",
    "               )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax1\n",
    "ax1.set_xlabel(\"total area\")\n",
    "ax1.set_ylabel(\"area infected\")\n",
    "\n",
    "# ---------- Plot 2 ----------#\n",
    "# Add the second plot - violin plot\n",
    "## 5.1.2 Use violinplot to make our split violin\n",
    "sns.violinplot(data = mean_embryo_data,\n",
    "               x = \"wormStrain\", y = \"numEmbryos\", \n",
    "               hue = \"doseLevel\",\n",
    "               split = True,              # This will create hybrid violin plots\n",
    "               inner = \"quartile\",        # Add quartile markers to each half of the violin\n",
    "               palette = {\"Mock\":\"crimson\", \"Medium\": \"deepskyblue\"},    # Update the palette of our violins\n",
    "               density_norm=\"width\",      # Make all the widths the same\n",
    "               alpha = 0.6,\n",
    "               ax = ...                   # Set this plot to ax2 within the fig object\n",
    "              )\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "sns.stripplot(data = mean_embryo_data,\n",
    "              x = \"wormStrain\", \n",
    "              y = \"numEmbryos\", \n",
    "              hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "              palette=\"dark:black\",     # Recolour all of the points to black\n",
    "              size=5,                   # Set the size of our points so they can be seen\n",
    "              dodge = True,\n",
    "              ax = ...\n",
    "             )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax2\n",
    "ax2.set_xlabel(\"worm strain\")\n",
    "ax2.set_ylabel(\"mean embryos per replicate\")\n",
    "\n",
    "# 5.1.2 Alter the xtick attributes\n",
    "ax2...(axis = 'x', rotation = 90)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5.1.3 Move or remove the legend from your plot with various methods\n",
    "\n",
    "So you can see there's a slight issue above with our legend in the violin plot. Due to plotting both a violin and strip plot together, we get a legend with both the colouring and the points. The points from the strip plot don't have much meaning so we can remove those easily in the call to `sns.stripplot()` by using the `legend = False` parameter.\n",
    "\n",
    "If you wanted to remove the legend altogether, you could use the `.get_legend().remove()` command. This would access the `Axes` object legend and remove it. \n",
    "\n",
    "You can also move the legend around with the `seaborn.move_legend()` function which requires 2 parameters:\n",
    "- `obj`: the axes object (plot) you want to move around\n",
    "- `loc`: the location you wish to place the legend. This can be in the form of a combination of upper/center/lower and left/center/right."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0), colspan=1, rowspan=2)\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1), colspan=2)\n",
    "ax3 = plt.subplot2grid((2, 3), (1, 1), colspan=2)\n",
    "\n",
    "# ---------- Plot 1 ----------#\n",
    "# Add the first plot - scatterplot\n",
    "# We'll build a scatterplot\n",
    "sns.scatterplot(data = infectionSig_data,   \n",
    "                x = \"area\", \n",
    "                y = \"area.infected\",     \n",
    "                hue = \"strain\",          # Set the point-colour by strain\n",
    "                alpha = 0.6,             # Set the transparency of the points\n",
    "                ax = ax1                 # Set this plot to ax1 within the fig object\n",
    "               )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax1\n",
    "ax1.set_xlabel(\"total area\")\n",
    "ax1.set_ylabel(\"area infected\")\n",
    "\n",
    "# ---------- Plot 2 ----------#\n",
    "# Add the second plot - violin plot\n",
    "# Use violinplot to make our split violin\n",
    "sns.violinplot(data = mean_embryo_data,\n",
    "               x = \"wormStrain\", y = \"numEmbryos\", \n",
    "               hue = \"doseLevel\",\n",
    "               split = True,              # This will create hybrid violin plots\n",
    "               inner = \"quartile\",        # Add quartile markers to each half of the violin\n",
    "               palette = {\"Mock\":\"crimson\", \"Medium\": \"deepskyblue\"},    # Update the palette of our violins\n",
    "               density_norm=\"width\",      # Make all the widths the same\n",
    "               alpha = 0.6,\n",
    "               ax = ax2                   # Set this plot to ax2 within the fig object\n",
    "              )\n",
    "\n",
    "sns.move_legend(ax2, ...)       ## 5.1.3 Move the legend to the lower right portion of the plot\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "sns.stripplot(data = mean_embryo_data,\n",
    "              x = \"wormStrain\", \n",
    "              y = \"numEmbryos\", \n",
    "              hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "              palette=\"dark:black\",     # Recolour all of the points to black\n",
    "              size=5,                   # Set the size of our points so they can be seen\n",
    "              dodge = True,\n",
    "              legend = False,           ## 5.1.3 Set the legend to false to remove it from the plot\n",
    "              ax = ax2\n",
    "             )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax2\n",
    "ax2.set_xlabel(\"worm strain\")\n",
    "ax2.set_ylabel(\"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "ax2.tick_params(axis = 'x', rotation = 90)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5.1.4 Add a KDE plot to our final panel\n",
    "\n",
    "Let's complete the set by adding a KDE plot to our final panel. We'll filter our dataset on the fly as we pass it to the `kdeplot()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0), colspan=1, rowspan=2)\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1), colspan=2)\n",
    "ax3 = plt.subplot2grid((2, 3), (1, 1), colspan=2)\n",
    "\n",
    "# ---------- Plot 1 ----------#\n",
    "# Add the first plot - scatterplot\n",
    "# We'll build a scatterplot\n",
    "sns.scatterplot(data = infectionSig_data,   \n",
    "                x = \"area\", \n",
    "                y = \"area.infected\",     \n",
    "                hue = \"strain\",          # Set the point-colour by strain\n",
    "                alpha = 0.6,             # Set the transparency of the points\n",
    "                ax = ax1                 # Set this plot to ax1 within the fig object\n",
    "               )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax1\n",
    "ax1.set_xlabel(\"total area\")\n",
    "ax1.set_ylabel(\"area infected\")\n",
    "\n",
    "# ---------- Plot 2 ----------#\n",
    "# Add the second plot - violin plot\n",
    "# Use violinplot to make our split violin\n",
    "sns.violinplot(data = mean_embryo_data,\n",
    "               x = \"wormStrain\", y = \"numEmbryos\", \n",
    "               hue = \"doseLevel\",\n",
    "               split = True,              # This will create hybrid violin plots\n",
    "               inner = \"quartile\",        # Add quartile markers to each half of the violin\n",
    "               palette = {\"Mock\":\"crimson\", \"Medium\": \"deepskyblue\"},    # Update the palette of our violins\n",
    "               density_norm=\"width\",      # Make all the widths the same\n",
    "               alpha = 0.6,\n",
    "               ax = ax2                   # Set this plot to ax2 within the fig object\n",
    "              )\n",
    "\n",
    "sns.move_legend(ax2, \"lower right\")       ## 5.1.3 Move the legend to the lower right portion of the plot\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "sns.stripplot(data = mean_embryo_data,\n",
    "              x = \"wormStrain\", \n",
    "              y = \"numEmbryos\", \n",
    "              hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "              palette=\"dark:black\",     # Recolour all of the points to black\n",
    "              size=5,                   # Set the size of our points so they can be seen\n",
    "              dodge = True,\n",
    "              legend = False,           ## 5.1.3 Set the legend to false to remove it from the plot\n",
    "              ax = ax2\n",
    "             )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax2\n",
    "ax2.set_xlabel(\"worm strain\")\n",
    "ax2.set_ylabel(\"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "ax2.tick_params(axis = 'x', rotation = 90)\n",
    "\n",
    "# ---------- Plot 3 ----------#\n",
    "# 5.1.4 rebuild a KDE plot\n",
    "...(data = N2_mock_data,                      \n",
    "            x = \"numEmbryos\",\n",
    "            hue = \"date\", \n",
    "            fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "            ax = ax3                 # Set this plot to ax2 within the fig object\n",
    "           )\n",
    "\n",
    "# Directly set the x-axis label\n",
    "ax3.set_xlabel(\"mean embryos per animal\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5.1.5 Adjust spacing between `Axes` using `tight_layout()`\n",
    "\n",
    "Nearly there, we can see there are some overlapping text issues from the y-axis labels of the KDE plot onto the scatter plot beside it. This is due to the size of the y-tick text itself pushing the title too far left. We can ask the plot to fix these spacing issues using the `plt.tight_layout()` method, which should fix the overlapping issues as best as it can."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a figure\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "# Generate 3 subplots onto our current figure\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0), colspan=1, rowspan=2)\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1), colspan=2)\n",
    "ax3 = plt.subplot2grid((2, 3), (1, 1), colspan=2)\n",
    "\n",
    "# ---------- Plot 1 ----------#\n",
    "# Add the first plot - scatterplot\n",
    "# We'll build a scatterplot\n",
    "sns.scatterplot(data = infectionSig_data,   \n",
    "                x = \"area\", \n",
    "                y = \"area.infected\",     \n",
    "                hue = \"strain\",          # Set the point-colour by strain\n",
    "                alpha = 0.6,             # Set the transparency of the points\n",
    "                ax = ax1                 # Set this plot to ax1 within the fig object\n",
    "               )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax1\n",
    "ax1.set_xlabel(\"total area\")\n",
    "ax1.set_ylabel(\"area infected\")\n",
    "\n",
    "# ---------- Plot 2 ----------#\n",
    "# Add the second plot - violin plot\n",
    "# Use violinplot to make our split violin\n",
    "sns.violinplot(data = mean_embryo_data,\n",
    "               x = \"wormStrain\", y = \"numEmbryos\", \n",
    "               hue = \"doseLevel\",\n",
    "               split = True,              # This will create hybrid violin plots\n",
    "               inner = \"quartile\",        # Add quartile markers to each half of the violin\n",
    "               palette = {\"Mock\":\"crimson\", \"Medium\": \"deepskyblue\"},    # Update the palette of our violins\n",
    "               density_norm=\"width\",      # Make all the widths the same\n",
    "               alpha = 0.6,\n",
    "               ax = ax2                   # Set this plot to ax2 within the fig object\n",
    "              )\n",
    "\n",
    "sns.move_legend(ax2, \"lower right\")       ## 5.1.3 Move the legend to the lower left portion of the plot\n",
    "\n",
    "# Overlay a swarmplot on our catplot\n",
    "sns.stripplot(data = mean_embryo_data,\n",
    "              x = \"wormStrain\", \n",
    "              y = \"numEmbryos\", \n",
    "              hue = \"doseLevel\",        # split the points by dose level just like nested boxplots\n",
    "              palette=\"dark:black\",     # Recolour all of the points to black\n",
    "              size=5,                   # Set the size of our points so they can be seen\n",
    "              dodge = True,\n",
    "              legend = False,           ## 5.1.3 Set the legend to false to remove it from the plot\n",
    "              ax = ax2\n",
    "             )\n",
    "\n",
    "# Directly set the labels on the x-axis and y-axis of ax2\n",
    "ax2.set_xlabel(\"worm strain\")\n",
    "ax2.set_ylabel(\"mean embryos per replicate\")\n",
    "\n",
    "# Alter the xtick attributes\n",
    "ax2.tick_params(axis = 'x', rotation = 90)\n",
    "\n",
    "# ---------- Plot 3 ----------#\n",
    "# rebuild a KDE plot\n",
    "sns.kdeplot(data = N2_mock_data,                      \n",
    "            x = \"numEmbryos\",\n",
    "            hue = \"date\", \n",
    "            fill = True,             # The fill parameter is passed on to kdeplot()\n",
    "            ax = ax3                 # Set this plot to ax2 within the fig object\n",
    "           )\n",
    "\n",
    "# Directly set the x-axis label\n",
    "ax3.set_xlabel(\"mean embryos per animal\")\n",
    "\n",
    "## 5.1.5 Re-adjust the axes to remove overlap due to axis text\n",
    "...\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not too shabby! And we're done!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 6.0.0 Class summary\n",
    "\n",
    "That's our fourth class on Python! You've made it through and we've learned about taking advantage of in-built DataFrame methods for exploratory data analyis as well as how to finally visualize some of your data:\n",
    "\n",
    "1. Exploratory data analysis with `groupby()` and aggregation functions.\n",
    "2. Basic plots with the `matplotlib.pyplot` module.\n",
    "3. Advanced visualizations with the `seaborn` package.\n",
    "4. Saving your figures to file\n",
    "5. Plotting multi-panel `seaborn` figures through the `pyplot` package."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 6.1.0 Submit your completed skeleton notebook (2% of final grade)\n",
    "\n",
    "At the end of this lecture a Quercus assignment portal will be available to submit your completed skeletons from today (including the comprehension question answers!). These will be due one week later, before the next lecture. Each lecture skeleton is worth 2% of your final grade but a __bonus 0.5%__ will also be awarded for submissions made within 24 hours from the end of lecture (ie 1600 hours the following day). \n",
    "\n",
    "1. From the Jupyter Notebook, select `File> Save Notebook` or use `ctrl + S`\n",
    "2. Return to the Jupyter folder page (ie click on the top-left \"jupyterhub\" icon)\n",
    "3. Locate your .ipynb file and click on the check box to its left side.\n",
    "4. If it is still active it will have a green icon beside its name. Above the file list will be option to `Shut Down`. After clicking on this, you may re-select the file and above the file list choose `Download` to save to your local hard drive.\n",
    "5. Upload your .ipynb file to the appropriate Quercus assignment portal.\n",
    "\n",
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/save_ipynb.png?raw=true\" width=700>|\n",
    "|:--:|\n",
    "|A sample screen shot for saving your completed skeleton notebooks|\n",
    "\n",
    "## 6.2.0 Post-lecture DataCamp assessment (5% of final grade)\n",
    "\n",
    "Soon after the end of each lecture, a homework assignment will be available for you in [DataCamp](https://www.datacamp.com/). Your assignment is to complete the __Introduction to Data Visualization with Seaborn__ course (4 chapters, 4200 possible points). This is a pass-fail assignment, and in order to pass you need to achieve a least 3150 points (75%) of the total possible points. Note that when you take hints from the DataCamp chapter, it will reduce your total earned points for that chapter.\n",
    "\n",
    "In order to properly assess your progress on DataCamp, at the end of each chapter, please print a PDF of the summary. You can do so by following these steps:\n",
    "\n",
    "1. Navigate to the **`Learn`** section along the top menu bar of DataCamp. This will bring up a sidebar menu where you can access the various courses you have been assigned under **`Assignments`** or you can scroll down on the page to find the **`My Assignments`** section. Click on the relevant chapter or course and this may bring you back to within the course itself. \n",
    "2. You should now be on a Course Summary page (See figure below). You can expand each __chapter__ of the course by clicking on the **`VIEW CHAPTER DETAILS`** link. ___DO THIS FOR ALL COMPLETED SECTIONS ON THE PAGE!___ \n",
    "3. Select the visible text in the course Window (not from any top or side menus) using a click and drag technique. Using something like __Ctrl+A will not properly select the course text__. \n",
    "4. Print the page from your browser menu and make sure you have \"Selection Only\" from the Settings options on the Print dialogue. Save as a single PDF. Note: if you don't select the text area correctly (at least in Google Chrome) you may not be able to print the full page. Your results should look something like this:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/DataCamp.example.png?raw=true\" width=700>|\n",
    "|:--:|\n",
    "|A sample screen shot for one of the DataCamp assignments. You'll want to try and print off a single PDF of this section from **Learn > My Assignments**|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may need to take several screenshots if you cannot print it all in a single try. Submit the file(s) or a combined PDF for the homework to the assignment portal on Quercus. By submitting your scores for each section, and chapter, we can keep track of your progress, identify knowledge gaps, and produce a standardized way for you to check on your assignment \"grades\" throughout the course.\n",
    "\n",
    "You will have until 12:59 hours on Tuesday, February 3rd to submit your assignment (right before the next lecture)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 6.3.0 Acknowledgements\n",
    "\n",
    "__Revision 1.0.0__: materials prepared for __CSB1021H S LEC0140__, 01-2022 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.0__: edited and prepared for __CSB1021H S LEC0140__, 01-2023 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.2__: edited and prepared for __CSB1021H S LEC0140__, 01-2024 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.3__: edited and prepared for __CSB1021H S LEC0140__, 01-2025 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.4__: edited and prepared for __CSB1021H S LEC0140__, 01-2026 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 7.0.0 Appendix 1 - In case your kernel should crash\n",
    "\n",
    "In the case that your kernel crashes, you can use this code cell to recreate all of the data that is used in __section 5.0.0__. Convert the below cell into a coding cell by using the \"Y\" key when highlighting it in \"Command\" mode. Then simply run the cell and it should recreate the datasets needed."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# ----- Always run this at the beginning of class so we can get multi-command output ----- #\n",
    "\n",
    "# Access options from the iPython core\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "\n",
    "# Change the value of ast_node_interactivity\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "# ----- Additional packages we want to import for class ----- #\n",
    "\n",
    "# Import the pandas package\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# !pip install seaborn\n",
    "\n",
    "# Read in embryo_long_merged.csv\n",
    "embryo_merged = pd.read_csv('data/embryo_long_merged.csv') \n",
    "\n",
    "# Subset the data from our dataset\n",
    "embryo_merged_subset = embryo_merged.iloc[:, :19]\n",
    "\n",
    "# Use the insert method\n",
    "embryo_merged_subset.insert(loc = 4,                   # Location to insert at\n",
    "                            column = 'pathogenDose',   # The column name after insertion\n",
    "                            \n",
    "                            # To calculate the value, we'll pop from the subset\n",
    "                            value = (embryo_merged_subset.pop('pathogenDose')\n",
    "                            \n",
    "                                     # Break up the float from the \"M\"\n",
    "                                     .str.split(pat = \"M\", expand = True)\n",
    "                            \n",
    "                                     # Convert the first column to a float, and then provide only that to insert\n",
    "                                     .astype({0:'float64'})[0])\n",
    "                           )\n",
    "\n",
    "# Filter the data by N2 animals with a pathogenDose of 0\n",
    "N2_mock_data = (embryo_merged_subset.query('wormStrain == \"N2\" & pathogenDose == 0')\n",
    "                # Convert our date variable with the astype() method\n",
    "                .astype({'date':'category'})\n",
    "               )\n",
    "\n",
    "# Take the unique values and flip the order\n",
    "date_list = N2_mock_data.date.unique()[::-1].tolist()\n",
    "\n",
    "# We'll need to pull out and replace the date column\n",
    "N2_mock_data['date'] = (N2_mock_data['date']\n",
    "                        # access the categorical property\n",
    "                        .cat\n",
    "                        # Reorder the categorical data\n",
    "                        .reorder_categories(new_categories = date_list)\n",
    "                       )\n",
    "\n",
    "# Filter for only uninfected data\n",
    "mean_embryo_data = (embryo_merged_subset.query('doseLevel in [\"Mock\", \"Medium\"]')  # filter by doseLevel\n",
    " \n",
    "                    # Group by infection experiment \n",
    "                    .groupby(by = ['date', 'wormStrain', 'pathogenStrain', 'doseLevel'])\n",
    "\n",
    "                    # Create the frequency table on numEmbryos\n",
    "                    .agg({'numEmbryos': 'mean'})\n",
    "\n",
    "                    # Recall, when we reset the index, it converts the indices back into columns\n",
    "                    .reset_index()\n",
    "                    \n",
    "                    .astype({'wormStrain':'category', 'doseLevel':'category'})\n",
    ")\n",
    "\n",
    "# Pull out our doseLevel data and replace it with a newly categorized version\n",
    "mean_embryo_data['doseLevel'] = (mean_embryo_data['doseLevel']\n",
    "                                 .cat\n",
    "                                 .reorder_categories(new_categories = [\"Mock\", \"Medium\"])\n",
    "                                )\n",
    "\n",
    "# Read the infection signal data in from file\n",
    "infectionSig_data = pd.read_csv(\"data/infection_signal.tsv\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## The Center for the Analysis of Genome Evolution and Function (CAGEF)\n",
    "\n",
    "The Centre for the Analysis of Genome Evolution and Function (CAGEF) at the University of Toronto offers comprehensive experimental design, research, and analysis services in microbiome and metagenomic studies, genomics, proteomics, and bioinformatics.\n",
    "\n",
    "From targeted DNA amplicon sequencing to transcriptomes, whole genomes, and metagenomes, from protein identification to post-translational modification, CAGEF has the tools and knowledge to support your research. Our state-of-the-art facility and experienced research staff provide a broad range of services, including both standard analyses and techniques developed by our team. In particular, we have special expertise in microbial, plant, and environmental systems.\n",
    "\n",
    "For more information about us and the services we offer, please visit <https://www.cagef.utoronto.ca/>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroR/CAGEF_new.png?raw=true\" width=700></center>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "335.455px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
