{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/CAGEF_services_slide.png?raw=true\"  width=700></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Python for Data Science\n",
    "\n",
    "# Lecture 02: Lists, Arrays, and DataFrames - Oh My!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Student Name:\n",
    "\n",
    "## Student ID:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.1.0 About Introduction to Python\n",
    "\n",
    "Introduction to Python is brought to you by the **Centre for the Analysis of Genome Evolution & Function (CAGEF)** bioinformatics training initiative. This course was developed based on feedback on the needs and interests of the Department of Cell & Systems Biology and the Department of Ecology and Evolutionary Biology.\n",
    "\n",
    "The structure of this course is a code-along style so it is 100% hands on! A few hours prior to each lecture, the materials will be avaialable for download at [QUERCUS](https://q.utoronto.ca/). The teaching materials will consist of a Jupyter Lab Notebook with concepts, comments, instructions, and blank spaces that you will fill out with Python code along with the instructor. Other teaching materials include a live version of the notebook, and datasets to import into Python - when required. This learning approach will allow you to spend the time coding and not taking notes! \n",
    "\n",
    "As we go along, there will be some in-class challenge questions for you to solve either individually or in cooperation with your peers. Post lecture assessments will also be available (see syllabus for grading scheme and percentages of the final mark) through [DataCamp](https://Datacamp.com) to help cement and/or extend what you learn each week.\n",
    "\n",
    "### 0.1.1 Where is this course headed?\n",
    "\n",
    "We'll take a blank slate approach here to Python and assume that you pretty much know _nothing_ about programming. From the beginning of this course to the end, we want to get you from some potential scenarios:\n",
    "\n",
    "- A pile of data (like an excel file or tab-separated file) full of experimental observations and you don't know what to do with it. \n",
    "\n",
    "- Maybe you're manipulating large tables all in excel, making custom formulas and pivot table with graphs. Now you have to repeat similar experiments and do the analysis again. \n",
    "\n",
    "- You're generating high-throughput data and there aren't any bioinformaticians around to help you sort it out.\n",
    "\n",
    "- You heard about Python and what it _could_ do for your data analysis but don't know what that means or where to start.\n",
    "\n",
    "and get you to a point where you can:\n",
    "\n",
    "- Format your data correctly for analysis\n",
    "\n",
    "- Produce basic plots and perform exploratory analysis\n",
    "\n",
    "- Make functions and scripts for re-analysing existing or new data sets\n",
    "\n",
    "- Track your experiments in a digital notebook like Jupyter!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1.2 How do we get there? Step-by-step.\n",
    "\n",
    "In the first two lessons, we will talk about the basic data structures and objects in Python, get cozy with the Jupyter Notebook environment, and learn how to get help when you are stuck. Because everyone gets stuck - a lot! Then you will learn how to get your data in and out of Python, how to tidy our data (data wrangling), subset and merge data. We'll take a break from data wrangling to spend our fourth lecture learning how to generate exploratory data plots. Then we'll slide into using the power of python and programming to use flow control before visiting text manipulation techniques in lectures 5 and 6. Data cleaning and string manipulation is really the battleground of coding - getting your data into the format where you can analyse it. Lastly, we will learn to write customized functions to help scale up your analyses. <br><br>\n",
    "\n",
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/Draw_an_Owl.jpg?raw=true\"  width=700></center>\n",
    "\n",
    "The structure of the class is a __code-along__ style: It is fully hands on. At the end of each lecture, the complete notes will be made available in a PDF format through the corresponding Quercus module so you don't have to spend your attention on taking notes. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### 0.1.3 What kind of coding style will we learn?\n",
    "\n",
    "There is no single path correct from A to B - although some paths may be more elegant, or more efficient than others. With that in mind, the emphasis in this lecture series will be on:\n",
    "\n",
    "1. **Code simplicity** - learn helpful functions that allow you to focus on understanding the basic tenets of good data wrangling (reformatting) to facilitate quick exploratory data analysis and visualization. \n",
    "2. **Code readability** - format and comment your code for yourself and others so that even those with minimal experience in R will be able to quickly grasp the overall steps in your code.\n",
    "3. **Code stability** - while the core Python code is relatively stable, behaviours of functions can still change with updates. There are well-developed packages we'll focus on for our analyses. Namely, we'll become more familiar with the `pandas` series of packages for working with tabular data. This resource is well-maintained by a large community of developers. While not always the \"fastest\" approach this additional layer can help ensure your code still runs (somewhat) smoothly later down the road."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.2.0 Lecture objectives\n",
    "\n",
    "Welcome to this second lecture in a series of seven. Today you will dive into more detailed data structures, packages that works with them, and build up towards our a more standard data science structure - the DataFrame.\n",
    "\n",
    "At the end of this lecture we will aim to have covered the following topics:\n",
    "\n",
    "1. Python lists\n",
    "2. Python dictionaries\n",
    "3. The `NumPy` package and arrays\n",
    "4. The `Pandas` package and DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.3.0 A legend for text format in Jupyter markdown\n",
    "\n",
    "`grey background` - a package, function, code, command or directory. Backticks are also use for in-line code.      \n",
    "*italics* - an important term or concept or an individual file or folder     \n",
    "**bold** - heading or a term that is being defined      \n",
    "<span style=\"color:blue\">blue text</span> - named or unnamed hyperlink\n",
    "\n",
    "`...` - Within each coding cell this will indicate an area of code that students will need to complete for the code cell to run correctly. \n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Blue box:</b> A key concept that is being introduced\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Yellow box:</b> Risk or caution\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Green boxes:</b> Recommended reads and resources to learn Python\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>Red boxes:</b> A comprehension question which may or may not involve a coding cell. You usually find these at the end of a section. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.4.0 Lecture and data files used in this course\n",
    "\n",
    "### 0.4.1 Weekly Lecture and skeleton files\n",
    "\n",
    "Each week, new lesson files will appear within your JupyterHub folders. We are pulling from a GitHub repository using this [Repository git-pull link](https://jupyter.utoronto.ca/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2Fuoft-csb-datasci%2F2026-01-IntroPython&urlpath=tree%2F2026-01-IntroPython%2F&branch=main). Simply click on the link and it will take you to the University of Toronto JupyterHub. You will need to use your UTORid credentials to complete the login process. From there you will find each week's lecture files in the directory `/2026-01-IntroPython/Lecture_XX`. You will find a partially coded `skeleton.ipynb` file as well as all of the data files necessary to run the week's lecture.\n",
    "\n",
    "Alternatively, you can download the Jupyter Notebook (`.ipynb`) and data files from JupyterHub to your personal computer if you would like to run independently of the JupyterHub.\n",
    "\n",
    "### 0.4.2 Live-coding HTML page\n",
    "\n",
    "A live lecture version will be available at [uoft-csb-datasci.github.io](https://uoft-csb-datasci.github.io/2026-01.IntroPython/index.html) that will update as the lecture progresses. Be sure to refresh to take a look if you get lost!\n",
    "\n",
    "### 0.4.3 Post-lecture PDFs\n",
    "\n",
    "As mentioned above, at the end of each lecture there will be a completed version of the lecture code released as a PDF file under the Modules section of Quercus.\n",
    "### 0.4.4 Data used in this lesson\n",
    "\n",
    "No outside datasets will be imported or used in this lecture!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 0.5.0 Packages used in this lesson\n",
    "\n",
    "`IPython` and `InteractiveShell` will be access just to set the behaviour we want for iPython so we can see multiple code outputs per code cell.\n",
    "\n",
    "`random` is a package with methods to add pseudorandomness to programs \n",
    "\n",
    "`numpy` provides a number of mathematical functions as well as the special data class of `arrays` which we'll be learning about today."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----- Always run this at the beginning of class so we can get multi-command output ----- #\n",
    "\n",
    "# Access options from the iPython core\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "\n",
    "# Change the value of ast_node_interactivity\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "# ----- Additional packages we want to import for class ----- #\n",
    "\n",
    "import random         # Used to help us generate random data\n",
    "import numpy as np    # Contains data structures and functions for working with sets of numbers\n",
    "import pandas as pd   # Contains data structures and functions for working with tabular data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 1.0.0 Python data structures have properties and functions\n",
    "\n",
    "As discussed in __Lecture 01__, everything in Python is an object: \n",
    "\n",
    "- integers\n",
    "- floats\n",
    "- strings \n",
    "- booleans \n",
    "\n",
    "The above are all objects but also _data types_ in Python. We can combine these data types in **data structures** to properly store, format, and model our data. The decision of what data structure to use depends on the objective(s), data types, and the tasks to perform. For example, some data structures can handle only one data type at a time (all numeric or all character) but are computationally very fast. Others structures can store several data types but are computationally very expensive and slow, especially when we have large datasets (thousands of rows and columns). \n",
    "\n",
    "Another feature to look for on data structures is their mutability; some structures can be altered or updated after they are created (__mutable__), some are not (__immutable__). Let's take a look at some of Python's core data types (built-in data structures). \n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Further Reading:</b> If you'd like to learn in more detail about what makes a data object mutable or immutable, you can find a good breakdown and set of examples from <a href=\"https://medium.com/@meghamohan/mutable-and-immutable-side-of-python-c2145cf72747\">Megha Mohan</a>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.1.0 Lists are ordered collections of arbitrary data types\n",
    "\n",
    "The first Python data structure that we will introduce is `lists`, which are __ordered__ collections of data of one __or__ several types (strings, booleans, etc.) where each datum is called an ___element___ or ___item___. Lists are easily identifiable because of the _squared brackets_, `[ ]`, that enclose their elements. \n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Caution:</b> In other programming languages, Python's lists are usually called <b>arrays or vectors</b>, but be aware that Python has its own version of arrays as part of a package called NumPy (<b>section 3.0.0</b> of this lecture).\n",
    "</div>\n",
    "\n",
    "### 1.1.1 A mono-type list hold only a single data type\n",
    "\n",
    "We'll begin our journey into lists by building a couple of lists that have only single data types within them. These are the simplest kinds of lists to work with. We'll put our elements in the `[ ]` to initialize the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a basic list. What data types are these? \n",
    "..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make another mono-type list. \n",
    "\n",
    "...\n",
    "\n",
    "# What data type am I attempting to store and why is it not working?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.1.2 Lists of strings must be defined correctly!\n",
    "\n",
    "From above, we attempted to generate a mono-type list of strings. As you might recall from last week, however, strings are always enclosed by a set of single quotes `' '` or double quotes `\" \"`. Above, we tried to generate a list of _variables_ which we had not yet defined anywhere in memory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a proper mono-type list of strings\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.2.0 A list can contain _different_ data types\n",
    "\n",
    "When working with different data types and information that you'd like to pass around or hold onto, it's convenient to know that you can combine this information into a single list. As mentioned above, lists can be arbitrary types so long as you remember the order of your elements, you can put a variety of data types into the same list object and you won't be subject to coercion.\n",
    "\n",
    "Let's try!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a list of arbitrary elements\n",
    "\n",
    "[\"order\", 20, 30.5, ...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.2.1 Use `list()` to initialize an empty list\n",
    "\n",
    "When writing flow control programs (__Lecture 05__), we need to create empty structures in advance so the program has a place to write its output. This is called \"initialization\" and to create a list we use the `list()` function. In fact, all classes (which represent objects) will have some kind of initializer to create an object even if they are essentially \"empty\" containers at their outset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "... # this is an empty list\n",
    "\n",
    "... # another way to create an empty list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember you can inquire about the class of an object with `type()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type([]) # Retrieve a simple\n",
    "\n",
    "print(type([])) # Or a complex answer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.2.2 Split a string into individual characters using `list()`\n",
    "\n",
    "Recall that a `string` object can be conceptualized as an ordered queue of individual characters. Sometimes you may have a complex string that you'd like to break into it's individual characters. Although you can already access each character with the index `[ ]` operators, you may want to treat those characters _specifically_ as a list. You can create this kind of separated list by simply using the `list()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert your string into individual characters\n",
    "list(...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.3.0 Lists can be assigned to variables\n",
    "\n",
    "So far we've just been making lists that disappear from memory but you may want to make a list that you can pass around, grow, shrink, or pull information from. Let's make a few to work with now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make some lists for us to look at\n",
    "list_1 = [\"genome\", 20, 30.5, True]\n",
    "list_2 = [1, 100, 10, 25]\n",
    "list_3 = [3]\n",
    "\n",
    "# Take a look at those lists individually\n",
    "list_1\n",
    "list_2\n",
    "...\n",
    "\n",
    "# Print out all the lists in a single line. Look carefully at how they are displayed!\n",
    "print(list_1, list_2, ...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.4.0 Access portions of a list using the indexing `[ ]` operators\n",
    "\n",
    "Items in a list can be accessed using squared brackets the same way we did with strings (__Lecture 01__). Many of the same indexing methods and mechanisms work the same way as with strings. Recall we use the `[index]` syntax to access a single element of our list which is also ___zero-indexed___."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Access the first element of list_1\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The items in a list can be modified by indexing the item that we want to change (remember that lists are ___mutable___)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_1 [0] = ... # The space between list_1 and [0] is not needed but improves readability for some\n",
    "\n",
    "list_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Mapping:</b> The relationship between indices and items in data structures is many-to-one. Each index \"maps to\" one element/object within the list. Each element, however, can have two or three possible indices to access it: a positive and negative integer, and potentially the half-open intervals we discussed in <b>Lecture 1</b>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can also use negative indices like we do with strings!\n",
    "list_1[-3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.5.0 Check for presence of elements using `in`\n",
    "\n",
    "We can ask if a __single element__ is present within a list using the `in` keyword. This can be a quick way to determine if your list has the element or item you are seeking. The `in` keyword will produce a boolean result as its output. Looking for multiple elements will have to wait for _another lecture_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for \"genomics\" in your list\n",
    "\"genomics\" ... list_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for \"genome\" in your list\n",
    "... list_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.6.0 Perform basic functions on lists\n",
    "\n",
    "We can perform mathematical operations on lists and between lists as long as they hold data of the correct type! Let's explore `list_2` using some built-in functions such as \n",
    "- `len()` (length)\n",
    "- `max()` (maximum)\n",
    "- `min()` (minimum)\n",
    "- `sum()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recall that list_2 was a set of numeric values\n",
    "list_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "...(list_2) # 4 elements long\n",
    "\n",
    "...(list_2) # the max number in list_2 is 100\n",
    "\n",
    "...(list_2) # the min number is 1\n",
    "\n",
    "sum(list_2) # the sum of list_2 is 136\n",
    "\n",
    "sum(list_2) / len(list_2) # The mean of list_2 as a float\n",
    " \n",
    "sum(list_2) // len(list_2) # The mean of list_2 as an integer. Notice the double forward slash!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.7.0 Using \"math\" operators on lists\n",
    "\n",
    "Yes, like most data types you can use basic operators on lists but what will their behaviour be? Remember from last week that we saw different behaviours between numbers and strings! Let's explore further.\n",
    "\n",
    "### 1.7.1 Use the `+` operator to concatenate two or more lists\n",
    "\n",
    "Recall that ___context___ matters when the interpreter decides on the meaning of syntax. Much like with strings, the `+` operator takes on a different behaviour when working with lists versus numbers. Rather than being interpreted as an addition operation, it is a _concatenation_ symbol more akin to its use with strings, allowing you to combine two or more lists together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remind us, what is in list_2?\n",
    "list_2\n",
    "\n",
    "# Make another list of integers\n",
    "list_4 = [2, 20, 200]\n",
    "\n",
    "# Combine the two lists\n",
    "list_5 = list_4 + ...\n",
    "\n",
    "list_5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.7.2 Use the `*` operator to repeat your list\n",
    "\n",
    "Again we see a separate behaviour for a mathematical operator given the context of a list data structure. If you'd like to repeat your list one or more times, you can use the `*` operator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " # Repeat 22 five times\n",
    "[22] ...\n",
    "\n",
    "# Repeat an entire list three times\n",
    "list_2 ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.7.3 Yes, you can slice a list too with `[start:end]`\n",
    "\n",
    "I think we are seeing a pattern now between lists and strings (___why do you think that is?___). So we can slice our lists using the same notation we learned in __Lecture 1__. Just remember that we are working with `[inclusive:exclusive]` form. So what does that really mean? Let's review."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Full list\n",
    "list_2\n",
    "\n",
    "# grab the 2nd, and 3rd element\n",
    "list_2[...]\n",
    "\n",
    "# Grab from the 1st to the 3rd element (index 0-2)\n",
    "list_2[...]\n",
    "\n",
    "# Grab from the 3rd to the last element (index 2-3)\n",
    "list_2[...]\n",
    "\n",
    "# What does this get us?\n",
    "list_2[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.7.3.1 use the `random.seed()` command to set up reproducible \"random\" sequences\n",
    "\n",
    "Let's try something more \"practical\" by generating a random sequence of nucleotides and working with that. We'll introduce the `seed(n)` function from the `random` package. We'll also use the `sample(sequence, n)` function to take an item from our list ___without replacement___. Remember how to access functions from packages? \n",
    "\n",
    "Furthermore, we'll be seeing our first use of a `for` loop but we'll dig deeper into that when looking at flow control (__Lecture 05__).\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>A note on simulating \"random\" events:</b> The only random things you'll find in computer science are whether your programs will run on the first try and whether you'll understand them 6 months from now.<br><br>\n",
    "\n",
    "More seriously, the ability to generate a truly random sequence of numbers is not possible. We can <b><i>approximate</i></b> randomness - notably with special hardware but from a software perspective we can only mimic stochastic processes. Generally our approximations or pseudorandom algorithms are, to the casual observer, just as good as truly random events. They are, however, deterministic and can be repeated if we know the start state of the process. Usually a random number generator might use something like the system time as a seed to initialize its state but if we use a <b><i>specified seed</i></b>, we can get repeatable results.\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a list containing the DNA nucleotides\n",
    "nucleotides = [\"A\", \"C\", \"G\", \"T\"]\n",
    "type(nucleotides)\n",
    "\n",
    "# Random is a package with methods to add pseudorandomness to programs \n",
    "import random\n",
    "random.seed(...) # Seed setting is a good practice to make code reproducible when code involves randomness\n",
    "\n",
    "gene_sequence = [ ]    # Initialize with an empty list\n",
    "\n",
    "# Example of what a for loop looks like (more about flow control on lecture 5 so do not worry about the code right now)\n",
    "# This loop will repeat 4x since that is the length of \"nucleotides\"\n",
    "for base in nucleotides:\n",
    "\n",
    "    # Concatenation of a growing list of random nucleotides. It will grow by 3 each time.\n",
    "    # Notice this indentation? We'll talk more about that in later lectures\n",
    "    gene_sequence = gene_sequence + random.sample(...)\n",
    "    gene_sequence\n",
    "    \n",
    "type(gene_sequence)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Let's reuse that list to make an even bigger list of random basepairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just a little command to make our output look better in Jupyter!\n",
    "%pprint\n",
    "\n",
    "# Let's add a second gene_sequence statement to extend our DNA sequence. \n",
    "# This time, let's take the print statement out of the for loop\n",
    "random.seed(3)\n",
    "\n",
    "gene_sequence = [] # Initialize with an empty list\n",
    "\n",
    "# This loop will repeat 4x since that is the length of \"nucleotides\"\n",
    "for base in nucleotides:\n",
    "    \n",
    "    # Concatenation of a growing list of random nucleotides. It will grow by 6 each loop.\n",
    "    \n",
    "    # Sample 3 random bases from nucleotides\n",
    "    gene_sequence = gene_sequence + random.sample(nucleotides, 3)\n",
    "    # sample 3 random bases from the gene_sequence object\n",
    "    gene_sequence = gene_sequence + random.sample(..., 3)\n",
    "    # gene_sequence\n",
    "\n",
    "gene_sequence\n",
    "type(gene_sequence)\n",
    "\n",
    "### We'll be playing around with gene_sequence quite a bit so let's make a copy\n",
    "gene_sequence_copy = list(gene_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Slice the first 10 nucleotides which are **CGAGACACGG**. They will always be the same sequence because we set a specific `random.seed()`. Also remember that lists use zero-indexing!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember that Python uses zero indexation and the last index in a range is excluded (in this case, index 10)\n",
    "\n",
    "gene_sequence[...] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Can you think of another way to index the first 10 nucleotides?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sequence[ :10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analogously, omitting the second index, returns elements all the way to the last element - just like a `string`!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with this syntax, the last element is included\n",
    "gene_sequence[...] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.7.3.2 Replace elements in your list with slicing\n",
    "\n",
    "Slicing can also be used to update several elements at the time. Replace bases G and A (bases number 2 and 3 from `gene_sequence`) by **R** (R in a DNA sequence means that either adenine or guanine (puRines) can be found at that position) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Replace the 2nd and 3rd element of your list with R\n",
    "gene_sequence[...] = ['R', 'R']\n",
    "\n",
    "gene_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take a close look at the resulting output. Two things to note here are that:\n",
    "\n",
    "1. We inserted our to 'R' entries starting at the 3rd base (index position 2) in our sequence instead of the 2nd base (index position 1).\n",
    "2. We __did not__ replace the bases at indices 2 and 3 but rather replaced the element at index 2 with two new elements, causing the remaining elements to _shift_ over. This converts our sequence __CG-A-GAC__ to __CG-RR-GAC__.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Caution:</b> Whoops! Here is a reminder on the behavior of lists. Now that we've replaced some of the elements <b>we can't undo it!</b> Luckily we made a copy of the list (<b>gene_sequence_copy</b>) when we made it in <b>section 1.7.3.1</b>.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.7.3.3 Referencing an object versus copying it\n",
    "\n",
    "Before we can continue we need to get our backup copy of the original `gene_sequence` which was stored in `gene_sequence_copy`. There are a few ways to go about doing this:\n",
    "\n",
    "1. Initiating a new list with `list()`. This also works with objects like `dict` and `set` which we'll learn about later.\n",
    "2. Using index notation `[:]` which returns a shallow copy!\n",
    "3. Importing the `copy` package and use the `copy()` or ...\n",
    "4. `deepcopy()` functions.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Are you making a copy or pointing at the original?:</b> We have inadvertantly come to a very important point in Python about objects and memory. When we use the assignment statement  <b>object_X = object_Y</b>, we are actually just asking Python <i>\"Where is object_X located in memory?\"</i> and what returns is a reference to that location that is assigned to <b>object_Y</b>. That means we are not copying the object but rather just pointing to it using a different variable! In some cases, this can be problematic because alterations to <b>object_X</b> will also be seen in <b>object_Y</b> and vice versa. </br></br>\n",
    "Instead what we want is to duplicate (or copy) the information in <b>object_X</b> and store it in a new place in memory that is referenced by <b>object_Y</b>! That's where the above functions will help you out. The first three methods return a shallow copy (only the first layer of objects is duplicated) and the fourth will return a fully duplicated and independent copy of the object. </br></br>\n",
    "Learn more about the <a href=\"https://realpython.com/copying-python-objects/#:~:text=A%20shallow%20copy%20means%20constructing,of%20the%20child%20objects%20themselves.\">difference between shallow and deep copies here</a>.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/objectCopyingVsPointing.png?raw=true\"  width=700>|\n",
    "|:--:|\n",
    "|Learn the difference between a pointer and a copy!|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notice that we're making a new list with the list() command?\n",
    "gene_sequence = list(gene_sequence_copy)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Back to our task: to change the second and third nucleotides in `gene_sequence` we should start at index __1__, not at index 2. Recall that Python has zero indexation!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"gene_sequence before modification: \" + str(gene_sequence[0:10]))\n",
    "\n",
    "# Define the indices of interest and, the right side of the equal sign, without squared brackets, add the elements that you want to incorporate\n",
    "gene_sequence[...] =  [\"R\", \"R\"]\n",
    "\n",
    "print(\"gene_sequence after modification:  \" + str(gene_sequence[0:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.8.0 Use the `.join()` method to concatenate list elements\n",
    "\n",
    "Naturally, the most appropriate way to store a gene sequence is as a string, not a as list. Let's type convert `gene_sequence` into a string with the `.join()` method which takes the form of `separator.join(sequence)` where\n",
    "- `sequence` is the given list of elements\n",
    "- `separator` is the character, or string, that you wish to occur _between_ elements of your sequence\n",
    "\n",
    "Let's give it a try!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make gene_sequence a string\n",
    "# The quotes are empty because we are joining the elements \n",
    "# within gene_sequence instead of joining it to another list\n",
    "gene_sequence = ...\n",
    "\n",
    "# Now we have a gene sequence that we can use\n",
    "gene_sequence\n",
    "\n",
    "# What is the type of variable now?\n",
    "type(gene_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.9.0 Additional methods for working with lists\n",
    "\n",
    "Time to move to other aspects of working with lists. Do you know what __methods__ are? In this context, we are still talking about functions that perform operations __BUT__ more specifically, Python objects, such as a `list`, have __methods__ (behaviours) that can be applied on them to carry out functions. We access these with the `.` syntax much like we do the functions from a package. In this case, however, we are accessing methods that belong to the object which is referenced by our variable. Thus our general syntax is `variable.method()`. \n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Syntax Update!:</b> From here on out, we'll use the syntax <b>.method()</b> to describe functions belonging to an object while we will use just <b>function()</b> (no dot!) to describe Python base or package functions.\n",
    "</div>\n",
    "\n",
    "Let's say that we want to add \"biology\" as an element to `list_1`. We know that the `+` operator concatenates two lists together, so we need to covert \"biology\" into a list and then concatenate the two lists together. Easy peasy, right?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's revisit list_1\n",
    "list_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.0.1 The index `[ ]` operators will retain your string structure!\n",
    "\n",
    "Don't forget that when using the `list()` operator on a string, it will automatically break up your string into individual characters! We've seen the way to retain the original string above by using just the index `[ ]` operators to create our list. This can be combined with `list()` although it is a bit redundant:\n",
    "\n",
    "`bio = list([\"biology\"])` or `bio = [\"biology\"]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is rather redundant..\n",
    "bio = list(...)\n",
    "bio\n",
    "\n",
    "# compared to just making the list\n",
    "bio = ...\n",
    "bio\n",
    "\n",
    "# Now you can successfully add it to the list!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Now we just need to concatenate to our `list_1` object with a `+` symbol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate list_1 with bio\n",
    "list_1 + ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.1 Use the `.append()` or `.extend()` methods to quickly add to your list without type-casting!\n",
    "\n",
    "Rather than spend extra code instantiating a variable and type-casting it to a proper list as we did above, you can use the `.append()` method to add a __single element__ to your list. To add multiple elements use the `.extend()` method instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_1 = [\"genome\", 20, 30.5, True]\n",
    "\n",
    "list_1...\n",
    "list_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's add list_2 to list_1\n",
    "list_1...\n",
    "list_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that **list_1** was modified while **list_2** remains intact. No need to _reassign_ (aka use the = sign) the results of `append()` or `extend()` back to `list_1`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_2                                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.2 Use the `.sort()` method on your list to order the elements\n",
    "\n",
    "Another method, `sort()` lets you sort the elements in a list. As a caveat, `.sort()` requires all the elements to be of the same type in order to make proper comparisons. It can come in the form `myList.sort(reverse = (True|False), key = MyFunc)` where:\n",
    "\n",
    "- `reverse` (optional) determines if you sort in ascending (`False` is the default) __or__ descending (`True`) order.\n",
    "- `key ` (optional) can be a function that determines how to sort your elements.\n",
    "\n",
    "Note that as a method, `.sort()` will permanently update your list, just like `.append()` or `.extend()`!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign bio as a list\n",
    "bio = list(\"biology\")\n",
    "bio\n",
    "\n",
    "# Apply bio.sort()\n",
    "bio.sort()\n",
    "\n",
    "# print bio after sorting it\n",
    "bio  # Now bio has changed (mutated)\n",
    "\n",
    "# sort list_2\n",
    "list_2...\n",
    "\n",
    "list_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Recall that `.sort()`'s default behavior is `.sort(reverse = False)` (lowest-to-highest) but it can be overwritten. How would you call on `.sort()` to give you the reverse order?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's do a reverse sort on list_2\n",
    "list_2.sort(...)\n",
    "list_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Parameters vs. arguments:</b> Sometimes a method or function can be changed to alter its default behaviour. In the definition of the function, these are accessed through <b>parameters</b>. The value <i>assigned</i> to the parameter is known as the <b>argument</b> and this could have a default value or one supplied by the user. The default values of a method/function can be accessed using the <b>help()</b> function. Knowing the default behaviour of a method/function is very important, especially for statistical applications and regardless of what programming language or software you use.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.3 Removing elements from the list \n",
    "\n",
    "We've seen how to add and replace elements within your list but sometimes you want to delete an element completely. Let's look at some of the ways to accomplish this.\n",
    "\n",
    "### 1.9.3.1 To remove and retrieve a deleted element, use the `.pop()` method\n",
    "\n",
    "The `.pop(index)` method will ___return___ the removed element, while altering the list object. This way you can save that element, move or copy it to another list, or run analyses on the value/object as needed.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Caution:</b> If no index is provided, the default behaviour of <b>.pop()</b> is to remove the <i>last</i> element of the list like popping the top (last) pancake off a stack of pancakes. It will do so without issuing any warnings at all.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Original list: \", list_1)    # before\n",
    "\n",
    "# Remove the first element of list_1\n",
    "pop_item = list_1...     # pop (dropped element) is stored as a separate object in case we need it\n",
    "print(\"Popped element: \", pop_item)\n",
    "\n",
    "print(\"Updated list: \", list_1)    # after"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.3.2 Use the `del()` function to remove an element\n",
    "\n",
    "If you are not interested in the removed value then you can use the `del()` function. Note how we aren't using the same kind of notation as above? We're using it as a function instead. There are two forms of syntax to use this function but it basically works with slice notation as well!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_4 = [2, 20, 200]\n",
    "list_4\n",
    "\n",
    "# del is a function, reason why it does not use dot notation. \n",
    "# It can also be used with parenthesis: del(list_4[1])\n",
    "...(list_4[1])\n",
    "\n",
    "# alternatively, this syntax can also be used with functions: \n",
    "# del list_4[1] # same as del()\n",
    "\n",
    "list_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.9.3.3 Use the `.remove()` method to remove an element without knowing its index\n",
    "\n",
    "More often that not, we tend to remember what element we want to drop but not its index. The `.remove()` method will find and delete the ___first___ occurrence of an element mentioned in its argument\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Caution:</b> If the element you've provided is not in the list, it will produce an error.\n",
    "</div>\n",
    "\n",
    "Let's remove the number 100 from `list_2`, this time with `.remove()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's set our list again\n",
    "list_2 = [1, 100, 10, 25, 100]\n",
    "\n",
    "# Use the .remove() method\n",
    "list_2...\n",
    "\n",
    "list_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, only the first 100 in the list has been removed. Note that using `remove()` on a non-existent element, will throw an error.\n",
    "\n",
    "***\n",
    "## 1.10.0 Use the `.split()` method to break a string into a list of strings\n",
    "\n",
    "Sometimes you may have a complex string that you'd like to break up based on a pattern, specific character, delimiter etc. To accomplish this we can use the `string` method `.split()`. The resulting output returned, however, is a ___list of strings___ regardless of whether or not the pattern itself is found.\n",
    "\n",
    "Now, let's subset \"biology\" from `list_1`, then use `.split()` to split \"biology\" wherever there is an \"o\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Revisit and check the elements of list_1\n",
    "list_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolate the right element and split it\n",
    "list_1[3]... # Biology is at index 3\n",
    "\n",
    "# This is essentially what is happening\n",
    "\"biology\"...\n",
    "\n",
    "# What happens to list_1? It remains unaltered\n",
    "list_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.11.0 Nesting: making a list within a list\n",
    "\n",
    "As we mentioned in the beginning, a `list` can have an arbitrary grouping of data types/object/elements. That means we can make a `list` that contains a `list`. This is also known as a _nested list_. When it comes to making more complex structures, it means we can provide a hierarchy to a structure using lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Each bracket subsets the list\n",
    "# We can use [:] notation to shallow copy our list\n",
    "list_6 = list_1[...]\n",
    "\n",
    "list_6.append([\"extra\", \"elements\", 3])\n",
    "\n",
    "list_6\n",
    "list_1\n",
    "# How do we just add on that extra list to list_1?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's make a nested list of amino acid information\n",
    "list_aminoacids = [[\"Alanine\", \"Ala\", \"A\"],         \n",
    "                   [\"Arginine\", \"Arg\", \"R\"],\n",
    "                   [\"Asparagine\", \"Asn\", \"N\"],\n",
    "                   [\"Aspartic acid\", \"Asp\", \"D\"]]\n",
    "\n",
    "list_aminoacids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 1.11.1 Use chained indexing `[ ][ ]` to access individual elements of your nested list\n",
    "\n",
    "In order to access specific parts of your `list`, you'll need to remember its structure and use that information. When there is an ordered pattern to your data structure, it can be easier to generate code for traversing the nested list. \n",
    "\n",
    "When using the `[]` index operators, each is evaluated from left to right by retrieving the expected element first before evaluating the next `[]` index call. This step-by-step indexing is also known as chained indexing and can be used to access the elements of a nested list. For a 2D nested list, when use properly, you can think of it like a table or spreadsheet where you access using a `[row][column]` syntax.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Chained indexing deeper structures:</b> can be accomplished by continually using <b>[  ]</b> indexing notation. For every layer or next structure, you can apply the <b>[  ]</b> operator. Just remember that each is evaluated individually from left to right. If you try to access a non-existent level or element, it will return an error!\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the first index is the row number and the second is the column\n",
    "# get sub-list (row) 3, element (col) 1\n",
    "list_aminoacids[...][...]     \n",
    "\n",
    "# get sub-list 1, element 2\n",
    "list_aminoacids[...][...]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can still slice nested lists at the first level.\n",
    "list_aminoacids[...]\n",
    "\n",
    "# What will this retrieve?\n",
    "list_aminoacids[...][...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 1.12.0 Challenge 1: exploring additional list methods\n",
    "\n",
    "You are interested in finding all \"GCC\" codons __AND__ their location in the following object `list_alanine`. Tip: Run `dir(list_alanine)` to see a list of all the attributes and methods that are part of list `list_alanine`.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Caution:</b> <b>dir()</b> unfortunately returns all of the attributes and methods from a class but it does not tell us which items listed are attributes vs. methods. You can, however, explore those more with <b>help()</b> or the <b>getattr()</b> functions.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_alanine = [\"GCA\", \"GCC\", \"GCG\", \"GCU\", \"GCC\"]\n",
    "\n",
    "print(\"Methods available for lists: \" + str(...(list_alanine)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# is count a method or an attribute?\n",
    "...(list_alanine, \"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_alanine\n",
    "\n",
    "# First, a quick check to make sure that \"GCC\" is actually present in alanine\n",
    "\"GCC\" in list_alanine\n",
    "\n",
    "# Count the number of GCC codons with .count()\n",
    "# Note that you can get the output without using the print() command because of our iPython changes\n",
    "print(\"The codon GCC is present \" + str(list_alanine.count(\"GCC\")) + \" time(s) in list_alanine\")\n",
    "\n",
    "# Locate GCC in list_alanine with .index()\n",
    "print(\"The codon GCC can first be found at index \" + str(...(\"GCC\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Caution:</b> <b>.index()</b> Returns only the index of the <b><i>first match</i></b> (hit) it encounters. In the above code, the second instance of GCC was not returned.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.12.1 A brief description of list methods\n",
    "\n",
    "Here is a brief table of the `list` methods with descriptions on what they do:\n",
    "\n",
    "| Method call | Description | Alters the list? |\n",
    "| :- | :- | :-: |\n",
    "| append() |  Add an element to the end of the list | Yes |\n",
    "| extend() |  Add all elements of a list to the another list | Yes |\n",
    "| insert() |  Insert an item at the defined index | Yes |\n",
    "| remove() |  Removes an item from the list | Yes |\n",
    "| pop() |  Removes and returns an element at the given index | Yes |\n",
    "| clear() |  Removes all items from the list | Yes |\n",
    "| sort() |  Sort items in a list in ascending order | Yes |\n",
    "| reverse() |  Reverse the order of items in the list | Yes |\n",
    "| index() |  Returns the index of the _first matched_ item | No |\n",
    "| count() |  Returns the number of times the provided argument occurs in the list  | No |\n",
    "| copy() |  Returns a shallow (new memory) copy of the list | No |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "<b>Section 1.0.0 Comprehension Question:</b> How do you access the last element of the last list in <b>'list_aminoacids'</b>?\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comprehension answer code 1.0.0\n",
    "# Get the last element in the last list of list_aminoacids\n",
    "# hint: it's a D\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 2.0.0 Dictionaries are associative arrays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/Dictionary-Creation-1.jpg?raw=true\"  width=700>|\n",
    "|:--:|\n",
    "|Dictionaries consist of keys that point to values (elements) in an unordered manner. Elements can even be more dictionaries!|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So far, we have seen that lists are very useful to store diverse types of data in a single structure. However, lists are not that convenient to use when we need to extract data in groups or elements that \"belong together\". Let's revisit our nested list `list_aminoacids` from __section 1.11.0__. What if you were interested in getting any data associated with aspartic acid? Do you remember what index **Asp** was at? Under this circumstance, we are better off using dictionaries. \n",
    "\n",
    "Dictionaries are similar to lists and they can also take (almost) any data type. Unlike lists, however, dictionaries are not ordered. Instead, dictionaries map a **key** to a set of **values** that are related to that key, *i.e.* `keys:values` that belong together. In our example, **Aspartic acid** is the key and an associated value is \"Asp\". \n",
    "\n",
    "Dictionaries are created by using `{ }` (curly brackets), a feature that also makes them easily identifiable. Unlike lists, though, ___dictionary keys are immutable___ and have no defined order. As a side note, Python dictionaries are usually called __hashes__ in other programming languages. In summary:\n",
    "\n",
    "- Dictionaries use a `key:value` system.\n",
    "- Two entries __cannot__ share the same key.\n",
    "- __Multiple keys__ can have the same value.\n",
    "- Keys must be immutable data types (string, integer, or tuples, but NOT lists).\n",
    "- Values have few restrictions on data type.\n",
    "- Dictionaries themselves __are__ mutable ie they can be added to, removed from, and changed.\n",
    "\n",
    "For starters, let's create a dictionary called `aminoacids_dict`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a key:value pairing of amino acids\n",
    "aminoacids_dict = {\"Alanine\": \"Ala\",\n",
    "                   \"Arginine\": \"Arg\",\n",
    "                   \"Asparagine\": \"Asn\",\n",
    "                   \"Aspartic acid\": \"Asp\"}\n",
    "\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we created a dictionary, let's access its keys and values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_dict[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the problem? I am 100% certain that alanine is in that list..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember that Python is case sensitive\n",
    "aminoacids_dict[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.0.1 Use `{ }` or the `dict()` function to intialize an empty dictionary\n",
    "\n",
    "Analogously to lists, empty dictionaries can be created with the function `dict()` or just with { }. These can be assigned to variables when needed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "...     # This is an empty dictionary\n",
    "\n",
    "...     # Another way to create an empty dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Caution:</b> Dictionaries do not allow repeated keys. If you enter a <b>key:value</b> pair, only the value will be updated \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.1.0 Adding entries to a dictionary\n",
    "\n",
    "### 2.1.1 Add single entries using a combination of indexing `[ ]` and `=` operators\n",
    "\n",
    "__Single__ entries can be added to dictionaries using the `=` (equal) operator with the syntax `dictionary[key] = value`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Cysteine and Glutamic acid to our dictionary\n",
    "\n",
    "aminoacids_dict[\"Cysteine\"] = ...\n",
    "aminoacids_dict[\"Glutamic acid\"] = ...\n",
    "\n",
    "# View the dictionary after updating it\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.1.2 Add multiple entries with the `.update()` method\n",
    "\n",
    "If you have more than one entry you'd like to add to your dictionary at once, you can essentially create a second dictionary to add through the `.update()` method. Remember that `{key1:value1, key2:value2}` initializes a new dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's add a four more entries at to our dictionary in a single call.\n",
    "\n",
    "# It looks kind of like adding a new dictionary to our old one!\n",
    "aminoacids_dict.update({\"Glutamine\": \"Gln\", \n",
    "                        \"Glycine\": \"Gly\",\n",
    "                        \"Histidine\": \"His\",\n",
    "                        \"Isoleucine\": \"Ile\"})  \n",
    "\n",
    "# View the updated dictionary\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.2.0 Changing the values within a dictionary \n",
    "\n",
    "### 2.2.1 Make a dictionary containing `sets` as values\n",
    "\n",
    "Now, let's rebuild our aminoacids database, this time with their respective symbols and encoding codons stored as `sets`. The main differences between `sets` and `lists` are:\n",
    "\n",
    "- Lists are ordered while sets are not.\n",
    "- Sets cannot contain the same value more than once thus they must all be unique values.\n",
    "- Sets while mutable, cannot contain mutable elements.\n",
    "- Sets use the `set()` or `{}` syntax to initialize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoacids_dict = {\"Alanine\": {\"Ala\", \"A\",  \"GCA GCC GCG GCT\"},\n",
    "                         \"Cysteine\": {\"Cys\", \"C\", \"TGC TGT\"},\n",
    "                         \"Aspartic acid\": {\"Asp\", \"D\", \"GAC GAT\"},\n",
    "                         \"Glutamic acid\": {\"Glu\", \"E\", \"GAA GAG\"},\n",
    "                         \"Phenylalanine\": {\"Phe\", \"F\", \"TTC TTT\"},\n",
    "                         \"Glycine\": {\"Gly\", \"G\", \"GGA GGC GGG GGT\"},\n",
    "                         \"Histidine\": {\"His\", \"H\", \"CAC CAT\"},\n",
    "                         \"Isoleucine\": {\"Ile\", \"I\", \"ATA ATC ATT\"},\n",
    "                         \"Lysine\": {\"Lys\", \"K\", \"AAA AAG\"},\n",
    "                         \"Leucine\": {\"Leu\", \"L\", \"TTA TTG CTA CTC CTG CTT\"},\n",
    "                         \"Methionine\": {\"Met\", \"M\" \"ATG\"},\n",
    "                         \"Asparagine\": {\"Asn\", \"N\", \"AAC AAT\"},\n",
    "                         \"Proline\": {\"Pro\", \"P\", \"CCA CCC CCG CCT\"},\n",
    "                         \"Glutamine\": {\"Gln\", \"Q\", \"CAA CAG\"},\n",
    "                         \"Arginine\": {\"Arg\", \"R\", \"AGA AGG CGA CGC CGG CGT\"},\n",
    "                         \"Serine\": {\"Ser\", \"S\", \"AGC AGT TCA TCC TCG TCT\"},\n",
    "                         \"Threonine\": {\"Thr\", \"T\", \"ACA ACC ACG ACU\"},\n",
    "                         \"Valine\": {\"Val\", \"V\", \"GTA GTC GTG GTT\"},\n",
    "                         \"Tryptophan\": {\"Trp\", \"W\", \"TGG\"},\n",
    "                         \"Tyrosine\": {\"Tyr\", \"Y,\" \"TAC TAT\"}\n",
    "                        }\n",
    "\n",
    "# What is the data type of our new dictionary?\n",
    "type(...)\n",
    "\n",
    "# Print out our dictionary\n",
    "aminoacids_dict                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "As you can see, Python objects can be overwritten so be careful when you create new variables. ___Immutability___, in the case of dictionaries, makes reference to the fact that keys can be added or deleted but not changed.\n",
    "\n",
    "### 2.2.2 Alter key:value pairs with the `.update()` method as well\n",
    "\n",
    "Not only can you add new entries with `.update()` as seen in our previous example, but you can also _alter_ the `key:value` pairs within a dictionary. You will do this similarly to _section 2.1.2_ by supplying a _single_ dictionary object to update all `key:value` pairs.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace the Alanine values in our dictionary with .update()\n",
    "aminoacids_dict.update({\"Alanine\": ...})\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.2.3 Use the `=` operator to also update `key:value` pairs\n",
    "\n",
    "As with `.update()`, we can directly access a `key:value` pair and alter the value using the `dictionary[key]=value` syntax. \n",
    "\n",
    "Let's bring back the Alanine values 'Ala', 'A', and 'GCA GCC GCG GCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Revert the Alanine values in our dictionary using =\n",
    "aminoacids_dict[...] = {'Ala', 'A', 'GCA GCC GCG GCT'}\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.3.0 Remove values from a dictionary with `del()` or `.pop()`\n",
    "\n",
    "Similar to `list` objects, we can remove values from the dictionary directly with the `del()` function or one at a time using the `.pop()` method. These operate much in the same way they do for lists. Note, however, that `.pop()` will not return the final entry by default since dictionaries are not ordered! It will produce an error instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete the \"Cysteine\" entry\n",
    "... aminoacids_dict[\"Cysteine\"]    # del(dictionary_aminoacids[\"Cysteine\"])\n",
    "\n",
    "# pop \"Alanine\" from the dictionary\n",
    "aminoacids_dict...(\"Alanine\")\n",
    "\n",
    "# What does the dictionary look like now?\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.4.0 Additional dictionary functions\n",
    "\n",
    "There are a number of additional `dictionary` methods that can provide information about or alter a dictionary object. For example `len()` gives us the number of `key:value` pairs in a dictionary. \n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>Under the hood:</b> <b>len(s)</b> is a Python core function that can be used on many objects like lists, strings, and dictionaries. Each structure, <b>s</b>, has it's own version of the method <b>__len__()</b> which <b>len()</b> calls on.\n",
    "</div>\n",
    "\n",
    "Let's try out the `len()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many entries are in our dictionary?\n",
    "...(aminoacids_dict)\n",
    "\n",
    "# Equivalent to this awkward code\n",
    "aminoacids_dict.__len__()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.4.1 Use `in` to determine if a key is present in your dictionary\n",
    "\n",
    "We are missing STOP and START codons in our dictionary. We can use our knowledge of **Python** to demonstrate that STOP is not present in `dictionary_aminoacids`. Using the `in` keyword to check for inclusion will produce a boolean answer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just like checking a list, we can check our set of keys for a specific one\n",
    "\"Tryptophan\" ... aminoacids_dict\n",
    "\n",
    "# What returns if a key isn't in our dictionary? Remember we recently used \".pop()\" on alanine\n",
    "\"Alanine\" ... aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.4.2 Use `.get()` to retrieve all of the values associated with a key\n",
    "\n",
    "Use `.get()` to retrieve all values associated with glutamic acid. Tip: Use `help()` to find out more about the usage of `.get()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look up what the 'get' method does\n",
    "help(dict.get)\n",
    "\n",
    "# Use the get function to look up \"Glutamic acid\"\n",
    "aminoacids_dict...(\"Glutamic acid\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.4.3 Use the `.values()` method to extract _all_ values from the dictionary\n",
    "\n",
    "If you need to extract only the values from a dictionary, use the method `.values()`. This will return a `dict_values` object which can also be cast as a list object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the values from our dictionary AND cast it to a list at the same time\n",
    "list_values_only = list(aminoacids_dict...)\n",
    "\n",
    "list_values_only\n",
    "\n",
    "list_values_only[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.5.0 Use a `list` _element_ as a dictionary key\n",
    "\n",
    "All of our examples so far have used a few different types of objects that can be used as values. To remind you, the keys of a dictionary are made of immutable types such as integers, strings, booleans, or tuples (see __Appendix 2__), but not lists or dictionaries - remember these are mutable. However, if you have a `list` element, that is not itself a `list` but is a simple immutable data type, then you can use that.\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Read more:</b> You can find more details about dictionary implementation <a href=\"https://wiki.python.org/moin/DictionaryKeys\">here</a>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add keys and values using update() or squared-bracket notation\n",
    "dictionary_empty = {}    #empty\n",
    "\n",
    "# Use a string for a key\n",
    "dictionary_empty.update({\"string\" : 1})\n",
    "\n",
    "# Use an integer for a key\n",
    "dictionary_empty[...] = \"integer\"\n",
    "\n",
    "# Use a boolean for a key\n",
    "dictionary_empty.update({True: 3})\n",
    "\n",
    "dictionary_empty"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "However, elements of a `list` can be passed on as keys to a dictionary ___provided that they too are immutable elements___."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_1\n",
    "\n",
    "# Use a list element as a key\n",
    "dictionary_test = {...: 2.5}\n",
    "dictionary_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.6.0 Dictionaries can also have nested structures\n",
    "\n",
    "Just like lists, you could use dictionaries as values within your dictionaries, thus generating nested dictionaries much like in our diagram at the start of this section. Like nested lists, you can chain index hash keys at each level with progressive use of the `[key]` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a dictionary where values are also dictionaries!\n",
    "aminoacids_subset_dict =  {'Alanine': {'One letter symbol':'A', \n",
    "                                       'Three letter symbol':'Ala', \n",
    "                                       'codons':'GCA GCC GCG GCU', \n",
    "                                       'Number of codons':4},\n",
    "                           'Aspartic acid': {'One letter symbol':'D', \n",
    "                                             'Three letter symbol':'Asp', \n",
    "                                             'codons':'GAC GAU', \n",
    "                                             'Number of codons':2},\n",
    "                           'Glutamic acid': {'One letter symbol':'E', \n",
    "                                             'Three letter symbol':'Glu', \n",
    "                                             'codons':'GAA GAG', \n",
    "                                             'Number of codons':2}\n",
    "                          }\n",
    "\n",
    "# check it's type\n",
    "type(aminoacids_subset_dict)\n",
    "type(aminoacids_subset_dict[\"Alanine\"])\n",
    "\n",
    "# Pull out a few specific sub-entries\n",
    "aminoacids_subset_dict[\"Alanine\"][\"Number of codons\"]\n",
    "aminoacids_subset_dict[\"Glutamic acid\"][\"codons\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Dictionaries cannot be sliced with index `[ ]` notation:</b> Unlike lists which still have an ordered nature, the dictionary keys <b>do not</b> have an ordered index. Therefore you cannot access specific elements by their \"position\" within the dictionary. Subsequently, since the elements have no order, you cannot use the slice notation <b>[start:end]</b> with dictionaries either.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.7.0 Use the `.items()` method in dictionaries to make tuples\n",
    "\n",
    "Sometimes you may want to preserve or move around a `key:value` pair. You can do so using the `.items()` method which will return a `dict_items` object. From there, you can use `tuple()` to cast it over to a tuple object either individually, or as a tuple of tuples. \n",
    "\n",
    "Note that while dictionaries are not indexed, they do remember their insertion order! That means if you wanted to walk your way through all of the `key:value` pairs, you could do so using this method!\n",
    "\n",
    "In the case of `aminoacids_dict`, each item in the dictionary will take the format of `('key_name', {set_information})`.\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>What is a tuple?</b> A tuple is essentially a list in many ways except it is immutable and thus cannot grow in size. These data structures, while simpler than lists, are computationally faster. See <b>Appendix section 7.0.0</b> for more about the tuple data structure.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab the first inserted key:value pair in the dictionary\n",
    "tuple(aminoacids_dict.items())[...]\n",
    "\n",
    "# grab the last inserted key:value pair in the dictionary\n",
    "tuple(aminoacids_dict.items())[...]\n",
    "\n",
    "# grab the fifth inserted key:value pair in the dictionary\n",
    "tuple(aminoacids_dict.items())[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 2.8.0 A summary of dictionaries\n",
    "\n",
    "To summarize, list and dictionaries are very similar but have three key differences: Indexation, order, and mutability. If you need fast indexation to look for unique keys and their values, go with dictionaries.\n",
    "\n",
    "Run `dir(dict)` for more methods on dictionaries.\n",
    "\n",
    "Here is a more detailed description of the dictionary methods and what they do:\n",
    "\n",
    "| Method call | Description | Alters the dictionary? |\n",
    "| :- | :- | :-: |\n",
    "| .update() | Add multiple entries to a dictionary | Yes |\n",
    "| .pop() | Remove values from the dictionary | Yes |\n",
    "| .get() | Retrieve values associated with a dictionary key | No |\n",
    "| .keys()| Retrieve all of the keys associated with a dictionary | No |\n",
    "| .values() | Retrieve all values associated with a dictionary | No |\n",
    "| .setdefault() | Returns a value for a key if it exists, otherwise creates the key and sets a value | Yes|\n",
    "| .items() | Returns key-value pairs of a dictionary as a dict_items object| No |\n",
    "| del() | Removes an item from the list | Yes |\n",
    "| in | Check if a key is present in a dictionary | No |\n",
    "| len() | Check the length of a dictionary | No |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Section 2.0.0 Comprehension Question:</b> From our dictionary <b>aminoacids_subset_dict</b>, how do we query what the keys are for the Aspartic acid dictionary? How do we directly access the three letter symbol for Aspartic acid? Hint: look closely at the functions listed in the table above.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comprehension answer code 2.0.0\n",
    "# Find the keys for the Aspartic acid dictionary\n",
    "...\n",
    "\n",
    "# Find the 3-letter symbol for Aspartic acid\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 3.0.0 Introduction to `NumPy`\n",
    "\n",
    "Built-in Python objects are very versatile. They can hold different data types, you can do some basic mathematical operations with them, they can be mutable or not, and many more. However, they lack a key feature: performing simple and advanced **mathematical operations on data structures**, in ways that are time- and computationally-efficient.\n",
    "\n",
    "In a science settings there are some operations that we would like to do with data that cannot be accomplished with only a `list`:\n",
    "\n",
    "- add a constant value to a list of numbers or\n",
    "- add values in an element-wise manner from size-matched lists\n",
    "- access subsets of values from a list in a non-contiguous manner (eg the 2nd, 4th, and 7th elements)\n",
    "\n",
    "Enter `NumPy` (Numeric Python), a package for scientific computing developed by Travis Oliphant and first released in 2005, based on a pre-existing python package called `Numeric`. NumPy has advanced and changed a lot since its first release. This is all thanks to a very active community of programmers that have contributed their time and effort to improve NumPy.\n",
    "\n",
    "In terms of data structures, NumPy offers an alternative to built-in `lists` called NumPy `arrays`, which allow math operations across one- or multi-dimensional arrays. Let's start diving into NumPy's functionality. First, however, we need to install NumPy and `import` it as `np` (not required but it is the standard).\n",
    "\n",
    "## 3.1.0 One-dimensional NumPy arrays using `numpy.array()`\n",
    "\n",
    "A One-dimensional (1D) `array` object has all data stored in an ordered fashion as a \"single column\". Sounds like a list right? These structures, however, ___can only contain a single data type___ and will perform coercion _without any warnings_. Let's create our first NumPy arrays using the `numpy.array()` function. Note that the object type produced is the `numpy.ndarray`, short for n-dimensional array.\n",
    "\n",
    "Recall that we will import with alias `np` so our command will take the form of `np.array([element_1, element_2, ..., element_n])`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip3 install numpy # Always keep installation commands commented out to prevent unwanted installations\n",
    "import numpy as np # Already done in section 0.5.0\n",
    "\n",
    "# All integers\n",
    "np.array([5, 16, 23])\n",
    "\n",
    "# Coerce to float\n",
    "np.array([1.2, ..., 12.12])\n",
    "\n",
    "# Coerce to Unicode string\n",
    "np.array([5, 1.2, ...])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.1.1 Use the `.shape` _property_ to retrieve array dimensions\n",
    "\n",
    "If you're not sure how many elements, columns, rows etc., you have in your array, you can quickly retrieve it using the `.shape` property. Notice the lack of parentheses here? We'll talk more about what a property is in section __4.2.0__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_1 = np.array([5, 16, 23])\n",
    "\n",
    "# What is the shape of this 1D array?\n",
    "array_1...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(3,) means 3 rows and no columns (remember this is a 1D array)\n",
    "\n",
    "***\n",
    "## 3.2.0 Mathematical operations on arrays are performed element-wise\n",
    "\n",
    "Remember that using operators like __`+`__ and __`*`__ on lists end up producing behaviours related to the concatenation of the list objects themselves, with arrays we can perform _element-wise_ math operations __if__:\n",
    "1. the arrays are the same size and\n",
    "2. the elements are suitable for math operations.\n",
    "\n",
    "Note that element-wise in this case refers to performing operations on each element OR on index-matched elements, depending on the context."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Multiplication of an array to a scalar (single) value\n",
    "np.array([5, 16, 23]) ...\n",
    "\n",
    "# Element-wise multiplication (or division) of two size-matched arrays\n",
    "np.array([5, 16, 23]) * np.array([3, 19, 15])\n",
    "\n",
    "np.array([5, 16, 23]) / np.array([...])    # equivalent to: np.array([5, 16, 23]) / 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Numpy does not automatically recycle arrays:</b> If you are familiar with the programming language <b>R</b> you may be used to the idea of vector recycling. In certain math operations when vectors (similar to np.arrays) are of different sizes, the smaller one is automatically \"recycled\" to complete the operation. This is not the case with the numpy package but a <i>similar</i> kind of idea underlies the <b>numpy.resize()</b> function.\n",
    "</div>\n",
    "\n",
    "If we add up two NumPy arrays, their elements will be added up element-wise. The same rules apply as above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember what + does for a list?\n",
    "list([5, 16, 23]) + ...([3, 19, 15])\n",
    "\n",
    "# Compare to an array\n",
    "np.array([5, 16, 23]) + ...([3, 19, 15])\n",
    "\n",
    "# Add the same value to all elements. We call this broadcasting!\n",
    "np.array([5, 16, 23]) + np.array([3])\n",
    "\n",
    "# In this case order does not matter\n",
    "np.array([3]) + np.array([5, 16, 23])\n",
    "\n",
    "# This is equivalent to above!\n",
    "np.array([5, 16, 23]) + 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can subtract too!\n",
    "np.array([5, 16, 23]) - np.array([13])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What happens if the arrays are not matched in size for addition?\n",
    "np.array([5, 16, 23]) + np.array([3, 19, 15, 16])\n",
    "\n",
    "# You'll get a similar problem with the other math operators"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2.1 Addition of a `list` to a NumPy `array` coerces the `list`\n",
    "\n",
    "Adding a list and a NumPy array results in the list behaving as an array (coercion) to complete the operation. Remember that the addition has to make sense to the interpreter too. The list is simply coerced to an array object. Will its contents be coerced as needed?\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>Caution:</b> NumPy <b>arrays</b> may seem a lot like <b>lists</b> but don't expect core python operators to behave as dynamically with them as we see with base core objects like numbers vs strings vs lists!\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Coercion to arrays happens regardless of addition order\n",
    "list([5, 16, 23]) + ...([3, 19, 15])\n",
    "\n",
    "np.array([3, 19, 15]) + ...([5, 16, 23])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Will this work? \n",
    "np.array([3, 19, 15]) + list([\"5\", \"16\", \"23\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can we concatenate string arrays like this?\n",
    "np.array([\"this\", \"that\", \"those\"]) + np.array([\"who\", \"what\", \"where\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.3.0 NumPy `array` is a data structure\n",
    "\n",
    "Just like lists, tuples, and dictionaries, the NumPy `array` is a data structure that has its own properties, and _methods_. As you've seen, we initialize arrays with the `np.array()` function. Similar to lists, we can use `[]` index operations to subset and slice these with the expected behaviour."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a numeric array\n",
    "array_1 = np.array([3, 15, 21])\n",
    "\n",
    "# Index the second element\n",
    "array_1[...]\n",
    "\n",
    "# Slice from index 0 (inclusive) to 2 (exclusive)\n",
    "array_1[...]\n",
    "\n",
    "# Use negative notation to index from the end\n",
    "array_1[...]\n",
    "\n",
    "# Slice with a negative index\n",
    "array_1[...]\n",
    "\n",
    "# Slice implicitly from the beginning of the array\n",
    "array_1[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.4.0 Use conditional operators on Numpy `array`s \n",
    "\n",
    "Up until now we haven't really used conditional operators on our data structures but we can quickly ask certain objects about their elements and if they fulfill a condition with a boolean result (aka `True` or `False`). With NumPy arrays, we can perform these conditional statements in an element-wise manner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which elements in array_1 are larger than 17?\n",
    "array_1 ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.1 Index arrays with conditional statements\n",
    "\n",
    "Instead of just looking at an array of booleans, which can get quite long in large data sets, you can instead feed your conditional result back into the original array to _retrieve the values_ that return true. This effectively gives you an array of actual element values that you might \"want\" to work with. This is considered a form of __conditional indexing__ but may also be known as value filtering. We achieve this with the syntax `array[conditional statement]` in the above case `array_1 > 17` is our conditional statement. Let's try!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return the values greater than 17\n",
    "array_1[array_1 ...]\n",
    "\n",
    "array_1[array_1 ...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.4.2 Negate the values of your conditional array with `~`\n",
    "\n",
    "In Python there are a number of ways to negate booleans (create the opposite value). With the NumPy package and arrays in particular we can use the __`~`__ operator to perform a bitwise negation on the array. Like with all objects, however, if used outside this context the behaviour of the operator may not be as expected!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Practice our negation on a toy array\n",
    "... np.array([True, False, False, True, True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return the values <= 17\n",
    "array_1[... (array_1 > 17)]\n",
    "\n",
    "# Placement of the ~ is important and must be used directly on a boolean array!\n",
    "... array_1[(array_1 > 17)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.5.0 Two-dimensional arrays\n",
    "\n",
    "Put simply, a 2D NumPy array has rows and columns (hence the 2D name). Similar to their 1D counterparts, all data in a 2D array must be of the same type. We can make 2D arrays by combining two individual lists of the same size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The double squared bracket is required, otherwise you get a \"type not understood\" traceback\n",
    "\n",
    "array_2 = np.array([\n",
    "                    [5, 16, 23],   # The line break enhances readability but is not required\n",
    "                    [3, 19, 15]\n",
    "                   ])   \n",
    "\n",
    "array_2\n",
    "# What is the shape of our array?\n",
    "array_2.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Our 2D array is 2 rows by 3 columns. Although our original arrays were thought of as columns of values, when we put them into 2D arrays each \"column\" becomes an entire row! \n",
    "\n",
    "You can also type-convert a nested list into a 2D array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a short 2D array of amino acid information\n",
    "aminoacids_list = [[\"Alanine\", \"Ala\", \"A\"],         \n",
    "                   [\"Arginine\", \"Arg\", \"R\"],\n",
    "                   [\"Asparagine\", \"Asn\", \"N\"],\n",
    "                   [\"Aspartic acid\", \"Asp\", \"D\"]]\n",
    "\n",
    "# Assign our list to an array\n",
    "aminoacids_array = np.array(...)\n",
    "\n",
    "# Let's take a look at it\n",
    "aminoacids_array\n",
    "\n",
    "# What are its dimensions?\n",
    "aminoacids_array.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.5.1 Subset 2D NumPy arrays with `[row][column]` index notation\n",
    "\n",
    "To subset 2D arrays we use two sets of squared brackets: one for rows and one for columns which looks like `my_array[row][column]`. This behaves much in the same we would treat a nested list object to perform chain indexing! If we consider 2D arrays as an array of nested arrays, then we are again chain indexing where the first `[ ]` pulls a \"row\" which is a nested `array` element. Then our second index `[ ]` pulls elements from the `array` itself. Don't forget, that like all other Python objects, we've seen that arrays are also __zero-indexed__!\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b> Caution: </b> Subsetting arrays with only one index defaults to rows. E.g., <b>my_array[0]</b> returns row 0.\n",
    "</div>\n",
    "\n",
    "Let's select the element \"15\" from array_2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What does array_2 look like?\n",
    "array_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the \"15\" in the last row, last column\n",
    "array_2[...][...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.5.2 Subset 2D NumPy arrays with `[row, column]` single-index annotation\n",
    "\n",
    "The same result as above can be achieved using one set of squared brackets with a comma that separates rows from columns which looks like `my_array[row, column]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset from row 1, column 2 of array_2\n",
    "array_2[...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.5.3 Use slice notation `[rowStart:rowEnd, colStart:colEnd]` to select subsets of your array\n",
    "\n",
    "That's right! Slicing has been implemented for arrays so you can pull out proper subsets unlike nested lists! For example, to select all rows or columns, use a `:` (colon) at the respective side of the comma. Let's practice some slicing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Slice from all rows, 2nd column\n",
    "array_2[...]\n",
    "\n",
    "# Slice from second row, first two columns\n",
    "array_2[1, 0:2]\n",
    "\n",
    "# Slice from first two rows, first two columns\n",
    "array_2[0:2, 0:2]\n",
    "\n",
    "# This copies the array's reference from memory\n",
    "array_2[...]     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.6.0 Using NumPy for statistical applications\n",
    "\n",
    "So far we have been doing very basic operations to experiment with Python objects. For that reason, we have been using small datasets. In reality, we get big datasets that need to be analyzed in ways that are efficient and reliable. \n",
    "\n",
    "In order to explore NumPy's statistical capabilities, we are going to simulate a more complex, very popular dataset in the data science world called **iris**. We will showcase some exploratory data analysis capabilities but we encourage you to look into R to do advanced stats and data visualization (CAGEF also has an introductory R course!).\n",
    "\n",
    "Let's create the iris dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should we set a seed for our random function? Why or why not?\n",
    "np.random.seed(2025)\n",
    "\n",
    "# loc= mean of the distribution, scale=standard deviation, size= how many elements\n",
    "\n",
    "# Generate a few different sets of numbers\n",
    "# np.round() will return a np.array of values rounded to 2 decimal places\n",
    "sepal_length = np.round(np.random.normal(loc = 5, scale=0.2, size=14), decimals=2)  \n",
    "sepal_width = np.round(np.random.normal(loc = 3, scale=0.5, size=14), decimals=2)\n",
    "petal_length = np.round(np.random.normal(loc = 1.3, scale=0.2, size=14), decimals=2)\n",
    "petal_width = np.round(np.random.normal(loc = 0.15, scale=0.2, size=14), decimals=2)\n",
    "\n",
    "# Combine your 4 arrays into a single 2D array\n",
    "array_iris = np.column_stack([sepal_length,\n",
    "                             sepal_width,\n",
    "                             petal_length,\n",
    "                             petal_width\n",
    "                             ])\n",
    "\n",
    "# What is the results of using np.column_stack?\n",
    "type(array_iris)\n",
    "\n",
    "# What kind of data does the array carry?\n",
    "array_iris.dtype\n",
    "\n",
    "# What is the shape of our 2D array?\n",
    "array_iris.shape\n",
    "\n",
    "# What does the array look like?\n",
    "array_iris"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.1 Use the `mean()` function on arrays\n",
    "\n",
    "Remember that both arrays and many of the statistical functions we'll be introducing are part of the NumPy package. That being said, there are implicit expectations regarding behaviours and attributes present in objects like arrays. Without getting too much into the philosophy, it can make the introduction of additional functions to a package both more flexible and rigid.\n",
    "\n",
    "Let's take a closer look at the `mean()` function. Looking at the documentation we see the following information\n",
    "\n",
    "`numpy.mean(a, axis=None, dtype=None, out=None, keepdims=<no value>, *, where=<no value>)` \n",
    "\n",
    "which we can break down to:\n",
    "\n",
    "- `a` an array containing numbers - or something that can be converted to such.\n",
    "- `axis` defaults to \"None\" and take the mean of the flattened array but can also be set to 0, 1 or n (multi-dimensional arrays). \n",
    "- `where` is optional but can be used to include a conditional array to determine _which_ elements to use to make the calculation.\n",
    "\n",
    "Let's start with the basic use of `mean()` and build up from there. First, we can calculate the average for sepal length (column 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate mean across the entire array\n",
    "np.mean(array_iris)\n",
    "\n",
    "# Calculate the mean of our array but only the 1st column. Is that index 1?\n",
    "np.mean(array_iris[...])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the mean of row 2\n",
    "np.mean(array_iris[...])\n",
    "\n",
    "# Equivalent notation\n",
    "np.mean(array_iris[1 , : ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.1.1 Using the `axis` parameter to calculate the mean across dimensions\n",
    "\n",
    "As you can see from above, when we use the default behaviour of mean, it treats our array, regardless of dimension, like a flat list of numbers and calculates the mean of the whole set. What if, we want to calculate mean across rows or columns? Intuitively it can be a little confusing but recall that we use `[row, column]` notation to access arrays. If we are working with an array with __n__ rows and __m__ columns `[n,m]` then:\n",
    "\n",
    "- `axis = 0` will return an array of 1 x m length which will calculate the mean of each column and return an array of __m__ values.\n",
    "- `axis = 1` will return an array of n x 1 which will calculate the mean of each row and return an array of __n__ of values.\n",
    "\n",
    "Another way to think about it is that `axis=0` returns a _row_ of means, and `axis=1` returns a _column_ of means. \n",
    "\n",
    "Note that we aren't even talking about multi-dimensional arrays where `axis` can also be assigned as a ___tuple___ to identify different dimensions you'd like to perform the calculation across. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the mean of each column (of which there are 4!)\n",
    "np.mean(array_iris, axis = ...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the mean across each row (of which there are 14)\n",
    "np.mean(array_iris, axis = ...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.2 Other functions available in NumPy\n",
    "\n",
    "We've been looking at just the `mean()` function but we can achieve different calculations with similar behaviours from \n",
    "- `median()`\n",
    "- `corrcoef()`\n",
    "- `std()`\n",
    "- `sum()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the median of column 2\n",
    "np.median(array_iris[...])\n",
    "\n",
    "# Calculate the Pearson correlation coefficient between sepal length (column 1) and sepal width (column 2)\n",
    "np.corrcoef(array_iris[ : , 0], ...) \n",
    "\n",
    "# What is the standard deviation of column 1?\n",
    "np.std(array_iris[ : , 0])\n",
    "\n",
    "# What is the sum of column 4?\n",
    "np.sum(array_iris[ : , 3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.3 Perform linear algebra and `tranpose()` with arrays in NumPy\n",
    "\n",
    "Sometimes you'd like to manipulate the shape or order of elements in your arrays. A common method that you may wish to perform is the `transpose()` method. This is a method available through the array object and it will take care of all of the details even for multi-dimensional arrays. It's actually part of the array manipulation set of routines.\n",
    "\n",
    "For now we'll stick with the straightforward 2-dimensional array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look at the array again\n",
    "array_iris\n",
    "\n",
    "array_iris.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transpose the array\n",
    "array_iris.transpose() \n",
    "\n",
    "array_iris.transpose().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.4 Calculate the inverse of a matrix\n",
    "\n",
    "You may remember from first- or second-year algebra the following formula\n",
    "\n",
    "$$ A x A^{-1} = I $$\n",
    "\n",
    "Where $A$ is a ___square matrix___ we can find it's inverse $A^{-1}$ such that the multiplication of them produces the identity matrix which has all 0s except for a diagonal line of 1s.\n",
    "\n",
    "To calculate the inverse matrix, we can use the function `inv()` from the `linalg` module of NumPy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix up our call to inv() by making a square matrix\n",
    "np.linalg.inv([[5, 16, 23],\n",
    "              [3, 19, 15],\n",
    "              [17, 2, 5]\n",
    "              ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.6.5 Calculate the dot-product with the `dot()` function\n",
    "\n",
    "The dot-product of two matrices calculates the sum of the product of elements between rows of matrix A and columns of matrix B. Sound familiar?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/dot-product.jpg?raw=true\"  width=700>|\n",
    "|:-|\n",
    "|Source: https://algebra1course.wordpress.com/2013/02/19/3-matrix-operations-dot-products-and-inverses/|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simple case of dot-product\n",
    "\n",
    "np.dot(np.array([[5, 16, 23]]),\n",
    "       np.array([[1], [2], [3]])) # Notice how we are making a \"vertical\" array?\n",
    "                                  # np.array([1,2,3]) will also make a vertical array!\n",
    "\n",
    "# It can all be done in just one line but can be harder to read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_1\n",
    "\n",
    "# Calculate the dot-product a 3x3 matrix to a 1x3\n",
    "np.dot(np.array([[ 5, 16, 23], [ 3, 19, 15], [17, 2, 5]]),   \n",
    "       array_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 3.7.0 A summary of the Numpy `ndarray`\n",
    "\n",
    "To summarize, the Numpy array shares some similarities to a list (indexed and mutable) but it can only be made up of one object type.\n",
    "\n",
    "Run `dir(np.ndarray)` for more functions that can be used on ndarrays.\n",
    "\n",
    "Here is a quick list functions and what they do:\n",
    "\n",
    "| Method call | Description | Alters the array? |\n",
    "| :- | :- | :-: |\n",
    "| np.sum() | Calculate array sums across a specified axis | No |\n",
    "| np.cumsum() | Calculate array of cumulative sums across a specified axis | No |\n",
    "| np.prod() | Calculate array products across a specified axis | No  |\n",
    "| np.cumprod()| Calculate array of cumulative products across a specified axis | No  |\n",
    "| np.all() | Return TRUE if all elements evaluate to TRUE | No |\n",
    "| np.any() | Return TRUE if any elements evaluate to TRUE | No|\n",
    "| np.min() | Return the minimum value along a given axis | No |\n",
    "| np.max() | Return the maximum value along a given axis | No |\n",
    "| .copy() | Return a copy of the array | No |\n",
    "| .reshape() | Returns an array containing the same data with a new shape | No |\n",
    "| .resize() | Change the shape and size of an array in place | Yes |\n",
    "\n",
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>There's more to explore!:</b> Check out more documentation on <a href=\"https://numpy.org/doc/stable/reference/arrays.ndarray.html\">NumPy arrays here</a>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Section 3.0.0 Comprehension Question:</b> How do we calculate the cumulative sum of <b>array_iris</b> across each row of the array?\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "help(np.cumsum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comprehension answer code 3.0.0\n",
    "# Take a quick look at array_iris\n",
    "...\n",
    "\n",
    "# Calculate the cumulative sum for each row\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 4.0.0 Pandas gives us the `DataFrame` object\n",
    "\n",
    "Up until now we've been seeing glimpses of how we could work with and manipulate tabular data - the kind of data we might see from our experiments. A 2D nested `list` structure is problematic to navigate and isn't receptive to fast math operations. A `dictionary` can feel close in structure but its display and navigation is not ideal. A 2D `numpy.array` begins to look exactly exactly how we might imagine an excel spreadsheet of data. We can also update and manipulate the values in an element-wise manner but are limited to containing __single data types__. \n",
    "\n",
    "The next step in our journey to working with real-world data brings us to the `Pandas` package. To put it in context, the __`Pandas`__ package expands __`NumPy`__ capabilities in the same way that NumPy expands Python's. Pandas is a data manipulation tool developed by Wes McKinney, built on NumPy to simplify working with _tabular_ datasets. \n",
    "\n",
    "We'll cover this in more detail next week, but in properly formatted tabular datasets, each column is a __variable__ (a parameter that was measured) and each row is a set of __observations__ (the results of quantitatively or qualitatively measuring each parameter). \n",
    "\n",
    "There are two data structures in Pandas that we are interested in: \n",
    "\n",
    "| Structure | Description | Characteristics |\n",
    "| :- | :- | :- |\n",
    "|`Series`| A 1-dimensional array-like structure. | Contains a single data type |\n",
    "||| Values are mutable but size is not |\n",
    "|`DataFrame`| A 2D labeled, tabular container for `Series` objects | Resembles a spreadsheet\n",
    "||| Size-mutable by adding columns|\n",
    "\n",
    "<br> \n",
    "Based on the above description we should note there are a couple of rules regarding the structure of `DataFrames`: \n",
    "\n",
    "1. While we are allowed to mix different data types now (ie `integers` vs `strings` etc.), each column is considered a `Series` object meaning that data in that column ___must be of the same type___.\n",
    "\n",
    "2. Each `Series` or column should be the same length. Trying to add or use mismatched dataset sizes will result in a Python error.\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/structureHierarchy.png?raw=true\"  width=700>|\n",
    "|:--:|\n",
    "|A conceptual hierarchy of data objects and their relationship to the pandas.DataFrame|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is some information about Pandas library architecture:\n",
    "\n",
    "- `pandas.core`: data structures about Pandas library\n",
    "- `pandas.src`: Holds basic functionality of Pandas\n",
    "- `pandas.io`: Tools to input and output files, data, etc.\n",
    "- `pandas.tools`: Functional operations in Pandas\n",
    "\n",
    "Other architectures include `sparse` (missing values), `stats` (statistical applications), `util` (testing and debugging tools), and `rpy` (R2Py, connectivity with R programming language). \n",
    "\n",
    "First, install pandas using pip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install pandas\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 4.1.0 Create a `DataFrame` from a `dictionary` object\n",
    "\n",
    "We can start our journey into `Pandas` by creating data frames out of dictionaries! Next week, we'll learn how to import them from tabular data like `.csv` and `.xlsx` files. \n",
    "\n",
    "Recall that dictionaries use a `key:value` structure. We can use `keys` as variables (column names) and `values` as observations (rows). \n",
    "\n",
    "First, let's build a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a dictionary of amino acids with 4 keys: Amino acid, single letter symbol, DNA codons, and Compressed\n",
    "aminoacids_dict = {\n",
    "    \"Aminoacid\": [\"Alanine\",\"Arginine\",\"Asparagine\",\"Aspartic acid\",\"Cysteine\",\"Glutamine\",\n",
    "                  \"Glutamic acid\",\"Glycine\",\"Histidine\",\"Isoleucine\",\"Leucine\",\"Lysine\",\n",
    "                  \"Methionine\",\"Phenylalanine\",\"Proline\",\"Pyrrolysine\",\"Serine\",\"Threonine\",\n",
    "                  \"Tryptophan\",\"Tyrosine\",\"Valine\",\"START\",\"STOP\"],  \n",
    "    \"single letter symbol\": [\"A\",\"R\",\"N\",\"D\",\"C\",\"Q\",\"E\",\"G\",\"H\",\"I\",\"L\",\"K\",\"M\",\n",
    "                             \"F\",\"P\",\"O\",\"S\",\"T\",\"W\",\"Y\",\"V\",\"START\",\"STOP\"],\n",
    "    \"DNA codons\": [\"GCT GCC GCA GCG\",\"CGT CGC CGA CGG AGA AGG\",\"AAT AAC\",\"GAT GAC\",\"TGT TGC\",\"CAA CAG CAR\",\n",
    "                   \"GAA GAG\",\"GGT GGC GGA GGG\",\"CAT CAC\",\"ATT ATC ATA\",\"TTA TTG CTT CTC CTA CTG\",\"AAA AAG\",\n",
    "                   \"ATG\",\"TTT TTC\",\"CCT CCC CCA CCG\",\"TAG\",\"TCT TCC TCA TCG AGT AGC\",\"ACT ACC ACA ACG ACN\",\n",
    "                   \"TGG\",\"TAT TAC\",\"GTT GTC GTA GTG\",\"ATG\",\"TAA TGA TAG\"],\n",
    "    \"Compressed\": [\"GCN\",\"CGN AGR MGN CGY\",\"AAY\",\"GAY\",\"TGY\",\"CAR\",\"GAR\",\"GGN\",\"CAY\",\"ATH\",\"YTR CTY CTN TTR\",\n",
    "                   \"AAR\",\"ATG\",\"TTY\",\"CCN\",\"TAG\",\"TCN AGY\",\"ACN\",\"TGG\",\"TAY\",\"GTN\", \"ATG\", \"TRA TAG\"]\n",
    "}\n",
    "\n",
    "aminoacids_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.1.1 Use the `DataFrame()` function\n",
    "\n",
    "Hard to read, right? It will look much better to human eyes if we convert it into a `DataFrame` object. We'll use the function `DataFrame()` to accomplish our goal and we'll also introduce a way to take a quick look at your data with the `.head()` method. This will allow us to look at a specified number of rows from the beginning of our DataFrame. In this case, the default number of rows is 5.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "    <b>Take Note!</b> We'll be using the <b>variable_df</b> suffix notation in variable names to identify them as DataFrame objects.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert our dictionary to DataFrame\n",
    "aminoacids_df = ...\n",
    "\n",
    "# Take a peek at the first 5 rows of our DataFrame with the head() method\n",
    "aminoacids_df...\n",
    "\n",
    "# We'll just insert an extra line here for ease of readabilty\n",
    "print(\"\\nWe just made an object of type:\", type(aminoacids_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice how Jupyter has formatted the DataFrame for us into a nice readable table? Convenient!\n",
    "\n",
    "***\n",
    "\n",
    "## 4.2.0 Properties of objects\n",
    "\n",
    "Recall that objects can be composed of attributes (values) and methods. Up until now we have been collecting information or working with objects through their methods. Depending on _who_ has implemented the code for your object and the language you are using, most attributes are (by good practices) private. That means that under the hood, you can't simply alter these attributes directly. Instead, you'll call on **helper methods** to accomplish this. Python has developed a specific implementation whereby _if_ you are allowed to access/alter an attribute it will likely be a `property` of the object. Defining attributes this way, you can define specific ways to `get` and `set` information about them (with failsafes!). \n",
    "\n",
    "More importantly, all of the details are hidden from the user and we can simply access these properties with the `<object>.<property>` syntax. Note the use of the `.` (dot-notation) here that comes between our object and the property! While it may look like we are directly accessing an attribute, we are not.\n",
    "\n",
    "### 4.2.1 Revisit the `.shape` property to retrieve dimension information\n",
    "\n",
    "We've already used this property a number of times on our `np.arrays` but let's try it on our `pd.DataFrame` now. How many rows and columns do we have? That question can be answered by retrieving the `.shape` __property__ which will return a pair of values in the format of `(row, column)`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the shape of our dataframe\n",
    "aminoacids_df..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.2.2 Use the `.index` property to retrieve/add/change row names in your `DataFrame`\n",
    "\n",
    "From above you may notice a single unnamed column in our DataFrame. It looks like it is using the potential indices of our rows as labels. These are the ___row names___ of our data frame and they can be altered to carrying meaningful information like amino acid abbreviations instead. \n",
    "\n",
    "To alter our `DataFrame` index, we can access the property `.index` and assign it directly to a new list. Now, let's add indices (row names) to our data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the indices of our data frame\n",
    "aminoacids_df...\n",
    "\n",
    "# Alter the indices of our data frame\n",
    "aminoacids_df.index = [\"Ala\",\"Arg\",\"Asn\",\"Asp\",\"Cys\",\"Gln\",\"Glu\",\"Gly\",\"His\",\"Ile\",\n",
    "                       \"Leu\",\"Lys\",\"Met\",\"Phe\",\"Pro\",\"Pyl\",\"Ser\",\"Thr\",\"Trp\",\"Tyr\",\n",
    "                       \"Val\",\"START\",\"STOP\"]\n",
    "\n",
    "aminoacids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What will the index look like now?\n",
    "aminoacids_df.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "As you can see from the above coding cells we can both ___set___ and ___get___ data using the same property!\n",
    "\n",
    "### 4.2.3 Use the `.columns` property to retrieve/change column names in your `DataFrame`\n",
    "\n",
    "Much like the `.index` property, we can use the `.columns` property to get and set our column names. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the column information for our data frame\n",
    "aminoacids_df..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 4.3.0 Other helpful `DataFrame` methods\n",
    "\n",
    "So far we've seen some helpful properties for retrieving and altering attributes within our `DataFrame` objects. We've also seen the use of `.head()` to look at the first __n__ rows of our DataFrames. Similarly you can view the _last_ __n__ rows of your DataFrame with `.tail()` and you can retrieve an overall summary of the DataFrame with the `.info()` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the tail end of our data.\n",
    "# \"n = 7\" overwrites the default value of 5 rows\n",
    "\n",
    "aminoacids_df..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the information about our data frame\n",
    "aminoacids_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 4.4.0 Subset `DataFrame` and `Series` objects with the indexing `[ ]` operator\n",
    "\n",
    "Pandas objects can also be subset with the indexing `[ ]` operator but in a limited fashion depending on the object:\n",
    "\n",
    "- `Series`: `[label]` returns a scalar (single element) value matching the index label\n",
    "- `DataFrame`: `[colName]` returns an entire column\n",
    "\n",
    "\n",
    "We can pull information down using this notation in a number of ways like using the column names directly. Let's retrieve the column `Aminoacid` and see what kind of object is returned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take a look at just the first few entries of the Aminoacid column\n",
    "aminoacids_df[...].head()\n",
    "\n",
    "print(\"Indexing a colum name returns a:\")\n",
    "type(aminoacids_df[\"Aminoacid\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.4.1 Subsetting a `DataFrame` by column returns a `Series`\n",
    "\n",
    "A Pandas series is a 1D-labeled NumPy array that makes up rows and columns in data frames. Though they _inherit_ much of their structure from NumPy `array`, the Panda `Series` objects have their own attributes such as `.values`, which allows you to access the data contained in a `Series` but as a NumPy `ndarray` object.\n",
    "\n",
    "On the other hand, accessing a `Series` object with the indexing operators `[]` will return another `Series` object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the DNA codons colum\n",
    "codons_series = aminoacids_df[\"DNA codons\"]\n",
    "\n",
    "# Confirm that it does create a Series object\n",
    "type(codons_series)\n",
    "\n",
    "codons_series.head() # Notice that the indices are still inherited from the data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Access Series elements with the .values property\n",
    "codons_series.values[1:5]\n",
    "\n",
    "# .values returns a numpy.ndarray!\n",
    "type(codons_series.values[1:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directly access with an index\n",
    "codons_series[...]\n",
    "\n",
    "# Returns a Series object!\n",
    "type(codons_series[1:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.4.2 Subset multiple columns in a `DataFrame` by providing a `list`\n",
    "\n",
    "The values in a series, as well as data frame rows and columns, are 1D NumPy arrays. The behaviour of using `[ ]` is such that, by default of accessing a _single_ column in a `DataFrame`, a `Series` object is returned similar to what we saw in our last example above.\n",
    "\n",
    "If, however we want to access a sub-portion of a `DataFrame` we can also provide a list. Here's where the notation can get funny because we define a list using `[ ]` as well! Therefore to subset multiple columns we need to use syntax that looks like: \n",
    "\n",
    "`dataFrame[[\"colName1\", \"colName2\", \"colNameN\"]]`\n",
    "\n",
    "To top it all off, providing a `list` (regardless of size) to subset your `DataFrame` will __always__ return a `DataFrame` object. You can perform similar operations on a `Series` object as well!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Grab two columns from our data frame and look at just the beginning\n",
    "aminoacids_df...\n",
    "\n",
    "type(...)\n",
    "\n",
    "# What kind of object will be returned when asking for a single column?\n",
    "type(aminoacids_df[[\"Aminoacid\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directly access a Series with single or multiple index labels will return a Series!\n",
    "codons_series[[\"Arg\", \"Asn\"]]\n",
    "type(codons_series[[\"Arg\", \"Asn\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.4.3 Use the `.loc[]` and `.iloc[]` advanced data access methods for rows and columns\n",
    "\n",
    "As you can see from above, the standard `[ ]` notation can grant us access to parts of a `DataFrame` or `Series` but these aren't particularly optimized in their function. The `Pandas` package, however, provides optimized access using the `.loc` and `.iloc` methods which provide similar behaviour, so remember the difference between: \n",
    "- single `[row, col]` and double `[[rows], [cols]]`!\n",
    "\n",
    "| Method | Description | Examples to put within `[]`|\n",
    "| :- | :- | :- |\n",
    "|`.loc`| Used primarily for accessing using labels and will search for matches in this attribute | `['a', 'b', 'c']`\n",
    "||| `'a':'f'`|\n",
    "|`.iloc`| Used primarily for accessing with integer positions | `[4, 3, 0]`\n",
    "||| `1:7`|\n",
    "\n",
    "Output will depend on how many dimensions you have selected but generally a single-dimension is returned as a `Series` while multiple dimensions return a `DataFrame`. Note that both of these methods can also accept an array of booleans where `NA` is treated as `False`. So you have options but be sure to choose the correct one!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's remind ourselves what our DataFrame looks like\n",
    "aminoacids_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull down a subset based on the index name Ala\n",
    "aminoacids_df...\n",
    "\n",
    "# Indexing by a single dimension should return a Series\n",
    "type(...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Remember when you provide a list for indexing a `DataFrame` using the `[ ]` notation, it also returns a `DataFrame`. The same _mostly_ applies for the `loc` and `iloc` commands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return a Panda series broken into several lines with []\n",
    "aminoacids_df...\n",
    "\n",
    "# Return a Panda DataFrame with [[]]\n",
    "aminoacids_df.iloc[...]     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset rows and columns\n",
    "# Rows are a list, but column is not - return a series\n",
    "aminoacids_df.loc[[\"Ala\", \"Asn\"], \"Aminoacid\"] \n",
    "\n",
    "# Both rows and cols are in a list form - return a DataFrame\n",
    "aminoacids_df.loc[[\"Ala\", \"Asn\"], ...] \n",
    "\n",
    "# The equivalent DataFrame retrieval as above but using .iloc\n",
    "# The key to this is the extra brackets!\n",
    "aminoacids_df.iloc[[0,2], [0]] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.4.3.1 The `.loc[]` and `.iloc[]` methods can be used in multiple ways\n",
    "\n",
    "So far, we've been using our newfound methods to pull out single columns of data which may be composed of one or more rows. We can, however, subset dataframes that have multiple columns as well. We can effectively do this by subsetting multiple columns as would be expected. The syntax will look like the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset multiple rows and columns in a single index call\n",
    "aminoacids_df.loc[[\"Ala\", \"Asn\"], ...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, you can also chain-index in a similar fashion to achieve the same output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chained indexing gives us the same result but beware!\n",
    "# We are subsetting by index label AND THEN subsetting by col-name\n",
    "aminoacids_df.loc[[\"Ala\", \"Asn\"]] ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4.4.4 Is it a View or a Copy?\n",
    "\n",
    "Okay, it appears that we can use the `.loc[ ]` and `.iloc[]` to do almost anything but beware! We've been playing around a lot by pulling out sub-portions of our data frame. There is a lot happening under the hood but remember that we are dealing with __objects__! Depending on how we ask a Pandas object for access to it's data, it may be returning a ___view___ of the object or a ___copy___! \n",
    "\n",
    "In the above example we used `.loc[row, col]` notation to access our DataFrame but we also used a form of chain indexing with `.loc[row][col]`. We've seen this before in accessing arrays as well, the ability to subset _two_ ways. \n",
    "\n",
    "In the __first case__, we are ___multi-indexing___ by calling on a method and passing two parameters, row and column, to the method. For Python and the `DataFrame` object, it all happens in a single step by essentially going to that reference and pulling out what we want using the `DataFrame`'s internal methods.\n",
    "\n",
    "In the __second case__, we are ___chain-indexing___. Depending on a package's implementation this can give you very different results! In the case of a DataFrame, we are asking Python to first retrieve just the rows of the DataFrame we want. When that object is returned it needs to be assigned a temporary place in Python memory. At this point, it could become a separate object or entity from the original - a shallow copy! After that, we are then subsetting by `col` in a completely different command (by Python's viewpoint). If we were using these commands to set values in our DataFrame we could be setting them in an object that simply disappears!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/chain-indexing.jpg?raw=true\"  width=500>|\n",
    "| :--:|\n",
    "|Remember to try and use multi-indexing when possible. Chain-indexing may only bring sorrow.|\n",
    "\n",
    "Simply put, unless you have good reason to, with Pandas objects such as a DataFrame, __avoid chain-index notation__!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>Read more:</b> For some deeper examples of this problem, check out the <u><a = href\"https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-view-versus-copy\">Pandas documentation</a></u> on this specific issue.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 4.6.0 Subset `DataFrame` objects by their attributes\n",
    "\n",
    "That's right, although there are some limitations on just how well this can work, you can treat the columns of your `DataFrame` like an accessible property. You can do similar with a `Series` object.\n",
    "\n",
    "You'll run into problems and errors if you've used reserved keywords, other `Pandas` package-specific names, or if you break valid variable-naming convention but other than that, it is possible to use `.column` to access by non-numeric labels. \n",
    "\n",
    "Hint: Use tab-completion to see attributes of an object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Access a row from aminoacids_csv_df\n",
    "\n",
    "aminoacids_df..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 4.7.0 A summary of DataFrame Access methods\n",
    "\n",
    "We've covered a lot of different ways to pull data from your `DataFrame` objects. While not comprehensive, here's a list of the most common ways we'll access data from a `DataFrame`.<br>\n",
    "\n",
    "|Access Method| Description|Accepted Input|Output Type|\n",
    "|:-|:-|:-|:-|\n",
    "|`frame[column_name]`| Retrieve a single column | Existing column name | Series object |\n",
    "|`frame[[col1, col2, .. coln]]` | Retrieve multiple columns | Existing columnnames | DataFrame or Series (single column) |\n",
    "|`.loc[index_label, colname]`| Retrieve a subset of rows and columns by name | `start_label` | DataFrame |\n",
    "|||`start_label:end_label` (both inclusive)||\n",
    "|||`[list_of_labels]:[list_of_colnames]`||\n",
    "||| boolean array||\n",
    "|`.iloc[row_index, col_index]`| Retrieve a subset of rows and columns by position| `[row]` or `[row_start:row_end]`| DataFrame or Series (single column) |\n",
    "|||`[, col]` or `[, col_start:col_end]`||\n",
    "|||`[row_start:row_end, col_start:col_end]`||\n",
    "|||`[list_row_indices], [list_col_indices]`||\n",
    "||| boolean array||\n",
    "|`frame.column_name`| Retrieve a column based on it's name as an attribute| Existing column name | Series object | \n",
    "\n",
    "<br>We'll work on practicing these methods a lot more in the coming weeks!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\">\n",
    "    <b>Section 4.0.0 Comprehension Question:</b> Use the .iloc method to replicate the output of:\n",
    "    <p style=\"margin-left: 40px\"><b>aminoacids_df.loc[[\"Ala\", \"Ser\", \"Pyl\"]]</b><br>\n",
    "        <b>aminoacids_df.loc[[\"Ala\", \"Ser\", \"Pyl\"] , [\"Aminoacid\", \"DNA codons\"]]</b><br>\n",
    "        <b>aminoacids_df.loc[ : , [\"Aminoacid\", \"DNA codons\"]]</b> \n",
    "    </p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comprehension answer code 4.0.0\n",
    "# Replicate the output of the following .loc coding using the .iloc() method!\n",
    "aminoacids_df.loc[[\"Ala\", \"Ser\", \"Pyl\"]]\n",
    "aminoacids_df.loc[[\"Ala\", \"Ser\", \"Pyl\"] , [\"Aminoacid\", \"DNA codons\"]]\n",
    "aminoacids_df.loc[ : , [\"Aminoacid\", \"DNA codons\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 5.0.0 Class summary\n",
    "\n",
    "That's our second class on Python! You've made it through and we've learned about a lot of data structures:\n",
    "\n",
    "1. Lists\n",
    "2. Dictionaries (and a bit about sets!)\n",
    "3. Arrays\n",
    "4. Series and DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.0.1 A cheatsheet for data structures\n",
    "\n",
    "| Structure | Package | Description | Initializer(s) | Indexed? | Mutable? | Nestable? |\n",
    "| :- | :- | :- | :- | :- | :- | :- |\n",
    "| List | Python core | A 1D container of elements that can be made of anything | list() | Yes | Yes | Yes | \n",
    "|||| [value1, value2, ...] ||||\n",
    "| Tuple | Python core | A 1D container of elements that can be made of anything | tuple() | Yes | No | Yes | \n",
    "|||| (value1, value2, ...) ||||\n",
    "| Dictionary | Python core | A container of `key:value` pairs where keys must be immutable | dict() | No | keys = No | Yes | \n",
    "|||| {key1:value, key2:value2, ...} || values = Yes ||\n",
    "| Sets | Python core | A 1D container of unique, unordered, non-mutable elements | set() | No | Yes | No |\n",
    "|||| {value1, value2, ... } ||||\n",
    "| Array | Numpy | A multi-dimensional container of a single immutable data type| np.array() | Yes | Yes | NA | \n",
    "| Series | Pandas | A 1D container of a single immutable data type| pd.series() | Yes | size = No | No | \n",
    "|||||| values = Yes ||\n",
    "| DataFrame | Pandas | A 2D container of Series objects (columns)| pd.DataFrame(dict) | Yes | num rows = No* | Yes** | \n",
    "|||||| num cols = Yes ||\n",
    "|||||| values = Yes ||\n",
    "\n",
    "- \\* You can `.append()` rows to a DataFrame but this returns a new object whereas adding columns does not!\n",
    "- \\*\\* You can nest DataFrames objects but it's messy and there are perhaps better structures if you really need to do this. Reconsider how you have created your dataset!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 5.1.0 Submit your completed skeleton notebook (2% of final grade)\n",
    "\n",
    "At the end of this lecture a Quercus assignment portal will be available to submit your completed skeletons from today (including the comprehension question answers!). These will be due one week later, before the next lecture. Each lecture skeleton is worth 2% of your final grade but a __bonus 0.5%__ will also be awarded for submissions made within 24 hours from the end of lecture (ie 1600 hours the following day). \n",
    "\n",
    "1. From the Jupyter Notebook, select `File> Save Notebook` or use `ctrl + S`\n",
    "2. Return to the Jupyter folder page (ie click on the top-left \"jupyterhub\" icon)\n",
    "3. Locate your .ipynb file and click on the check box to its left side.\n",
    "4. If it is still active it will have a green icon beside its name. Above the file list will be option to `Shut Down`. After clicking on this, you may re-select the file and above the file list choose `Download` to save to your local hard drive.\n",
    "5. Upload your __`.ipynb`__ file to the appropriate Quercus assignment portal.\n",
    "\n",
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/save_ipynb.png?raw=true\" width=700>|\n",
    "|:--:|\n",
    "|A sample screen shot for saving your completed skeleton notebooks|\n",
    "\n",
    "## 5.2.0 Post-lecture DataCamp assessment (5% of final grade)\n",
    "\n",
    "Soon after the end of each lecture, a homework assignment will be available for you in [DataCamp](https://www.datacamp.com/). Your assignment is to complete chapters 2 (Python Lists, 1300 possible points) and 4 (NumPy, 1400 possible points) from the Introduction to Python course. __This is a pass-fail assignment__, and in order to pass you need to achieve a least 2025 points (75%) of the total possible points. Note that when you take hints from the DataCamp chapter, it will reduce your total earned points for that chapter.\n",
    "\n",
    "In order to properly assess your progress on DataCamp, at the end of each chapter, please print a PDF of the summary. You can do so by following these steps:\n",
    "\n",
    "1. Navigate to the **`Learn`** section along the top menu bar of DataCamp. This will bring up a sidebar menu where you can access the various courses you have been assigned under **`Assignments`** or you can scroll down on the page to find the **`My Assignments`** section. Click on the relevant chapter or course and this may bring you back to within the course itself. \n",
    "2. You should now be on a Course Summary page (See figure below). You can expand each __chapter__ of the course by clicking on the **`VIEW CHAPTER DETAILS`** link. ___DO THIS FOR ALL COMPLETED SECTIONS ON THE PAGE!___ \n",
    "3. Select the visible text in the course Window (not from any top or side menus) using a click and drag technique. Using something like __Ctrl+A will not properly select the course text__. \n",
    "4. Print the page from your browser menu and make sure you have \"Selection Only\" from the Settings options on the Print dialogue. Save as a single PDF. Note: if you don't select the text area correctly (at least in Google Chrome) you may not be able to print the full page. Your results should look something like this:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroPython/DataCamp.example.png?raw=true\" width=700>|\n",
    "|:--:|\n",
    "|A sample screen shot for one of the DataCamp assignments. You'll want to try and print off a single PDF of this section from **Learn > My Assignments**|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may need to take several screenshots if you cannot print it all in a single try. Submit the file(s) or a combined PDF for the homework to the assignment portal on Quercus. By submitting your scores for each section, and chapter, we can keep track of your progress, identify knowledge gaps, and produce a standardized way for you to check on your assignment \"grades\" throughout the course.\n",
    "\n",
    "You will have until 0959 hours on Tuesday, January 20th to submit your assignment (right before the next lecture)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "***\n",
    "## 5.3.0 Acknowledgements\n",
    "\n",
    "__Revision 1.0.0__: materials prepared by Oscar Montoya, M.Sc. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.1.0__: edited and prepared for __CSB1021H S LEC0140__, 06-2021 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.0__: edited and prepared for __CSB1021H S LEC0140__, 01-2022 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.1__: edited and prepared for __CSB1021H S LEC0140__, 01-2023 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.2__: edited and prepared for __CSB1021H S LEC0140__, 01-2024 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.3__: edited and prepared for __CSB1021H S LEC0140__, 01-2025 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._\n",
    "\n",
    "__Revision 1.2.4__: edited and prepared for __CSB1021H S LEC0140__, 01-2026 by Calvin Mok, Ph.D. _Bioinformatician, Education and Outreach, CAGEF._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 6.0.0 Appendix 1: More on slicing\n",
    "\n",
    "## 6.1.0 What happens when we slice a list?\n",
    "\n",
    "Here's a small example of list slicing and what we get back. Remember we talked about getting a view or copy back in section 4.5.4? \n",
    "\n",
    "Let's recall that every object is assigned an integer ID in Python. You can find this with the `id()` function. When we assign a variable to an object, Python is providing a link or ___reference___ back to the ID of that object. Python essentially uses the IDs to assign memory space to an object where it's __value__ can be stored. Python uses ___references___ to keep track of IDs so that when there are no more (0) references to an object ID, Python can reclaim the memory where its value is heald and its ID to use for something else in the program.\n",
    "\n",
    "Whenever we slice through a list, it will return a copy of the ___references___ to those elements. Rather than copying each element to a new space in memory (and object ID), it uses the reference to find where those individual objects are in memory. This can bring up some rather tricky concepts when we have nested lists that are made using other list object references and what happens to them.\n",
    "\n",
    "Depending on how you have sliced your original list, you may be getting back a direct reference (ID) to the original object, or references for the elements of the original list!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "\n",
    "# A crazy example of how list references work!\n",
    "\n",
    "list_1 = [\"genome\", 20, 30.5, True]\n",
    "list_2 = [list_1, \"this\", \"that\", \"those\"]\n",
    "\n",
    "# reference the first 3 elements of list_2 (a list, and two strings)\n",
    "list_3 = list_2[0:3] \n",
    "\n",
    "# reference the first 4 elements of list_1 (4 different references)\n",
    "list_4 = list_1[0:4]\n",
    "\n",
    "# reference the first 4 elements of the first element in list_2 (4 different references)\n",
    "list_5 = list_2[0][0:4]\n",
    "\n",
    "# reference the first element of list_2\n",
    "list_6 = list_2[0]\n",
    "\n",
    "# Copy all the references of list_2\n",
    "list_7 = list_2[:]\n",
    "\n",
    "# make a shallow copy of list_2\n",
    "list_8 = list_2.copy()\n",
    "\n",
    "# Use the copy.copy() function\n",
    "list_9 = copy.copy(list_2)\n",
    "\n",
    "# Use the copy.deepcopy() function\n",
    "list_10 = copy.deepcopy(list_2)\n",
    "\n",
    "# Change the first element of list_1\n",
    "list_1[0] = \"genomic\"\n",
    "\n",
    "# What is the resulting output?\n",
    "list_1\n",
    "list_2\n",
    "list_3\n",
    "list_4\n",
    "list_5\n",
    "list_6\n",
    "list_7\n",
    "list_8\n",
    "list_9\n",
    "list_10\n",
    "\n",
    "# Look at the corresponding IDs of these variables!\n",
    "id(list_1)\n",
    "id(list_2)\n",
    "id(list_3)\n",
    "id(list_4) \n",
    "id(list_5)\n",
    "id(list_6)\n",
    "\n",
    "# What is list_7? Is it it's own object?\n",
    "id(list_7)\n",
    "\n",
    "# What about the first element of list 7?\n",
    "id(list_7[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at what just happened so we can break down the code in a bit more detail. \n",
    "\n",
    "### 6.1.1 Assigning and slicing our lists\n",
    "\n",
    "1. `list_1 = [\"genome\", 20, 30.5, True]`: we generate a list with 4 elements.\n",
    "\n",
    "2. `list_2 = [list_1, \"this\", \"that\", \"those\"]`: we generate a second list of 4 elements with the first element being a list itself\n",
    "\n",
    "3. `list_3 = list_2[0:3]`: make a list copying the first 3 references to `list_2`; this includes the reference to the `list_1` object.\n",
    "\n",
    "4. `list_4 = list_1[0:4]`: make a list by copying references for the first 4 elements in `list_1`; we do not copy the reference to `list_1` object itself.\n",
    "\n",
    "5. `list_5 = list_2[0][0:4]`: make a list by copying references for the first 4 elements in the first element of `list_2`; it should be very similar to `list_4`.\n",
    "\n",
    "6. `list_6 = list_2[0]`: copy the reference to the first element of `list_2` which is the reference to the `list_1` object.\n",
    "\n",
    "7. `list_7 = list_2[:]`: copy all of the references to the elements of `list_2` which includes a reference to the `list_1` object.\n",
    "\n",
    "8. `list_8 = list_2.copy()`: Use the `list.copy()` method to create a shallow copy of the contents of `list_2` which includes a reference to the `list_1` object.\n",
    "\n",
    "9. `list_9 = copy.copy(list_2)`: Use the `copy.copy()` function to create a __shallow__ copy of `list_2` which includes a reference to the `list_1` object.\n",
    "\n",
    "10. `list_10 = copy.deepcopy(list_2)`: Use the `copy.deepcopy()` function to create a __deep__ copy of `list_2` which will duplicate new versions of all elements, _including_ making a copy in memory of `list_1`!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### 6.1.2 Make a change directly to `list_1`\n",
    "\n",
    "8. `list_1[0] = \"genomic\"`: Now we've changed the first element to `list_1`. Any other objects __directly__ referencing `list_1` __will__ propogate this change. Objects that merely reference the elements of `list_1` __will not__ propagate this change.\n",
    "\n",
    "### 6.1.3 What is the resulting output?\n",
    "\n",
    "Here's a summary of the list objects and their theoretical output\n",
    "\n",
    "| Object | Direct reference to `list_1`? | example object ID | Contents at assignment | Contents after changing list_1 |\n",
    "| :- | :- | :-: | :- | :- | \n",
    "| list_1 | Yes, of course! | 1 |`['genome', 20, 30.5, True]`|`['genomic', 20, 30.5, True]`|\n",
    "| list_2 | Yes, at index = 0 | 2 |`[['genome', 20, 30.5, True], 'this', 'that', 'those']`|`[['genomic', 20, 30.5, True], 'this', 'that', 'those']`|\n",
    "| list_3 | Yes, at index = 0 | 3 |`[['genome', 20, 30.5, True], 'this', 'that']`|`[['genomic', 20, 30.5, True], 'this', 'that']`|\n",
    "| list_4 | No | 4 |`['genome', 20, 30.5, True]`|`['genome', 20, 30.5, True]`|\n",
    "| list_5 | No | 5 |`['genome', 20, 30.5, True]`|`['genome', 20, 30.5, True]`|\n",
    "| list_6 | Yes, a direct reference | 1 |`['genome', 20, 30.5, True]`|`['genomic', 20, 30.5, True]`|\n",
    "| list_7 | Yes, at index = 0 | 6 |`[['genome', 20, 30.5, True], 'this', 'that', 'those']`|`[['genomic', 20, 30.5, True], 'this', 'that', 'those']`|\n",
    "| list_8 | Yes, at index = 0 | 7 |`[['genome', 20, 30.5, True], 'this', 'that', 'those']`|`[['genomic', 20, 30.5, True], 'this', 'that', 'those']`|\n",
    "| list_9 | Yes, at index = 0 | 8 |`[['genome', 20, 30.5, True], 'this', 'that', 'those']`|`[['genomic', 20, 30.5, True], 'this', 'that', 'those']`|\n",
    "| list_10 | __No__, all references are new objects | 9 |`[['genome', 20, 30.5, True], 'this', 'that', 'those']`|__`[['genome', 20, 30.5, True], 'this', 'that', 'those']`__|\n",
    "\n",
    "### 6.1.4 Know when you are viewing or copying when slicing\n",
    "\n",
    "So we've taken a close look at lists and how they handle slicing. Remember to avoid getting a direct reference (view) for a simple list object when using the `=` assignment operator, you can use `[:]` or the `.copy()` method to copy all of the element references. When working with nested lists which contain references to other lists, however, beware of what you'll get. To get a fully independent copy, use the `copy.deepcopy()` function instead.\n",
    "\n",
    "From our above examples, what happens if we assigned `list_6[0] = \"GENOMIC\"`. Can you guess?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## 6.2.0 What happens when we slice a NumPy array?\n",
    "\n",
    "Now that we've covered lists and their potential complications, should we take a closer look at NumPy arrays? Similar in concept to lists, recall that these objects are implemented in a package independent of the Python kernel development."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The double squared bracket is required, otherwise you get a \"type not understood\" traceback\n",
    "\n",
    "# First build our 2D array: 2 rows, 3 columns\n",
    "array_1 = np.array([[5, 16, 23],   \n",
    "                   [3, 19, 15]])   \n",
    "\n",
    "# Slice all of the elements and assign to array_2\n",
    "array_2 = array_1[:, :]\n",
    "\n",
    "# Slice a subset of array_2. Is it a copy of the references?\n",
    "array_3 = array_2[0:2, 0:2]\n",
    "\n",
    "# Make a copy of array_3\n",
    "array_4 = array_3\n",
    "\n",
    "# Make a shallow copy of array_1\n",
    "array_5 = array_1.copy()\n",
    "\n",
    "# Change the first element of array_4\n",
    "array_4[0,0] = 50 \n",
    "\n",
    "# Multiply all of array_3 with 2\n",
    "array_3 = array_3 * 2\n",
    "\n",
    "# What do array_2 and array_3 look like?\n",
    "array_2\n",
    "array_3\n",
    "array_4\n",
    "array_5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 6.2.1 Slicing an array with `[ ]` returns a reference to the original array\n",
    "\n",
    "So, the way arrays handle slicing is different to how lists handle the same commands. From the simple example above, we see that the underlying behaviour is that when you slice an array with `[]` you are _always_ returned a reference or view of that array. That being said, altering an array slice's reference can be quirky. \n",
    "\n",
    "1. `array_4[0,0] = 50` is a direct slicing call with an assignment and so the change is propagated to `array_1` and `array_2`\n",
    "\n",
    "\n",
    "2. `array_3 = array_3 * 2` does not use any slicing notation. Instead the Python interpreter makes a _new copy_ of array_3, completes the math on the array and then _assigns_ this to the variable `array_3`. The reference that `array_3` had to `array_1` is released and replaced with one for this new object.\n",
    "\n",
    "These are intentional behaviours of the array object. If a reference to an array is not what you want, then you can use the `.copy()` method to generate a shallow copy of the array. Any changes in the original will not be propagated to the copy and vice versa!\n",
    "\n",
    "What do you think `array_3[:,:] = array_3 * 2` would do?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# 7.0.0 Appendix 2: Tuples are immutable lists"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|<img src=\"https://github.com/camok/CSB_Course_Materials/blob/main/IntroPython/LordOfTuples.png?raw=true\"  width=700>|\n",
    "|:--:|\n",
    "|Tuples are less flexible but faster versions of lists.|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've already mentioned the concept of tuples but haven't discussed clearly what these objects are. Also known as structured arrays, the tuple is simply put, an immutable list. Let's compare them to lists for a better understanding.\n",
    "\n",
    "| Feature | List | Tuple |\n",
    "| :-: | :-: | :-: |\n",
    "| Can have any object as an element |$$\\checkmark$$|$$\\checkmark$$|\n",
    "| Indexed |$$\\checkmark$$|$$\\checkmark$$|\n",
    "| Notation to initialize |`[ ]`|`( )`|\n",
    "| Mutable |$$\\checkmark$$|$$X$$|\n",
    "| Useable as a dictionary key? |$$X$$|$$\\checkmark$$|\n",
    "\n",
    "- Tuples have fewer methods (since they are immutable).\n",
    "- Tuples are _faster_ which makes a BIG difference on large sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# When creating a one-element tuple, the comma after the element is required\n",
    "# Otherwise it makes a string.\n",
    "\n",
    "tuple_string = \"a\"\n",
    "type(tuple_string)\n",
    "\n",
    "# equivalent\n",
    "tuple_string = (\"a\")\n",
    "type(tuple_string)\n",
    "\n",
    "# What does tuple_string look like?\n",
    "tuple_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a single tuple now\n",
    "tuple_single = \"a\",\n",
    "type(tuple_single)\n",
    "\n",
    "tuple_single = (\"a\",)\n",
    "type(tuple_single)\n",
    "\n",
    "tuple_single"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Time to make a tuple with multiple elements:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a tuple with multiple entries\n",
    "tuple_aminoacids = (\"Alanine\", \"ala\", \"A\", \"GCA GCC GCG GCU\", \"A\")\n",
    "\n",
    "tuple_aminoacids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you try to change an element in a tuple, you get a traceback (Python error). Let's make Alanine all lowercase:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tuple_aminoacids[0]\n",
    "\n",
    "tuple_aminoacids[0] = \"alanine\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 7.0.1 Initialize using the `tuple()` function\n",
    "\n",
    "Like dictionaries, the immutability of tuples makes them good candidates to reliably store data that you do not want to be changed (like by accident). Similarly to lists, the function `tuple()` will split a single string into its component characters and will store them as elements:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tuple_alanine = tuple(\"Alanine\")\n",
    "\n",
    "print(tuple_alanine)\n",
    "print(type(tuple_alanine))\n",
    "len(tuple_alanine)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7.1.0 Tuples only have two methods: `.count()` and `.index()`\n",
    "\n",
    "As mentioned above, given the immutablitiy of the tuple, it doesn't need some of the same methods as a list. As a consequence, tuples have only two methods: `.count()` and `.index()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir(tuple_aminoacids)[-2]     # Tuple methods\n",
    "\n",
    "tuple_aminoacids\n",
    "\n",
    "tuple_aminoacids.index(\"ala\")     # Ala is at index position 1\n",
    "\n",
    "tuple_aminoacids.count(\"A\")    # There are two instances of the string \"A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1.1 Work around immutability with lists and the `.append()` method\n",
    "\n",
    "Methods such as `.append()` do not operate on tuples because of their immutability. A workaround to append elements to a tuple is to first create a list, append elements to it, and then use type conversion to make a tuple."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a nested list\n",
    "list_aminoacids = [[\"Alanine\", \"Ala\", \"A\",  \"GCA GCC GCG GCU\"],\n",
    "                    [\"Cysteine\", \"Cys\", \"C\", \"UGC, UGU\"],\n",
    "                    [\"Aspartic acid\", \"Asp\", \"D\", \"GAC GAU\"],\n",
    "                    [\"Glutamic acid\", \"Glu\", \"E\", \"GAA GAG\"]                    \n",
    "                   ]\n",
    "\n",
    "list_aminoacids\n",
    "\n",
    "# Append the desired elements to the list \n",
    "list_aminoacids.append([\"Phenylalanine\", \"Phe\", \"F\", \"UUC UUU\"])      \n",
    "\n",
    "print() # Use empty Prints to make the output more readable\n",
    "list_aminoacids\n",
    "\n",
    "# Type convert the list into a tuple\n",
    "tuple_aminoacids = tuple(list_aminoacids)\n",
    "\n",
    "print()\n",
    "tuple_aminoacids # Note the output difference with lists is () vs []\n",
    "type(tuple_aminoacids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "As you can see, post-conversion our tuple elements are output with `( )` surrounding the group vs a list which used the `[ ]` notation.\n",
    "\n",
    "## 7.2.0 Retrieve elements from a tuple by index\n",
    "\n",
    "Accessing tuple elements is the same as a list and you can use the `[]` notation to assign elements to individual variables one by one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_1 = tuple_aminoacids[0]\n",
    "var_2 = tuple_aminoacids[1]\n",
    "\n",
    "print(var_1, var_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also use multiple assignment (remember lecture 1?) to assign several variables at once. Just be sure that each side balances out! Note that you can also use slice notation on a tuple."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_a, var_b, var_c, var_d, var_e, var_f = tuple_alanine[0:6]\n",
    "\n",
    "# var_a, var_b, var_c, var_d, var_e, var_f = tuple_alanine\n",
    "\n",
    "print(var_a, var_b, var_c, var_d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "<b>Python trick: </b>Variables can be swapped with the syntax a, b = b, a. That will assign a to b and b to a.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## The Center for the Analysis of Genome Evolution and Function (CAGEF)\n",
    "\n",
    "The Centre for the Analysis of Genome Evolution and Function (CAGEF) at the University of Toronto offers comprehensive experimental design, research, and analysis services in microbiome and metagenomic studies, genomics, proteomics, and bioinformatics.\n",
    "\n",
    "From targeted DNA amplicon sequencing to transcriptomes, whole genomes, and metagenomes, from protein identification to post-translational modification, CAGEF has the tools and knowledge to support your research. Our state-of-the-art facility and experienced research staff provide a broad range of services, including both standard analyses and techniques developed by our team. In particular, we have special expertise in microbial, plant, and environmental systems.\n",
    "\n",
    "For more information about us and the services we offer, please visit <https://www.cagef.utoronto.ca/>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center><img src=\"https://github.com/uoft-csb-datasci/CSBdatasci_Course_Materials/blob/main/IntroR/CAGEF_new.png?raw=true\" width=700></center>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "414.986px",
    "width": "160px"
   },
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "408.082px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
